<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Distributions – Bayesian Inference (BI)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./api_diag.html" rel="next">
<link href="./api_manip.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="stylesheet" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@latest/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="MyStyles.css">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Bayesian Inference (BI)</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./start/Installation.html"> 
<span class="menu-text">Get started</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./0. Introduction.html"> 
<span class="menu-text">Models</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./api_dist.html" aria-current="page"> 
<span class="menu-text">API Reference</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-help" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Help</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-help">    
        <li>
    <a class="dropdown-item" href="https://github.com/BGN-for-ASNA/BI/issues"><i class="bi bi-bug" role="img">
</i> 
 <span class="dropdown-text">Report a Bug</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://github.com/BGN-for-ASNA/BI/discussions"><i class="bi bi-chat-dots" role="img">
</i> 
 <span class="dropdown-text">Ask a Question</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://twitter.com" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/BGN-for-ASNA/BI" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./api_dist.html">Distributions</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./api_manip.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Handle data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./api_dist.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Distributions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./api_diag.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Diagnostics</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#asymmetric-laplace" id="toc-asymmetric-laplace" class="nav-link active" data-scroll-target="#asymmetric-laplace">Asymmetric Laplace</a></li>
  <li><a href="#asymmetric-laplace-quantile" id="toc-asymmetric-laplace-quantile" class="nav-link" data-scroll-target="#asymmetric-laplace-quantile">Asymmetric Laplace Quantile</a></li>
  <li><a href="#bernoulli" id="toc-bernoulli" class="nav-link" data-scroll-target="#bernoulli">Bernoulli</a></li>
  <li><a href="#bernoulli-logits" id="toc-bernoulli-logits" class="nav-link" data-scroll-target="#bernoulli-logits">Bernoulli Logits</a></li>
  <li><a href="#bernoulli-probs" id="toc-bernoulli-probs" class="nav-link" data-scroll-target="#bernoulli-probs">Bernoulli Probs</a></li>
  <li><a href="#beta" id="toc-beta" class="nav-link" data-scroll-target="#beta">Beta</a></li>
  <li><a href="#betabinomial" id="toc-betabinomial" class="nav-link" data-scroll-target="#betabinomial">BetaBinomial</a></li>
  <li><a href="#beta-proportion" id="toc-beta-proportion" class="nav-link" data-scroll-target="#beta-proportion">Beta Proportion</a></li>
  <li><a href="#binomial" id="toc-binomial" class="nav-link" data-scroll-target="#binomial">Binomial</a></li>
  <li><a href="#binomial-logits" id="toc-binomial-logits" class="nav-link" data-scroll-target="#binomial-logits">Binomial Logits</a></li>
  <li><a href="#binomial-probs" id="toc-binomial-probs" class="nav-link" data-scroll-target="#binomial-probs">Binomial Probs</a></li>
  <li><a href="#conditional-autoregressive-car" id="toc-conditional-autoregressive-car" class="nav-link" data-scroll-target="#conditional-autoregressive-car">Conditional Autoregressive (CAR)</a></li>
  <li><a href="#categorical-distribution." id="toc-categorical-distribution." class="nav-link" data-scroll-target="#categorical-distribution.">Categorical distribution.</a></li>
  <li><a href="#categorical-logits" id="toc-categorical-logits" class="nav-link" data-scroll-target="#categorical-logits">Categorical Logits</a></li>
  <li><a href="#categorical-probs-distribution." id="toc-categorical-probs-distribution." class="nav-link" data-scroll-target="#categorical-probs-distribution.">Categorical Probs distribution.</a></li>
  <li><a href="#cauchy" id="toc-cauchy" class="nav-link" data-scroll-target="#cauchy">Cauchy</a></li>
  <li><a href="#chi-squared" id="toc-chi-squared" class="nav-link" data-scroll-target="#chi-squared">Chi-squared</a></li>
  <li><a href="#circulant-normal-multivariate-normal" id="toc-circulant-normal-multivariate-normal" class="nav-link" data-scroll-target="#circulant-normal-multivariate-normal">Circulant Normal Multivariate normal</a></li>
  <li><a href="#delta" id="toc-delta" class="nav-link" data-scroll-target="#delta">Delta</a></li>
  <li><a href="#dirichlet" id="toc-dirichlet" class="nav-link" data-scroll-target="#dirichlet">Dirichlet</a></li>
  <li><a href="#dirichlet-multinomial" id="toc-dirichlet-multinomial" class="nav-link" data-scroll-target="#dirichlet-multinomial">Dirichlet-Multinomial</a></li>
  <li><a href="#discrete-uniform" id="toc-discrete-uniform" class="nav-link" data-scroll-target="#discrete-uniform">Discrete Uniform</a></li>
  <li><a href="#doubly-truncated-power-law" id="toc-doubly-truncated-power-law" class="nav-link" data-scroll-target="#doubly-truncated-power-law">Doubly Truncated Power Law</a></li>
  <li><a href="#eulermaruyama" id="toc-eulermaruyama" class="nav-link" data-scroll-target="#eulermaruyama">Euler–Maruyama</a></li>
  <li><a href="#exponential" id="toc-exponential" class="nav-link" data-scroll-target="#exponential">Exponential</a></li>
  <li><a href="#folded" id="toc-folded" class="nav-link" data-scroll-target="#folded">Folded</a></li>
  <li><a href="#gamma" id="toc-gamma" class="nav-link" data-scroll-target="#gamma">Gamma</a></li>
  <li><a href="#gamma-poisson" id="toc-gamma-poisson" class="nav-link" data-scroll-target="#gamma-poisson">Gamma Poisson</a></li>
  <li><a href="#gaussian-copula" id="toc-gaussian-copula" class="nav-link" data-scroll-target="#gaussian-copula">Gaussian Copula</a></li>
  <li><a href="#gaussian-copula-beta" id="toc-gaussian-copula-beta" class="nav-link" data-scroll-target="#gaussian-copula-beta">Gaussian Copula Beta</a></li>
  <li><a href="#gaussian-random-walk" id="toc-gaussian-random-walk" class="nav-link" data-scroll-target="#gaussian-random-walk">Gaussian Random Walk</a></li>
  <li><a href="#gaussian-state-space" id="toc-gaussian-state-space" class="nav-link" data-scroll-target="#gaussian-state-space">Gaussian State Space</a></li>
  <li><a href="#geometric-distribution." id="toc-geometric-distribution." class="nav-link" data-scroll-target="#geometric-distribution.">Geometric distribution.</a></li>
  <li><a href="#geometriclogits" id="toc-geometriclogits" class="nav-link" data-scroll-target="#geometriclogits">GeometricLogits</a></li>
  <li><a href="#geometricprobs" id="toc-geometricprobs" class="nav-link" data-scroll-target="#geometricprobs">GeometricProbs</a></li>
  <li><a href="#gompertz" id="toc-gompertz" class="nav-link" data-scroll-target="#gompertz">Gompertz</a></li>
  <li><a href="#gumbel" id="toc-gumbel" class="nav-link" data-scroll-target="#gumbel">Gumbel</a></li>
  <li><a href="#halfcauchy" id="toc-halfcauchy" class="nav-link" data-scroll-target="#halfcauchy">HalfCauchy</a></li>
  <li><a href="#halfnormal" id="toc-halfnormal" class="nav-link" data-scroll-target="#halfnormal">HalfNormal</a></li>
  <li><a href="#improper-uniform" id="toc-improper-uniform" class="nav-link" data-scroll-target="#improper-uniform">Improper Uniform</a></li>
  <li><a href="#inverse-gamma" id="toc-inverse-gamma" class="nav-link" data-scroll-target="#inverse-gamma">Inverse Gamma</a></li>
  <li><a href="#kumaraswamy" id="toc-kumaraswamy" class="nav-link" data-scroll-target="#kumaraswamy">Kumaraswamy</a></li>
  <li><a href="#laplace" id="toc-laplace" class="nav-link" data-scroll-target="#laplace">Laplace</a></li>
  <li><a href="#left-truncated" id="toc-left-truncated" class="nav-link" data-scroll-target="#left-truncated">Left Truncated</a></li>
  <li><a href="#levy" id="toc-levy" class="nav-link" data-scroll-target="#levy">Levy</a></li>
  <li><a href="#lewandowski-kurowicka-joe-lkj" id="toc-lewandowski-kurowicka-joe-lkj" class="nav-link" data-scroll-target="#lewandowski-kurowicka-joe-lkj">Lewandowski Kurowicka Joe (LKJ)</a></li>
  <li><a href="#lkj-cholesky" id="toc-lkj-cholesky" class="nav-link" data-scroll-target="#lkj-cholesky">LKJ Cholesky</a></li>
  <li><a href="#log-normal" id="toc-log-normal" class="nav-link" data-scroll-target="#log-normal">Log-Normal</a></li>
  <li><a href="#log-uniform" id="toc-log-uniform" class="nav-link" data-scroll-target="#log-uniform">Log-Uniform</a></li>
  <li><a href="#logistic" id="toc-logistic" class="nav-link" data-scroll-target="#logistic">Logistic</a></li>
  <li><a href="#low-rank-multivariate-normal" id="toc-low-rank-multivariate-normal" class="nav-link" data-scroll-target="#low-rank-multivariate-normal">Low Rank Multivariate Normal</a></li>
  <li><a href="#lower-truncated-power-law" id="toc-lower-truncated-power-law" class="nav-link" data-scroll-target="#lower-truncated-power-law">Lower Truncated Power Law</a></li>
  <li><a href="#matrix-normal" id="toc-matrix-normal" class="nav-link" data-scroll-target="#matrix-normal">Matrix Normal</a></li>
  <li><a href="#a-marginalized-finite-mixture-of-component-distributions." id="toc-a-marginalized-finite-mixture-of-component-distributions." class="nav-link" data-scroll-target="#a-marginalized-finite-mixture-of-component-distributions.">A marginalized finite mixture of component distributions.</a></li>
  <li><a href="#mixture-general" id="toc-mixture-general" class="nav-link" data-scroll-target="#mixture-general">Mixture General</a></li>
  <li><a href="#finite-mixture-of-component-distributions-from-the-same-family." id="toc-finite-mixture-of-component-distributions-from-the-same-family." class="nav-link" data-scroll-target="#finite-mixture-of-component-distributions-from-the-same-family.">Finite mixture of component distributions from the same family.</a></li>
  <li><a href="#multinomial" id="toc-multinomial" class="nav-link" data-scroll-target="#multinomial">Multinomial</a></li>
  <li><a href="#multinomial-logits" id="toc-multinomial-logits" class="nav-link" data-scroll-target="#multinomial-logits">Multinomial Logits</a></li>
  <li><a href="#multinomial-probs" id="toc-multinomial-probs" class="nav-link" data-scroll-target="#multinomial-probs">Multinomial Probs</a></li>
  <li><a href="#multivariate-normal" id="toc-multivariate-normal" class="nav-link" data-scroll-target="#multivariate-normal">Multivariate Normal</a></li>
  <li><a href="#multivariate-students-t" id="toc-multivariate-students-t" class="nav-link" data-scroll-target="#multivariate-students-t">Multivariate Student’s t</a></li>
  <li><a href="#negative-binomial" id="toc-negative-binomial" class="nav-link" data-scroll-target="#negative-binomial">Negative Binomial</a></li>
  <li><a href="#section" id="toc-section" class="nav-link" data-scroll-target="#section"></a></li>
  <li><a href="#negative-binomial-logits" id="toc-negative-binomial-logits" class="nav-link" data-scroll-target="#negative-binomial-logits">Negative Binomial Logits</a></li>
  <li><a href="#negative-binomial-with-probabilities." id="toc-negative-binomial-with-probabilities." class="nav-link" data-scroll-target="#negative-binomial-with-probabilities.">Negative Binomial with probabilities.</a></li>
  <li><a href="#normal" id="toc-normal" class="nav-link" data-scroll-target="#normal">Normal</a></li>
  <li><a href="#ordered-logistic" id="toc-ordered-logistic" class="nav-link" data-scroll-target="#ordered-logistic">Ordered Logistic</a></li>
  <li><a href="#pareto" id="toc-pareto" class="nav-link" data-scroll-target="#pareto">Pareto</a></li>
  <li><a href="#poisson" id="toc-poisson" class="nav-link" data-scroll-target="#poisson">Poisson</a></li>
  <li><a href="#projected-normal" id="toc-projected-normal" class="nav-link" data-scroll-target="#projected-normal">Projected Normal</a></li>
  <li><a href="#relaxed-bernoulli" id="toc-relaxed-bernoulli" class="nav-link" data-scroll-target="#relaxed-bernoulli">Relaxed Bernoulli</a></li>
  <li><a href="#relaxed-bernoulli-logits" id="toc-relaxed-bernoulli-logits" class="nav-link" data-scroll-target="#relaxed-bernoulli-logits">Relaxed Bernoulli Logits</a></li>
  <li><a href="#right-truncated" id="toc-right-truncated" class="nav-link" data-scroll-target="#right-truncated">Right Truncated</a></li>
  <li><a href="#sine-bivariate-von-mises" id="toc-sine-bivariate-von-mises" class="nav-link" data-scroll-target="#sine-bivariate-von-mises">Sine Bivariate Von Mises</a></li>
  <li><a href="#sine-skewing" id="toc-sine-skewing" class="nav-link" data-scroll-target="#sine-skewing">Sine-skewing</a></li>
  <li><a href="#softlaplace" id="toc-softlaplace" class="nav-link" data-scroll-target="#softlaplace">SoftLaplace</a></li>
  <li><a href="#students-t" id="toc-students-t" class="nav-link" data-scroll-target="#students-t">Student’s t</a></li>
  <li><a href="#truncated-cauchy" id="toc-truncated-cauchy" class="nav-link" data-scroll-target="#truncated-cauchy">Truncated Cauchy</a></li>
  <li><a href="#truncated" id="toc-truncated" class="nav-link" data-scroll-target="#truncated">Truncated</a></li>
  <li><a href="#truncated-normal" id="toc-truncated-normal" class="nav-link" data-scroll-target="#truncated-normal">Truncated Normal</a></li>
  <li><a href="#truncated-polyagamma" id="toc-truncated-polyagamma" class="nav-link" data-scroll-target="#truncated-polyagamma">Truncated PolyaGamma</a></li>
  <li><a href="#two-sided-truncated" id="toc-two-sided-truncated" class="nav-link" data-scroll-target="#two-sided-truncated">Two Sided Truncated</a></li>
  <li><a href="#uniform" id="toc-uniform" class="nav-link" data-scroll-target="#uniform">Uniform</a></li>
  <li><a href="#unit" id="toc-unit" class="nav-link" data-scroll-target="#unit">Unit</a></li>
  <li><a href="#weibull" id="toc-weibull" class="nav-link" data-scroll-target="#weibull">Weibull</a></li>
  <li><a href="#wishart" id="toc-wishart" class="nav-link" data-scroll-target="#wishart">Wishart</a></li>
  <li><a href="#wishart-cholesky" id="toc-wishart-cholesky" class="nav-link" data-scroll-target="#wishart-cholesky">Wishart Cholesky</a></li>
  <li><a href="#generic-zero-inflated" id="toc-generic-zero-inflated" class="nav-link" data-scroll-target="#generic-zero-inflated">Generic Zero Inflated</a></li>
  <li><a href="#zero-inflated-negative-binomial" id="toc-zero-inflated-negative-binomial" class="nav-link" data-scroll-target="#zero-inflated-negative-binomial">Zero-Inflated Negative Binomial</a></li>
  <li><a href="#a-zero-inflated-poisson" id="toc-a-zero-inflated-poisson" class="nav-link" data-scroll-target="#a-zero-inflated-poisson">A Zero Inflated Poisson</a></li>
  <li><a href="#zero-sum-normal" id="toc-zero-sum-normal" class="nav-link" data-scroll-target="#zero-sum-normal">Zero Sum Normal</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="api_dist.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Distributions</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p><code>Utils.np_dists.UnifiedDist</code> is a class to unify various distribution methods and provide a consistent interface for sampling and inference.</p>
<section id="asymmetric-laplace" class="level3">
<h3 class="anchored" data-anchor-id="asymmetric-laplace">Asymmetric Laplace</h3>
<p>Samples from an Asymmetric Laplace distribution.</p>
<p>The Asymmetric Laplace distribution is a generalization of the Laplace distribution, where the two sides of the distribution are scaled differently. It is defined by a location parameter (loc), a scale parameter (scale), and an asymmetry parameter (asymmetry).</p>
<p><span class="math display">
f(x, \kappa) = \frac{1}{\kappa+\kappa^{-1}}\exp(-x\kappa),\quad x\ge0\\
= \frac{1}{\kappa+\kappa^{-1}}\exp(x/\kappa),\quad x&lt;0\\
</span></p>
<p>for <span class="math inline">-\infty &lt; x &lt; \infty, \kappa &gt; 0</span>.</p>
<p>laplace_asymmetric takes ‘kappa’ as a shape parameter for <span class="math inline">\kappa</span>. For :<span class="math inline">\kappa = 1</span>, it is identical to a Laplace distribution</p>
<section id="args" class="level4">
<h4 class="anchored" data-anchor-id="args">Args:</h4>
<div class="sourceCode" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>bi.dist.asymmetric_laplace(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>asymmetry<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray or float): Location parameter of the distribution.</p></li>
<li><p><em>scale</em> (jnp.ndarray or float): Scale parameter of the distribution.</p></li>
<li><p><em>asymmetry</em> (jnp.ndarray or float): Asymmetry parameter of the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
<li><p>validate_args (bool, optional): Whether to enable validation of distribution parameters. Defaults to <code>None</code>.</p></li>
</ul>
</section>
<section id="returns" class="level4">
<h4 class="anchored" data-anchor-id="returns">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI AsymmetricLaplace distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the AsymmetricLaplace distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI AsymmetricLaplace distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage" class="level4">
<h4 class="anchored" data-anchor-id="example-usage">Example Usage:</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>m.dist.asymmetric_laplace(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, asymmetry<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#asymmetriclaplace</p>
</section>
<section id="references" class="level4">
<h4 class="anchored" data-anchor-id="references">References</h4>
<p>[1] “Asymmetric Laplace distribution”, Wikipedia https://en.wikipedia.org/wiki/Asymmetric_Laplace_distribution</p>
<p>[2] Kozubowski TJ and Podgórski K. A Multivariate and Asymmetric Generalization of Laplace Distribution, Computational Statistics 15, 531–540 (2000). :doi:<code>10.1007/PL00022717</code></p>
<p>–</p>
</section>
</section>
<section id="asymmetric-laplace-quantile" class="level3">
<h3 class="anchored" data-anchor-id="asymmetric-laplace-quantile">Asymmetric Laplace Quantile</h3>
<p>Samples from an AsymmetricLaplaceQuantile distribution.</p>
<p>This distribution is an alternative parameterization of the AsymmetricLaplace distribution, commonly used in Bayesian quantile regression. It utilizes a <code>quantile</code> parameter to define the balance between the left- and right-hand sides of the distribution, representing the proportion of probability density that falls to the left-hand side.</p>
<section id="args-1" class="level4">
<h4 class="anchored" data-anchor-id="args-1">Args:</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>bi.dist.asymmetric_laplace_quantile(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>quantile<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (float): The location parameter of the distribution.</p></li>
<li><p><em>sample</em> (float): The scale parameter of the distribution.</p></li>
<li><p><em>quantile</em> (float): The quantile parameter, representing the proportion of probability density to the left of the median. Must be between 0 and 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-1" class="level4">
<h4 class="anchored" data-anchor-id="returns-1">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI AsymmetricLaplaceQuantile distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the AsymmetricLaplaceQuantile distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI AsymmetricLaplaceQuantile distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-1" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-1">Example Usage:</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>m.dist.asymmetric_laplace_quantile(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, quantile<span class="op">=</span><span class="fl">0.5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-1" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-1">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#asymmetriclaplacequantile</p>
<hr>
</section>
</section>
<section id="bernoulli" class="level3">
<h3 class="anchored" data-anchor-id="bernoulli">Bernoulli</h3>
<p>The Bernoulli distribution models a single trial with two possible outcomes: success or failureIt models a single trial with exactly two possible outcomes:</p>
<ul>
<li><strong>Success</strong> (often coded as 1) with probability (p),</li>
<li><strong>Failure</strong> (often coded as 0) with probability (1 - p).</li>
</ul>
<p>It is a building block for many more complex discrete distributions (e.g.&nbsp;binomial, geometric).</p>
<p>We write <span class="math display">
X \sim {Bernoulli}(p)
</span> with <span class="math inline">0 \le p \le 1</span>.</p>
<hr>
<section id="probability-mass-function-pmf" class="level4">
<h4 class="anchored" data-anchor-id="probability-mass-function-pmf">Probability Mass Function (PMF)</h4>
<p><span class="math display">
P(X = x) =
\begin{cases}
p,  \text{if } x = 1, \
1 - p,  \text{if } x = 0, \
0,  \text{otherwise.}
\end{cases}
</span></p>
<p>Alternatively, using an “indicator” exponent form:</p>
<p><span class="math display">
P(X = x) = p^x (1 - p)^{1 - x}, \quad x \in {0,1}.
</span></p>
</section>
<section id="args-2" class="level4">
<h4 class="anchored" data-anchor-id="args-2">Args:</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>bi.dist.bernoulli(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray, optional): Probability of success for each Bernoulli trial. Must be between 0 and 1.</p></li>
<li><p><em>logits</em> (jnp.ndarray, optional): Log-odds of success for each Bernoulli trial. <code>probs = sigmoid(logits)</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-2" class="level4">
<h4 class="anchored" data-anchor-id="returns-2">Returns:</h4>
<p>BI Bernoulli distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the Bernoulli distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-2" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-2">Example Usage:</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>m.dist.bernoulli(probs<span class="op">=</span><span class="fl">0.7</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlbernoulli" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlbernoulli">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#bernoulli</h4>
</section>
<section id="references-1" class="level4">
<h4 class="anchored" data-anchor-id="references-1">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Bernoulli_distribution “Bernoulli distribution”</li>
<li>https://stats.libretexts.org/Courses/Saint_Mary%27s_College_Notre_Dame/MATH_345__-<em>Probability</em>%28Kuter%29/3%3A_Discrete_Random_Variables/3.3%3A_Bernoulli_and_Binomial_Distributions “3.3: Bernoulli and Binomial Distributions - Statistics LibreTexts”</li>
</ul>
<hr>
</section>
</section>
<section id="bernoulli-logits" class="level3">
<h3 class="anchored" data-anchor-id="bernoulli-logits">Bernoulli Logits</h3>
<p>Samples from a Bernoulli distribution parameterized by logits.</p>
<p>The Bernoulli distribution models a single binary event (success or failure), parameterized by the log-odds ratio of success. The probability of success is given by the sigmoid function applied to the logit.</p>
<p><span class="math display">
p = \sigma(\eta) = \frac{1}{1 + e^{-\eta}}
</span></p>
<p>where () is the log-odds (the <em>logit</em>).</p>
<section id="args-3" class="level4">
<h4 class="anchored" data-anchor-id="args-3">Args:</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>bi.dist.bernoulli_logits(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>logits</em> (jnp.ndarray, optional): Log-odds ratio of success. Must be real-valued.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as <em>event</em> dimensions (used in model building). Defaults to 0.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. Defaults to None.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-3" class="level4">
<h4 class="anchored" data-anchor-id="returns-3">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BernoulliLogits distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-3" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-3">Example Usage:</h4>
<div class="sourceCode" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>m.dist.bernoulli_logits(logits<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="dv">1</span>, <span class="dv">2</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-2" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-2">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#bernoulli-logits</p>
</section>
<section id="references-2" class="level4">
<h4 class="anchored" data-anchor-id="references-2">References</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Bernoulli_distribution “Bernoulli distribution”</li>
<li>https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Bernoulli “tfp. distributions.Bernoulli - Probability”</li>
<li>https://mc-stan.org/docs/2_19/functions-reference/bernoulli-logit-distribution.html “11.2 Bernoulli Distribution, Logit Parameterization - Stan”</li>
</ul>
<hr>
</section>
</section>
<section id="bernoulli-probs" class="level3">
<h3 class="anchored" data-anchor-id="bernoulli-probs">Bernoulli Probs</h3>
<p>Samples from a Bernoulli distribution parameterized by probabilities.</p>
<p>The Bernoulli distribution models the probability of success in a single trial, where the outcome is binary (success or failure). It is characterized by a single parameter, <code>probs</code>, representing the probability of success.</p>
<p><span class="math display">
Pr(X=1) = p = 1- Pr(X=0) = 1 - q, \\
f(x)=p \text{ if } k = 1 \text{ else } q \text{ if } k = 0
</span></p>
<p>where: p is the probability of success (0 &lt;= p &lt;= 1)</p>
<section id="args-4" class="level4">
<h4 class="anchored" data-anchor-id="args-4">Args:</h4>
<div class="sourceCode" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>bi.dist.bernoulli_probs(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): The probability of success for each Bernoulli trial. Must be between 0 and 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-4" class="level4">
<h4 class="anchored" data-anchor-id="returns-4">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BernoulliProbs distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI BernoulliProbs object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-4" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-4">Example Usage:</h4>
<div class="sourceCode" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>m.dist.bernoulli_probs(probs<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.7</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-3" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-3">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#bernoulliprobs</p>
<hr>
</section>
</section>
<section id="beta" class="level3">
<h3 class="anchored" data-anchor-id="beta">Beta</h3>
<p>Samples from a Beta distribution, defined on the interval [0, 1]. The Beta distribution is a versatile distribution often used to model probabilities or proportions. It is parameterized by two positive shape parameters, usually denoted <span class="math inline">\alpha</span> and <span class="math inline">\beta&gt;0</span>, control the shape of the density (how much mass is pushed toward 0, 1, or intermediate).</p>
<p><span class="math display">
X \sim Beta(\alpha,\beta), \\f(x)=\frac{ x^{\alpha-1}(1-x)^{\beta-1}}{B(\alpha,\beta)}, \\
B(\alpha,\beta)=\frac{\Gamma(\alpha)\Gamma(\beta)}{\Gamma(\alpha+\beta)}, \\F(x)=I_{x}(\alpha+\beta)
</span></p>
<p>where <span class="math inline">B(\alpha, \beta)</span> is the Beta function:</p>
<p><span class="math display">
B(\alpha, \beta) = \int_0^1 x^{\alpha - 1} (1 - x)^{\beta - 1} , dx = \frac{\Gamma(\alpha),\Gamma (\beta)}{\Gamma(\alpha + \beta)}.
</span></p>
<p>where :num$` and <span class="math inline">\beta</span> are the concentration parameters, and <span class="math inline">B(x, y)</span> is the Beta function.</p>
<section id="args-5" class="level4">
<h4 class="anchored" data-anchor-id="args-5">Args:</h4>
<div class="sourceCode" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>bi.dist.beta(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>concentration1,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>concentration0,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration1</em> (jnp.ndarray): The first concentration parameter (shape parameter). Must be positive.</p></li>
<li><p><em>concentration0</em> (jnp.ndarray): The second concentration parameter (shape parameter). Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-5" class="level4">
<h4 class="anchored" data-anchor-id="returns-5">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Beta distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI Beta object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-5" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-5">#### Example Usage:</h4>
<div class="sourceCode" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>m.dist.beta(concentration1<span class="op">=</span><span class="fl">1.0</span>, concentration0<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-4" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-4">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#beta</p>
</section>
<section id="references-3" class="level4">
<h4 class="anchored" data-anchor-id="references-3">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Beta_distribution “Beta distribution - Wikipedia”</li>
<li>https://www.statlect.com/probability-distributions/beta-distribution “Beta distribution | Properties, proofs, exercises - StatLect”</li>
<li>https://reference.wolfram.com/language/ref/BetaDistribution.html “BetaDistribution - Wolfram Language Documentation”</li>
</ul>
<hr>
</section>
</section>
<section id="betabinomial" class="level3">
<h3 class="anchored" data-anchor-id="betabinomial">BetaBinomial</h3>
<p>Samples from a BetaBinomial distribution, a compound distribution where the probability of success in a binomial experiment is drawn from a Beta distribution. This models situations where the underlying probability of success is not fixed but varies according to a prior belief represented by the Beta distribution. It is often used to model over-dispersion relative to the binomial distribution.</p>
<p>For <span class="math inline">X \sim {BetaBinomial}(N,\alpha,\beta)</span>, for <span class="math inline">k = 0,1,\dots,N</span>:</p>
<p><span class="math display">
\Pr(X = k) = \binom{N}{k} ; \frac{B(k + \alpha,; N - k + \beta)}{B(\alpha,\beta)},
</span></p>
<p>where (<span class="math inline">B(u,v)</span>) is the Beta function <span class="math inline">(B(u,v) = \frac{\Gamma(u),\Gamma(v)}{\Gamma(u+v)})</span>.</p>
<section id="args-6" class="level4">
<h4 class="anchored" data-anchor-id="args-6">Args:</h4>
<div class="sourceCode" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>bi.dist.beta_binomial(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>concentration1,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>concentration0,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration1</em> (jnp.ndarray): The first concentration parameter (alpha) of the Beta distribution.</p></li>
<li><p><em>concentration0</em> (jnp.ndarray): The second concentration parameter (beta) of the Beta distribution.</p></li>
<li><p><em>total_count</em> (jnp.ndarray): The number of Bernoulli trials in the Binomial part of the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-6" class="level4">
<h4 class="anchored" data-anchor-id="returns-6">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BetaBinomial distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BetaBinomial distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI BetaBinomial distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-6" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-6">Example Usage:</h4>
<div class="sourceCode" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>m.dist.beta_binomial(concentration1<span class="op">=</span><span class="fl">1.0</span>, concentration0<span class="op">=</span><span class="fl">1.0</span>, total_count<span class="op">=</span><span class="dv">10</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-5" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-5">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#betabinomial</p>
</section>
<section id="references-4" class="level4">
<h4 class="anchored" data-anchor-id="references-4">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Beta-binomial_distribution “Beta-binomial distribution”</li>
<li>https://www.statisticshowto.com/beta-binomial-distribution/ “Beta-Binomial Distribution: Definition - Statistics How To”</li>
<li>https://mc-stan.org/docs/2_19/functions-reference/beta-binomial-distribution.html “12.3 Beta-Binomial Distribution | Stan Functions Reference”</li>
</ul>
<hr>
</section>
</section>
<section id="beta-proportion" class="level3">
<h3 class="anchored" data-anchor-id="beta-proportion">Beta Proportion</h3>
<p>The Beta Proportion distribution is a reparameterization of the conventional Beta distribution in terms of a the variate mean and a precision parameter. It’s useful for modeling rates and proportions. It’s essentially the same family as the standard Beta((,)), but the mapping is:</p>
<p><span class="math display">
\alpha = \mu , \kappa,\quad \beta = (1 - \mu), \kappa.
</span></p>
<p>Thus the pdf is:</p>
<p><span class="math display">
f(\theta \mid \mu, \kappa) = \frac{1}{B(\mu\kappa,,(1-\mu)\kappa)} ; \theta^{\mu\kappa - 1} ; (1 - \theta)^{(1 - \mu)\kappa         - 1}, \\
\theta \in (0,1).
</span></p>
<p>This parameterization is useful in hierarchical models, since one can put hyperpriors on () (the “center”) and () (the “spread”).</p>
<section id="args-7" class="level4">
<h4 class="anchored" data-anchor-id="args-7">Args:</h4>
<div class="sourceCode" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>bi.dist.beta_proportion(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>mean,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>mean</em> (jnp.ndarray): The mean of the BetaProportion distribution,must be between 0 and 1.</p></li>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter of the BetaProportion distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-7" class="level4">
<h4 class="anchored" data-anchor-id="returns-7">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BetaProportion distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BetaProportion distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI BetaProportion distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-7" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-7">Example Usage:</h4>
<div class="sourceCode" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> m.dist.beta_proportion(mean<span class="op">=</span><span class="fl">0.5</span>, concentration<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>, shape<span class="op">=</span>(<span class="dv">1000</span>,))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-6" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-6">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#beta_proportion</p>
</section>
<section id="references-5" class="level4">
<h4 class="anchored" data-anchor-id="references-5">References:</h4>
<p>https://mc-stan.org/docs/2_19/functions-reference/beta-proportion-distribution.html “19.2 Beta Proportion Distribution | Stan Functions Reference”</p>
<hr>
</section>
</section>
<section id="binomial" class="level3">
<h3 class="anchored" data-anchor-id="binomial">Binomial</h3>
<p>The Binomial distribution models the number of successes in a sequence of independent Bernoulli trials. It represents the probability of obtaining exactly <em>k</em> successes in <em>n</em> trials, where each trial has a probability <em>p</em> of success. It assumes:</p>
<ol type="1">
<li>Each trial has exactly two possible outcomes (success or failure).</li>
<li>The probability of success $p is constant across trials.</li>
<li>Trials are independent.</li>
</ol>
<p>We denote: <span class="math display">
X \sim {Binomial}(n, p).
</span></p>
<section id="probability-mass-function-pmf-1" class="level4">
<h4 class="anchored" data-anchor-id="probability-mass-function-pmf-1">Probability Mass Function (PMF)</h4>
<p>For <span class="math inline">k = 0,1,2,\dots,n</span>:</p>
<p><span class="math display">
\Pr(X = k) = \binom{n}{k} ; p^k ; (1 - p)^{,n - k},
</span></p>
<p>where <span class="math inline">\binom{n}{k} = \frac{n!}{k!(n-k)!}</span>.</p>
<p>This formula counts the number of ways to choose which <span class="math inline">k</span> out of <span class="math inline">n</span> trials are the successes (the binomial coefficient) and then multiplies by the probability of each such configuration <span class="math inline">p^k (1-p)^{n-k}</span>.</p>
</section>
<section id="args-8" class="level4">
<h4 class="anchored" data-anchor-id="args-8">Args:</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>bi.dist.binomial(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>total_count</em> (int): The number of trials <em>n</em>.</p></li>
<li><p><em>probs</em> (jnp.ndarray, optional): The probability of success <em>p</em> for each trial. Must be between 0 and 1.</p></li>
<li><p><em>logits</em> (jnp.ndarray, optional): The log-odds of success for each trial. <code>probs = jax.nn.sigmoid(logits)</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-8" class="level4">
<h4 class="anchored" data-anchor-id="returns-8">Returns:</h4>
<p>Binomial distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the Binomial distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-8" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-8">Example Usage:</h4>
<div class="sourceCode" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>m.dist.binomial(total_count<span class="op">=</span><span class="dv">10</span>, probs<span class="op">=</span><span class="fl">0.5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlbinomial" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlbinomial">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#binomial</h4>
</section>
<section id="references-6" class="level4">
<h4 class="anchored" data-anchor-id="references-6">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Binomial_distribution “Binomial distribution”</li>
<li>https://www.itl.nist.gov/div898/handbook/eda/section3/eda366i.htm “1.3.6.6.18. Binomial Distribution - Information Technology Laboratory”</li>
</ul>
<hr>
</section>
</section>
<section id="binomial-logits" class="level3">
<h3 class="anchored" data-anchor-id="binomial-logits">Binomial Logits</h3>
<p>The BinomialLogits distribution represents a binomial distribution parameterized by logits. It is useful when the probability of success is not directly known but is instead expressed as logits, which are the natural logarithm of the odds ratio.</p>
<p><span class="math display">
p = {logit}^{-1}(\alpha) = \frac{1}{1 + e^{-\alpha}}.
</span></p>
<section id="args-9" class="level4">
<h4 class="anchored" data-anchor-id="args-9">Args:</h4>
<div class="sourceCode" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>bi.dist.binomial_logits(</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>logits</em> (jnp.ndarray): Log-odds of each success.</p></li>
<li><p><em>total_count</em> (int): Number of trials.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-9" class="level4">
<h4 class="anchored" data-anchor-id="returns-9">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BinomialLogits distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI BinomialLogits object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-9" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-9">Example Usage:</h4>
<div class="sourceCode" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>m.dist.binomial_logits(logits<span class="op">=</span>jnp.zeros(<span class="dv">10</span>), total_count<span class="op">=</span><span class="dv">5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-7" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-7">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#binomialllogits</p>
<hr>
</section>
</section>
<section id="binomial-probs" class="level3">
<h3 class="anchored" data-anchor-id="binomial-probs">Binomial Probs</h3>
<p>Samples from a Binomial distribution with specified probabilities for each trial.</p>
<p>The Binomial distribution models the number of successes in a sequence of independent Bernoulli trials, where each trial has the same probability of success.</p>
<p>For <span class="math inline">k = 0,1,2,\dots,N</span>: <span class="math display">   
\Pr(X = k) = \binom{N}{k} ; p^k ; (1 - p)^{,N - k}
</span></p>
<p>where <span class="math inline">\binom{N}{k} = \frac{N!}{k!(N-k)!}</span>.</p>
<section id="args-10" class="level4">
<h4 class="anchored" data-anchor-id="args-10">Args:</h4>
<div class="sourceCode" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>bi.dist.binomial_probs(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): The probability of success for each trial. Must be between 0 and 1.</p></li>
<li><p><em>total_count</em> (int): The number of trials in each sequence.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-10" class="level4">
<h4 class="anchored" data-anchor-id="returns-10">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI BinomialProbs distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI BinomialLogits object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-10" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-10">Example Usage:</h4>
<div class="sourceCode" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>m.dist.binomial_probs(probs<span class="op">=</span><span class="fl">0.5</span>, total_count<span class="op">=</span><span class="dv">10</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-8" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-8">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#binomialprobs</p>
<hr>
</section>
</section>
<section id="conditional-autoregressive-car" class="level3">
<h3 class="anchored" data-anchor-id="conditional-autoregressive-car">Conditional Autoregressive (CAR)</h3>
<p>The CAR distribution models a vector of variables where each variable is a linear combination of its neighbors in a graph. The CAR model captures spatial dependence in areal data by modeling each observation as conditionally dependent on its neighbors.It specifies a joint distribution of a vector of random variables <span class="math inline">{y} = (y_1, y_2, \dots, y_N)</span> based on their conditional distributions, where each <span class="math inline">y_i</span> is conditionally independent of all other variables given its neighbors.</p>
<ul>
<li><p><strong>Application</strong>: Widely used in disease mapping, environmental modeling, and spatial econometrics to account for spatial autocorrelation.</p></li>
<li><p><strong>Conditional Distribution</strong>: Each <span class="math inline">y_i</span> given its neighbors <span class="math inline">y_{-i}</span> follows a normal distribution:</p></li>
</ul>
<p><span class="math display">
y_i | y*{-i} \sim N\left( \frac{\sum*{j \in N(i)} w_{ij} y_j}{\sum_{j \in N(i)} w_{ij}}, \sigma^2 \right)
</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">{N}(i)</span> denotes the set of neighbors of unit <span class="math inline">i</span>,</li>
<li><span class="math inline">w_{ij}</span> are the weights representing the strength of the spatial relationship between units <span class="math inline">i</span> and <span class="math inline">j</span>,</li>
<li><span class="math inline">\sigma^2</span> is the variance term. where <span class="math inline">\mu_i</span> is a function of the values of the neighbors of site <span class="math inline">i</span> and <span class="math inline">\Sigma_i</span> is the variance of site <span class="math inline">i</span>.</li>
</ul>
<section id="args-11" class="level4">
<h4 class="anchored" data-anchor-id="args-11">Args:</h4>
<div class="sourceCode" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>bi.dist.car(</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>loc,</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>correlation,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>conditional_precision,</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>adj_matrix,</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>is_sparse<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (Union[float, Array]): Mean of the distribution.</p></li>
<li><p><em>correlation</em> (Union[float, Array]): Correlation between variables.</p></li>
<li><p><em>conditional_precision</em> (Union[float, Array]): Precision of the distribution.</p></li>
<li><p><em>adj_matrix</em> (Union[Array, scipy.sparse.spmatrix]): Adjacency matrix defining the graph. is_sparse (bool): Whether the adjacency matrix is sparse. Defaults to False.</p></li>
<li><p>*validate_args/ (bool): Whether to validate arguments. Defaults to None.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-11" class="level4">
<h4 class="anchored" data-anchor-id="returns-11">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI CAR distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI CAR object (for advanced use cases).</p></li>
</ul>
</section>
<section id="references-7" class="level4">
<h4 class="anchored" data-anchor-id="references-7">References:</h4>
<ul>
<li>https://www.pymc.io/projects/examples/en/latest/spatial/conditional_autoregressive_priors.html “Conditional Autoregressive (CAR) Models for Spatial Data”</li>
<li>https://mc-stan.org/learn-stan/case-studies/icar_stan.html “Intrinsic Auto-Regressive Models for Areal Data - Stan”</li>
<li>https://onlinelibrary.wiley.com/doi/abs/10.1002/9781118445112.stat08048 “Conditional Autoregressive (CAR) Model - Schmidt”</li>
</ul>
<hr>
</section>
</section>
<section id="categorical-distribution." class="level3">
<h3 class="anchored" data-anchor-id="categorical-distribution.">Categorical distribution.</h3>
<p>The Categorical distribution, also known as the multinomial distribution, describes the probability of <span class="math inline">K</span> different outcomes with probabilities <span class="math inline">{p_1, p_2, \dots, p_K}</span>.There’s no inherent ordering required among the categories.</p>
<ul>
<li>The probabilities satisfy (p_i ) for all (i) and $ _{i=1}^K p_i = 1$ .</li>
<li>It’s essentially the distribution of a single draw from a discrete set of categories.</li>
<li>It generalises the Bernoulli distribution (which is the special case <span class="math inline">K=2</span>).</li>
</ul>
<p>It is commonly used to model discrete choices or classifications.</p>
<p>If <span class="math inline">X</span> is a categorical random variable taking values in <span class="math inline">{1, 2, \dots, K}</span> with probabilities <span class="math inline">p_1, \dots, p_K</span>, then:</p>
<p><span class="math display">
\Pr(X = i) = p_i, \quad i = 1,\dots, K.
</span></p>
<p>where <span class="math inline">p_k</span> is the probability of outcome <span class="math inline">k</span>, and the sum is over all possible outcomes.</p>
<section id="args-12" class="level4">
<h4 class="anchored" data-anchor-id="args-12">Args:</h4>
<div class="sourceCode" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>bi.dist.categorical(</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): A 1D array of probabilities for each category. Must sum to 1.</p></li>
<li><p><em>logits</em> (jnp.ndarray): A 1D array of unnormalized log probabilities for each category.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-12" class="level4">
<h4 class="anchored" data-anchor-id="returns-12">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Categorical distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Categorical distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-11" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-11">Example Usage:</h4>
<div class="sourceCode" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>m.dist.categorical(probs<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlcategorical" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlcategorical">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#categorical</h4>
</section>
<section id="references-8" class="level4">
<h4 class="anchored" data-anchor-id="references-8">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Categorical_distribution”Categorical distribution”</li>
<li>https://distribution-explorer.github.io/discrete/categorical.html”Categorical distribution — Probability Distribution Explorer documentation”</li>
<li>https://static.hlt.bme.hu/semantics/external/pages/n-gram/en.wikipedia.org/wiki/Categorical_distribution.html”Categorical distribution - Wikipedia”</li>
</ul>
<hr>
</section>
</section>
<section id="categorical-logits" class="level3">
<h3 class="anchored" data-anchor-id="categorical-logits">Categorical Logits</h3>
<p>Samples from a Categorical distribution with logits. This distribution represents a discrete probability distribution over a finite set of outcomes, where the probabilities are determined by the logits. The probability of each outcome is given by the softmax function applied to the logits.</p>
<p>Given logits $ = (_1, , _K) $, the probability of each category $ i $ is: <span class="math display">
p_i = \frac{e^{\gamma_i}}{\sum_{j=1}^{K} e^{\gamma_j}}
</span></p>
<p>where $ p_i is the probability of category $ i , and the sum in the denominator ensures that the probabilities sum to 1.</p>
<section id="args-13" class="level4">
<h4 class="anchored" data-anchor-id="args-13">Args:</h4>
<div class="sourceCode" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>bi.dist.categorical_logits(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>logits</em> (jnp.ndarray): Log-odds of each category.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-13" class="level4">
<h4 class="anchored" data-anchor-id="returns-13">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI CategoricalLogits distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI CategoricalLogits object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-12" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-12">Example Usage:</h4>
<div class="sourceCode" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>m.dist.categorical_logits(logits<span class="op">=</span>jnp.zeros(<span class="dv">5</span>), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-9" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-9">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#categoricallogits</p>
<hr>
</section>
</section>
<section id="categorical-probs-distribution." class="level3">
<h3 class="anchored" data-anchor-id="categorical-probs-distribution.">Categorical Probs distribution.</h3>
<p>Samples from a Categorical distribution.</p>
<p>The Categorical distribution is a discrete probability distribution that represents the probability of each outcome from a finite set of possibilities. It is often used to model the outcome of a random process with a fixed number of possible outcomes, such as the roll of a die or the selection of an item from a list.</p>
<p><span class="math display">   
P(x) = \frac{probs_i}{\sum_{k=1}^{K} probs_k}
</span></p>
<section id="args-14" class="level4">
<h4 class="anchored" data-anchor-id="args-14">Args:</h4>
<div class="sourceCode" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>bi.dist.categorical_probs(</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): Probabilities for each category. Must sum to 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-14" class="level4">
<h4 class="anchored" data-anchor-id="returns-14">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI CategoricalProbs distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI CategoricalProbs object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-13" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-13">Example Usage:</h4>
<div class="sourceCode" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>m.dist.categorical_probs(probs<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-10" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-10">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#categoricalprobs</p>
<hr>
</section>
</section>
<section id="cauchy" class="level3">
<h3 class="anchored" data-anchor-id="cauchy">Cauchy</h3>
<p>Samples from a Cauchy</p>
<p>The Cauchy distribution, also known as the Lorentz distribution, is a continuous probability distribution that arises frequently in various fields, including physics and statistics. It is characterized by its heavy tails, which extend indefinitely. This means it has a higher probability of extreme values compared to the normal distribution.</p>
<ul>
<li><strong>Parameters</strong>:</li>
<li><strong>Location parameter</strong> (<span class="math inline">\mu</span>): Determines the peak of the distribution.</li>
<li><strong>Scale parameter</strong> (<span class="math inline">\gamma</span>): Describes the half-width at half-maximum (HWHM) of the peak. <span class="math display">
f(x; \mu, \gamma) = \frac{1}{\pi \gamma \left[1 + \left(\frac{x - \mu}{\gamma}\right)^2\right]}
</span></li>
</ul>
<p>where:</p>
<ul>
<li><span class="math inline">x</span> is the random variable,</li>
<li><span class="math inline">\mu</span> is the location parameter,</li>
<li><span class="math inline">\gamma</span> is the scale parameter.</li>
</ul>
<section id="args-15" class="level4">
<h4 class="anchored" data-anchor-id="args-15">Args:</h4>
<div class="sourceCode" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>bi.dist.cauchy(</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray or float, optional): Location parameter. Defaults to 0.0.</p></li>
<li><p><em>sample</em> (jnp.ndarray or float, optional): Scale parameter. Must be positive. Defaults to 1.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as event dimensions (used in model building). Defaults to None.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. Defaults to None.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-15" class="level4">
<h4 class="anchored" data-anchor-id="returns-15">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Cauchy distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Cauchy distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-14" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-14">Example Usage:</h4>
<div class="sourceCode" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>m.dist.cauchy(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-11" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-11">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#cauchy</p>
<hr>
</section>
</section>
<section id="chi-squared" class="level3">
<h3 class="anchored" data-anchor-id="chi-squared">Chi-squared</h3>
<p>The Chi-squared distribution is a continuous probability distribution that arises frequently in hypothesis testing, particularly in ANOVA and chi-squared tests. It is defined by a single positive parameter, degrees of freedom (df), the number of independent standard normal variables squared and summed, which determines the shape of the distribution.</p>
<p><span class="math display">   
f(x; k) = \begin{cases}
\frac{x^{\frac{k}{2} - 1} e^{-x/2}}{2^{\frac{k}{2}} \Gamma\left(\frac{k}{2}\right)},  x &gt; 0 \
0,  \text{otherwise}
\end{cases}
</span></p>
<p>where: * ( x ) is the random variable, * ( k ) is the degrees of freedom, * ( () ) is the gamma function.</p>
<section id="args-16" class="level4">
<h4 class="anchored" data-anchor-id="args-16">Args:</h4>
<div class="sourceCode" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>bi.dist.chi2(</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>df,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>df</em> (jnp.ndarray): Degrees of freedom. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-16" class="level4">
<h4 class="anchored" data-anchor-id="returns-16">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Chi2 distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI Chi2 object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-15" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-15">Example Usage:</h4>
<div class="sourceCode" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu)</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="er">m.dist.chi2</span>(df<span class="op">=</span><span class="fl">3.0</span>, sample <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-12" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-12">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#chi2</p>
<hr>
</section>
</section>
<section id="circulant-normal-multivariate-normal" class="level3">
<h3 class="anchored" data-anchor-id="circulant-normal-multivariate-normal">Circulant Normal Multivariate normal</h3>
<p>Circulant Normal Multivariate normal distribution with covariance matrix ${C}<code>that is positive-definite and circulant [1], i.e., has periodic boundary conditions. The density of a sample ${x}\in{R}^n</code> is the standard multivariate normal density</p>
<p><span class="math display">
p\left(x\mid \mu,{C}\right) =
\frac{\left({det}\,{C}\right)^{-1/2}}{\left(2\pi\right)^{n / 2}}
\exp\left(-\frac{1}{2}\left({x}- \mu\right)^\intercal
{C}^{-1}\left({x}-\mu\right)\right),
</span></p>
<p>where <span class="math inline">{det}</span> denotes the determinant and <span class="math inline">^\intercal</span> the transpose. Circulant matrices can be diagnolized efficiently using the discrete Fourier transform, allowing the log likelihood to be evaluated in <span class="math inline">n \log n</span> time for <span class="math inline">n</span> observations.</p>
<ul>
<li><p><em>loc:</em> Mean of the distribution <span class="math inline">\mu</span>.</p></li>
<li><p><em>covariance_row:</em> First row of the circulant covariance matrix <span class="math inline">C</span>. Because of periodic boundary conditions, the covariance matrix is fully determined by its first row (see <code>jax.scipy.linalg.toeplitz</code> for further details).</p></li>
<li><p><em>covariance_rfft:</em> Real part of the real fast Fourier transform of :code:<code>covariance_row</code>, the first row of the circulant covariance matrix <span class="math inline">C</span>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
<p>References:</p>
<ol type="1">
<li><p>Wikipedia. (n.d.). Circulant matrix. Retrieved March 6, 2025, from https://en.wikipedia.org/wiki/Circulant_matrix</p></li>
<li><p>Wood, A. T. A., &amp; Chan, G. (1994). Simulation of Stationary Gaussian Processes in <span class="math inline">\left[0, 1\right]^d</span>. <em>Journal of Computational and Graphical Statistics</em>, 3(4), 409–432. https://doi.org/10.1080/10618600.1994.10474655</p></li>
</ol>
<section id="args-17" class="level4">
<h4 class="anchored" data-anchor-id="args-17">Args:</h4>
<div class="sourceCode" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>bi.dist.circulant_normal(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>loc: jax.Array,</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>covariance_row: jax.Array <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>covariance_rfft: jax.Array <span class="op">=</span> <span class="va">None</span>,</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> : jnp.ndarray Mean of the distribution <span class="math inline">\mu</span>.</p></li>
<li><p><em>covariance_row</em> : jnp.ndarray, optional. First row of the circulant covariance matrix <span class="math inline">C</span>. Defaults to None.</p></li>
<li><p><em>covariance_rfft</em> : jnp.ndarray, optional Real part of the real fast Fourier transform of <code>covariance_row</code>. Defaults to None.</p></li>
</ul>
</section>
<section id="returns-17" class="level4">
<h4 class="anchored" data-anchor-id="returns-17">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Circulant Normal Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI Circulant Normal Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlcirculantnormal" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlcirculantnormal">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#circulantnormal</h4>
</section>
<section id="references-9" class="level4">
<h4 class="anchored" data-anchor-id="references-9">References:</h4>
<p>Wikipedia. (n.d.). Circulant matrix. Retrieved March 6, 2025, from https://en.wikipedia.org/wiki/Circulant_matrix</p>
<p>Wood, A. T. A., &amp; Chan, G. (1994). Simulation of Stationary Gaussian Processes in <span class="math inline">[0,1]^d</span>. Journal of Computational and Graphical Statistics, 3(4), 409–432. https://doi.org/10.1080/10618600.1994.10474655</p>
<hr>
</section>
</section>
<section id="delta" class="level3">
<h3 class="anchored" data-anchor-id="delta">Delta</h3>
<p>The Delta distribution, also known as a point mass distribution, assigns probability 1 to a single point and 0 elsewhere. It’s useful for representing deterministic variables or as a building block for more complex distributions.</p>
<p><span class="math display">   
F(x) =
\begin{cases}
0,  x &lt; a, \
1,  x \ge a.
\end{cases}
</span></p>
<section id="args-18" class="level4">
<h4 class="anchored" data-anchor-id="args-18">Args:</h4>
<div class="sourceCode" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>bi.dist.delta(</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>v<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>log_density<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>event_dim<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>v</em> (jnp.ndarray): The location of the point mass.</p></li>
<li><p><em>log_density</em> (float, optional): The log probability density of the point mass. This is primarily for creating distributions that are non-normalized or for specific advanced use cases. For a standard delta distribution, this should be 0. Defaults to 0.0.</p></li>
<li><p><em>event_dim</em> (int, optional): The number of rightmost dimensions of <code>v</code> to interpret as event dimensions. Defaults to 0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-18" class="level4">
<h4 class="anchored" data-anchor-id="returns-18">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Circulant Delta Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the BernoulliLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI Circulant Delta Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-16" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-16">Example Usage:</h4>
<div class="sourceCode" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>m.dist.delta(v<span class="op">=</span><span class="fl">0.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-13" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-13">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#delta</p>
<hr>
</section>
</section>
<section id="dirichlet" class="level3">
<h3 class="anchored" data-anchor-id="dirichlet">Dirichlet</h3>
<p>Samples from a Dirichlet distribution.</p>
<p>The Dirichlet distribution is a multivariate generalization of the Beta distribution. It is a probability distribution over a simplex, which is a set of vectors where each element is non-negative and sums to one. It is often used as a prior distribution for categorical distributions.</p>
<p>For <span class="math inline">X = (X_1, \dots, X_K \sim {Dir}(\alpha)</span>, the density is</p>
<p><span class="math display">
f( x; \alpha) = \frac{1}{B(\alpha)} \prod_{i=1}^K x_i^{\alpha_i - 1}
</span></p>
<p>for <span class="math inline">x</span> in the simplex, and 0 otherwise. Here <span class="math inline">B(\alpha)</span> is the multivariate Beta function (the normalization constant):</p>
<p><span class="math display">
B(\alpha) = \frac{\prod_{i=1}^K \Gamma(\alpha_i)}{\Gamma\left(\sum_{i=1}^K \alpha_i\right)}.
</span></p>
<section id="args-19" class="level4">
<h4 class="anchored" data-anchor-id="args-19">Args:</h4>
<div class="sourceCode" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>bi.dist.dirichlet(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter(s) of the Dirichlet distribution. Must be a positive array.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-19" class="level4">
<h4 class="anchored" data-anchor-id="returns-19">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Dirichlet Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Dirichlet distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI Dirichlet Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-17" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-17">Example Usage:</h4>
<div class="sourceCode" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>m.dist.dirichlet(concentration<span class="op">=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">1.0</span>, <span class="fl">1.0</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmldirichlet" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmldirichlet">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#dirichlet</h4>
</section>
<section id="references-10" class="level4">
<h4 class="anchored" data-anchor-id="references-10">References</h4>
<p>https://distribution-explorer.github.io/multivariate_continuous/dirichlet.html https://en.wikipedia.org/wiki/Dirichlet_distribution</p>
<hr>
</section>
</section>
<section id="dirichlet-multinomial" class="level3">
<h3 class="anchored" data-anchor-id="dirichlet-multinomial">Dirichlet-Multinomial</h3>
<p>Creates a Dirichlet-Multinomial compound distribution, which is a Multinomial distribution with a Dirichlet prior on its probabilities. It is often used in Bayesian statistics to model count data where the proportions of categories are uncertain.</p>
<p>The <strong>Dirichlet-Multinomial</strong> distribution describes a model in which you first draw a probability vector <span class="math display">
{p} = (p_1, \dots, p_K) \sim {Dirichlet}(\alpha)
</span> and then you draw counts <span class="math display">
{x} = (x_1, \dots, x_K) \sim {Multinomial}(n, {p}).
</span></p>
<p>Integrating out <span class="math inline">{p}</span> gives you the marginal distribution of ${x}). * It generalizes the <strong>Beta-Binomial</strong> model to multiple categories (i.e.&nbsp;(K &gt; 2)). * A key feature is that it introduces <strong>overdispersion</strong> relative to a standard multinomial: counts are more “clumped” because the underlying probabilities vary according to the Dirichlet prior.</p>
<section id="args-20" class="level4">
<h4 class="anchored" data-anchor-id="args-20">Args:</h4>
<div class="sourceCode" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>bi.dist.dirichlet_multinomial(</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb39-13"><a href="#cb39-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb39-14"><a href="#cb39-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter (alpha) for the Dirichlet prior. Values must be positive. The last dimension is interpreted as the number of categories.</p></li>
<li><p><em>total_count</em> (int, jnp.ndarray, optional): The total number of trials (n). This must be a non-negative integer. Defaults to 1.</p></li>
<li><p><em>validate_args</em> (bool, optional): Whether to enable validation of distribution parameters. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. If provided, events with a <code>True</code> mask will be conditioned on <code>obs</code>, while the remaining events will be treated as latent variables. Defaults to <code>None</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>. Defaults to 0.</p></li>
<li><p><em>shape</em> (tuple, optional): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to <code>False</code>.</p></li>
</ul>
</section>
<section id="returns-20" class="level4">
<h4 class="anchored" data-anchor-id="returns-20">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI dirichlet_multinomial Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the dirichlet_multinomial distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI dirichlet_multinomial Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-18" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-18">Example Usage:</h4>
<div class="sourceCode" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> jax.numpy <span class="im">as</span> jnp</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Direct sampling</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample a single vector of counts for 10 trials from 3 categories</span></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>counts <span class="op">=</span> m.dist.dirichlet_multinomial(concentration<span class="op">=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">1.0</span>, <span class="fl">1.0</span>]),total_count<span class="op">=</span><span class="dv">10</span>,sample<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage within a model</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> my_model(obs_data<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a prior on the concentration parameter</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>alpha <span class="op">=</span> m.dist.half_cauchy(scale<span class="op">=</span>jnp.ones(<span class="dv">5</span>), name<span class="op">=</span><span class="st">'alpha'</span>, shape<span class="op">=</span>(<span class="dv">5</span>,))</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Model observed counts</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> m.plate(<span class="st">'data'</span>, <span class="bu">len</span>(obs_data)):</span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> m.dist.dirichlet_multinomial(</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a>concentration<span class="op">=</span>alpha,</span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">100</span>,</span>
<span id="cb40-19"><a href="#cb40-19" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'y'</span>,</span>
<span id="cb40-20"><a href="#cb40-20" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span>obs_data</span>
<span id="cb40-21"><a href="#cb40-21" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-14" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-14">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#dirichletmultinomial</p>
</section>
<section id="references-11" class="level4">
<h4 class="anchored" data-anchor-id="references-11">References:</h4>
<p>https://en.wikipedia.org/wiki/Dirichlet-multinomial_distribution https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.dirichlet_multinomial.html</p>
<hr>
</section>
</section>
<section id="discrete-uniform" class="level3">
<h3 class="anchored" data-anchor-id="discrete-uniform">Discrete Uniform</h3>
<p>Samples from a Discrete Uniform distribution.</p>
<p>The Discrete Uniform distribution defines a uniform distribution over a range of integers. It is characterized by a lower bound (<code>low</code>) and an upper bound (<code>high</code>), inclusive.</p>
<p><span class="math display">   
P(X = k) = \frac{1}{high - low + 1}, \quad k \in \{low, low+1, ..., high\}
</span></p>
<p>Otherwise (if <span class="math inline">k</span> is outside the range), $ P(X = k) = 0 $.</p>
<section id="args-21" class="level4">
<h4 class="anchored" data-anchor-id="args-21">Args:</h4>
<div class="sourceCode" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>bi.dist.discrete_uniform(</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>low</em> (jnp.ndarray): The lower bound of the uniform range, inclusive.</p></li>
<li><p><em>high</em> (jnp.ndarray): The upper bound of the uniform range, inclusive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-21" class="level4">
<h4 class="anchored" data-anchor-id="returns-21">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI DiscreteUniform Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the DiscreteUniform distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI DiscreteUniform Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-19" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-19">Example Usage:</h4>
<div class="sourceCode" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>m.dist.discrete_uniform(low<span class="op">=</span><span class="dv">0</span>, high<span class="op">=</span><span class="dv">5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-15" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-15">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#discreteuniform</p>
</section>
<section id="references-12" class="level4">
<h4 class="anchored" data-anchor-id="references-12">References:</h4>
<hr>
</section>
</section>
<section id="doubly-truncated-power-law" class="level3">
<h3 class="anchored" data-anchor-id="doubly-truncated-power-law">Doubly Truncated Power Law</h3>
<p>This distribution represents a continuous power law with a finite support bounded between <code>low</code> and <code>high</code>, and with an exponent <code>alpha</code>. It is normalized over the interval <code>[low, high]</code> to ensure the area under the density function is 1.</p>
<p>Key parameters:</p>
<ul>
<li><span class="math inline">\alpha &gt; 0</span>: shape (exponent) of the power law</li>
<li><span class="math inline">x_{min}</span>: lower cutoff (minimum possible value)</li>
<li><span class="math inline">x_{\max}</span>: upper cutoff (maximum possible value)</li>
</ul>
<p>The probability density function (PDF) is defined as:</p>
<p><span class="math display">
f(x) = \begin{cases}
\displaystyle \frac{(1-\alpha), x^{-\alpha}}{x_{\max}^{1-\alpha} - x_{min}^{1-\alpha}},  \alpha \neq 1, ; x_{min} \le x         \le x_{\max},\
\displaystyle \frac{1}{x, \ln\left(x_{\max}/x_{min}\right)},  \alpha = 1, ; x_{min} \le x \le x_{\max},\
0,  \text{otherwise}.
\end{cases}
</span></p>
<p>where the normalization constant <span class="math inline">Z(\alpha,a,b)</span> is given by</p>
<p><span class="math display">
Z(\alpha,a,b) =
\begin{cases}
\log(b) - \log(a),  \text{if } \alpha = -1, \\
\dfrac{b^{1+\alpha} - a^{1+\alpha}}{1+\alpha},  \text{otherwise}.
\end{cases}
</span></p>
<p>This distribution is useful for modeling data that follows a power-law behavior but is naturally bounded due to measurement or theoretical constraints (e.g., finite-size systems).</p>
<section id="args-22" class="level4">
<h4 class="anchored" data-anchor-id="args-22">Args:</h4>
<div class="sourceCode" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>bi.dist.doubly_truncated_power_law(</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>alpha,</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>low,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>high,</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>alpha</em> (float or array-like): Power-law exponent.</p></li>
<li><p><em>low</em> (float or array-like): Lower bound of the distribution (must be <span class="math inline">≥</span> 0).</p></li>
<li><p><em>high</em> (float or array-like): Upper bound of the distribution (must be <span class="math inline">&gt;</span> 0).</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p>*validate_args (bool, optional): Whether to validate the arguments. Defaults to True.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-22" class="level4">
<h4 class="anchored" data-anchor-id="returns-22">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI doubly_truncated_power_law Distribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the doubly_truncated_power_law distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI doubly_truncated_power_law Distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-20" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-20">Example Usage:</h4>
<div class="sourceCode" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)        </span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>m.dist.doubly_truncated_power_law(low<span class="op">=</span><span class="fl">0.1</span>, high<span class="op">=</span><span class="fl">10.0</span>, alpha<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="references-13" class="level4">
<h4 class="anchored" data-anchor-id="references-13">References:</h4>
<p>https://reference.wolframcloud.com/language/ref/TruncatedDistribution.html</p>
<hr>
</section>
</section>
<section id="eulermaruyama" class="level3">
<h3 class="anchored" data-anchor-id="eulermaruyama">Euler–Maruyama</h3>
<p>Euler–Maruyama methode is a method for the approximate numerical solution of a stochastic differential equation (SDE). It simulates the solution to an SDE by iteratively applying the Euler method to each time step, incorporating a random perturbation to account for the diffusion term.</p>
<p><span class="math display">
dX_t = f(X_t, t) dt + g(X_t, t) dW_t
</span></p>
<p>where: - $X_t<code>is the state of the system at time $t</code>. - $f(X_t, t)<code>is the drift coefficient. - $g(X_t, t)</code> is the diffusion coefficient. - $dW_t` is a Wiener process (Brownian motion).</p>
<section id="args-23" class="level4">
<h4 class="anchored" data-anchor-id="args-23">Args:</h4>
<div class="sourceCode" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>bi.dist.euler_maruyama(</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>t,</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>sde_fn,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>init_dist,</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>t</em> (jnp.ndarray): Discretized time steps.</p></li>
<li><p><em>sde_fn</em> (callable): A function that takes the current state and time as input and returns the drift and diffusion coefficients.</p></li>
<li><p><em>init_dist</em> (Distribution): The initial distribution of the system.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p>*sample_shape (tuple, optional): The shape of the samples to draw. Defaults to None.</p></li>
<li><p>*validate_args (bool, optional): Whether to validate the arguments. Defaults to True.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-23" class="level4">
<h4 class="anchored" data-anchor-id="returns-23">Returns:</h4>
<p>jnp.ndarray: Samples drawn from the Euler–Maruyama distribution.</p>
</section>
<section id="example-usage-21" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-21">Example Usage:</h4>
<div class="sourceCode" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>m.dist.euler_maruyama(t<span class="op">=</span>jnp.array([<span class="fl">0.0</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>]), sde_fn<span class="op">=</span><span class="kw">lambda</span> x, t: (x, <span class="fl">1.0</span>), init_dist<span class="op">=</span>m.dist.normal(<span class="fl">0.0</span>, <span class="fl">1.0</span>, create_obj<span class="op">=</span><span class="va">True</span>), sample <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="exponential" class="level3">
<h3 class="anchored" data-anchor-id="exponential">Exponential</h3>
<p>The Exponential distribution is a continuous probability distribution that models the time until an event occurs in a Poisson process, where events occur continuously and independently at a constant average rate. It is often used to model the duration of events, such as the time until a machine fails or the length of a phone call.</p>
<p>For $x $ <span class="math display">   
F(x; \lambda) = 1 - e^{-\lambda x}
</span></p>
<p>Zhere ( &gt; 0 ) is the rate parameter (the mean is ( 1/)).</p>
<section id="args-24" class="level4">
<h4 class="anchored" data-anchor-id="args-24">Args:</h4>
<div class="sourceCode" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>bi.dist.exponential(</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter, $`. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-24" class="level4">
<h4 class="anchored" data-anchor-id="returns-24">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Exponential distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Exponential distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-22" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-22">Example Usage:</h4>
<div class="sourceCode" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>m.dist.exponential(rate<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-16" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-16">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#exponential</p>
</section>
<section id="references-14" class="level4">
<h4 class="anchored" data-anchor-id="references-14">References</h4>
<p>https://reference.wolfram.com/language/ref/ExponentialDistribution.html https://en.wikipedia.org/wiki/Exponential_distribution</p>
<hr>
</section>
</section>
<section id="folded" class="level3">
<h3 class="anchored" data-anchor-id="folded">Folded</h3>
<p>Samples from a Folded distribution, which is the absolute value of a base univariate distribution. This distribution reflects the base distribution across the origin, effectively taking the absolute value of each sample.</p>
<p><span class="math display">
p(x) = \sum_{k=-\infty}^{\infty} p(x - 2k)
</span></p>
<section id="args-25" class="level4">
<h4 class="anchored" data-anchor-id="args-25">Args:</h4>
<div class="sourceCode" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>bi.dist.folded_distribution(</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (float, optional): Location parameter of the base distribution. Defaults to 0.0.</p></li>
<li><p><em>sample</em> (float, optional): Scale parameter of the base distribution. Defaults to 1.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-25" class="level4">
<h4 class="anchored" data-anchor-id="returns-25">Returns:</h4>
<p>BI FoldedDistribution distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the FoldedDistribution distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-23" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-23">Example Usage:</h4>
<div class="sourceCode" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>m.dist.folded_distribution(m.dist.normal(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, create_obj <span class="op">=</span> <span class="va">True</span>), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-17" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-17">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#foldeddistribution</p>
<hr>
</section>
</section>
<section id="gamma" class="level3">
<h3 class="anchored" data-anchor-id="gamma">Gamma</h3>
<p>Samples from a Gamma distribution.</p>
<p>The Gamma distribution is a continuous probability distribution frequently used in Bayesian statistics, particularly as a prior distribution for variance parameters. It is defined by two positive shape parameters: concentration (<span class="math inline">k</span>) and rate (<span class="math inline">\lambda</span>).</p>
<p>There are two common parameterizations:</p>
<ol type="1">
<li><p><strong>Shape–scale</strong>: parameters (<span class="math inline">\alpha, \theta</span>), with density <span class="math display">
f(x) = \frac{1}{\Gamma(\alpha) \theta^\alpha} x^{\alpha - 1} e^{-x/\theta}, \quad x \ge 0
</span></p></li>
<li><p><strong>Shape–rate</strong>: parameters (<span class="math inline">\alpha, \lambda</span>) where <span class="math inline">\lambda = 1/\theta</span>, with density <span class="math display">
f(x) = \frac{\lambda^\alpha}{\Gamma(\alpha)} x^{\alpha - 1} e^{-\lambda x}, \quad x \ge 0
</span></p></li>
</ol>
<section id="args-26" class="level4">
<h4 class="anchored" data-anchor-id="args-26">Args:</h4>
<div class="sourceCode" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gamma(</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb51-9"><a href="#cb51-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb51-10"><a href="#cb51-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb51-11"><a href="#cb51-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb51-12"><a href="#cb51-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb51-13"><a href="#cb51-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb51-14"><a href="#cb51-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The shape parameter of the Gamma distribution (k &gt; 0).</p></li>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter of the Gamma distribution (theta &gt; 0).</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-26" class="level4">
<h4 class="anchored" data-anchor-id="returns-26">Returns:</h4>
<p>Gamma: A BI Gamma distribution object (for model building).</p>
<p>jnp.ndarray: A JAX array of samples drawn from the Gamma distribution (for direct sampling).</p>
<p>Gamma: The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-24" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-24">Example Usage:</h4>
<div class="sourceCode" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>m.dist.gamma(concentration<span class="op">=</span><span class="fl">2.0</span>, rate<span class="op">=</span><span class="fl">0.5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-18" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-18">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gamma</p>
</section>
<section id="references-15" class="level4">
<h4 class="anchored" data-anchor-id="references-15">References</h4>
<p>https://reference.wolfram.com/language/ref/GammaDistribution.html https://en.wikipedia.org/wiki/Gamma_distribution</p>
<hr>
</section>
</section>
<section id="gamma-poisson" class="level3">
<h3 class="anchored" data-anchor-id="gamma-poisson">Gamma Poisson</h3>
<p>The Gamma-Poisson distribution, also known as the Negative Binomial distribution, models overdispersed count data. It arises from a hierarchical process where the rate parameter of a Poisson distribution is itself a random variable following a Gamma distribution. This structure allows the model to capture variability in count data that exceeds what is predicted by the Poisson distribution, making it suitable for applications like modeling RNA-sequencing data and microbial count.</p>
<p><span class="math display">
f(x) = \frac{\Gamma(x + \beta) \alpha^x}{\Gamma(\beta)(1 + \alpha)^{\beta + x} x!}, \quad x = 0, 1, 2, \ldots
</span> Where <span class="math inline">\Gamma</span> denotes the gamma function.</p>
<p>This CDF expresses the probability that the count variable ( X ) takes a value less than or equal to ( x ), incorporating the overdispersion introduced by the Gamma distribution.</p>
<section id="args-27" class="level4">
<h4 class="anchored" data-anchor-id="args-27">Args:</h4>
<div class="sourceCode" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gamma_poisson(</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): Shape parameter (alpha) of the Gamma distribution.</p></li>
<li><p><em>rate</em> (jnp.ndarray): Rate parameter (beta) for the Gamma distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-27" class="level4">
<h4 class="anchored" data-anchor-id="returns-27">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI GammaPoisson distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the GammaPoisson distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-25" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-25">Example Usage:</h4>
<div class="sourceCode" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>m.dist.gamma_poisson(concentration<span class="op">=</span><span class="fl">1.0</span>, rate<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-19" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-19">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gammapoisson</p>
</section>
<section id="references-16" class="level4">
<h4 class="anchored" data-anchor-id="references-16">References</h4>
<p>[1] https://www.statisticshowto.com/gamma-poisson-distribution https://en.wikipedia.org/wiki/Negative_binomial_distribution</p>
<hr>
</section>
</section>
<section id="gaussian-copula" class="level3">
<h3 class="anchored" data-anchor-id="gaussian-copula">Gaussian Copula</h3>
<p>A distribution that links the <code>batch_shape[:-1]</code> of a marginal distribution with a multivariate Gaussian copula, modelling the correlation between the axes. A copula is a multivariate distribution over the uniform distribution on [0, 1]. The Gaussian copula links the marginal distributions through a multivariate normal distribution.</p>
<section id="args-28" class="level4">
<h4 class="anchored" data-anchor-id="args-28">Args:</h4>
<div class="sourceCode" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gaussian_copula(</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>marginal_dist,</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>correlation_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>correlation_cholesky<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb55-10"><a href="#cb55-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb55-11"><a href="#cb55-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb55-12"><a href="#cb55-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb55-13"><a href="#cb55-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb55-14"><a href="#cb55-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb55-15"><a href="#cb55-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>marginal_dist</em> (Distribution): Distribution whose last batch axis is to be coupled.</p></li>
<li><p><em>correlation_matrix</em> (array_like, optional): Correlation matrix of the coupling multivariate normal distribution. Defaults to None.</p></li>
<li><p><em>correlation_cholesky</em> (array_like, optional): Correlation Cholesky factor of the coupling multivariate normal distribution. Defaults to None.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. Defaults to None.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>Mi  xtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-28" class="level4">
<h4 class="anchored" data-anchor-id="returns-28">Returns:</h4>
<p>BI GaussianCopula distribution object: When <code>sample=False</code> (for model building). JAX array: When <code>sample=True</code> (for direct sampling). BI distribution object: When <code>create_obj=True</code> (for advanced use cases).</p>
</section>
<section id="example-usage-26" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-26">Example Usage:</h4>
<div class="sourceCode" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>m.dist.gaussian_copula(</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>marginal_dist <span class="op">=</span> m.dist.beta(<span class="fl">2.0</span>, <span class="fl">5.0</span>, create_obj <span class="op">=</span> <span class="va">True</span>), </span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>correlation_matrix <span class="op">=</span> jnp.array([[<span class="fl">1.0</span>, <span class="fl">0.7</span>],[<span class="fl">0.7</span>, <span class="fl">1.0</span>]]), </span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>sample <span class="op">=</span> <span class="va">True</span></span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-20" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-20">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gaussiancopula</p>
<hr>
</section>
</section>
<section id="gaussian-copula-beta" class="level3">
<h3 class="anchored" data-anchor-id="gaussian-copula-beta">Gaussian Copula Beta</h3>
<p>This distribution combines a Gaussian copula with a Beta distribution. The Gaussian copula models the dependence structure between random variables, while the Beta distribution defines the marginal distributions of each variable.</p>
<section id="args-29" class="level4">
<h4 class="anchored" data-anchor-id="args-29">Args:</h4>
<div class="sourceCode" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gaussian_copula_beta(</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>concentration1,</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>concentration0,</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>correlation_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>correlation_cholesky<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration1</em> (jnp.ndarray): The first shape parameter of the Beta distribution.</p></li>
<li><p><em>concentration0</em> (jnp.ndarray): The second shape parameter of the Beta distribution.</p></li>
<li><p><em>correlation_matrix</em> (array_like, optional): Correlation matrix of the coupling multivariate normal distribution. Defaults to None.</p></li>
<li><p><em>correlation_cholesky</em> (jnp.ndarray): The Cholesky decomposition of the correlation matrix.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-29" class="level4">
<h4 class="anchored" data-anchor-id="returns-29">Returns:</h4>
<p>GaussianCopulaBeta: A BI GaussianCopulaBeta distribution object (for model building). jnp.ndarray: A JAX array of samples drawn from the GaussianCopulaBeta distribution (for direct sampling). Distribution: The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-27" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-27">Example Usage:</h4>
<div class="sourceCode" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>m.dist.gaussian_copula_beta(</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>concentration1 <span class="op">=</span> jnp.array([<span class="fl">2.0</span>, <span class="fl">3.0</span>]), </span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>concentration0 <span class="op">=</span> jnp.array([<span class="fl">5.0</span>, <span class="fl">3.0</span>]),</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>correlation_cholesky <span class="op">=</span> jnp.linalg.cholesky(jnp.array([[<span class="fl">1.0</span>, <span class="fl">0.7</span>],[<span class="fl">0.7</span>, <span class="fl">1.0</span>]])), </span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>sample <span class="op">=</span> <span class="va">True</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-21" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-21">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gaussiancopulabetadistribution</p>
<hr>
</section>
</section>
<section id="gaussian-random-walk" class="level3">
<h3 class="anchored" data-anchor-id="gaussian-random-walk">Gaussian Random Walk</h3>
<p>Creates a distribution over a Gaussian random walk of a specified number of steps. This is a discrete-time stochastic process where the value at each step is the previous value plus a Gaussian-distributed increment. The distribution is over the entire path.</p>
<section id="args-30" class="level4">
<h4 class="anchored" data-anchor-id="args-30">Args:</h4>
<div class="sourceCode" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gaussian_random_walk(</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>num_steps<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb59-14"><a href="#cb59-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (float, jnp.ndarray, optional): The standard deviation ($`) of the Gaussian increments. Must be positive. Defaults to 1.0.</p></li>
<li><p>*num_steps (int, optional): The number of steps in the random walk, which determines the event shape of the distribution. Must be positive. Defaults to 1.</p></li>
<li><p>*validate_args (bool, optional): Whether to enable validation of distribution parameters. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. If provided, events with a <code>True</code> mask will be conditioned on <code>obs</code>, while the remaining events will be treated as latent variables. Defaults to <code>None</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>. Defaults to 0.</p></li>
<li><p><em>shape</em> (tuple, optional): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to <code>False</code>.</p></li>
</ul>
</section>
<section id="returns-30" class="level4">
<h4 class="anchored" data-anchor-id="returns-30">Returns:</h4>
<p>BI.primitives.Messenger: A BI sample site object when used in a model context (<code>sample=False</code>). jnp.ndarray: A JAX array of samples drawn from the GaussianRandomWalk distribution (for direct sampling, <code>sample=True</code>). numpyro.distributions.Distribution: The raw BI distribution object (if <code>create_obj=True</code>).</p>
</section>
<section id="example-usage-28" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-28">Example Usage:</h4>
<div class="sourceCode" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> jax.numpy <span class="im">as</span> jnp</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Direct sampling of a random walk with 100 steps</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> m.dist.gaussian_random_walk(scale<span class="op">=</span><span class="fl">0.5</span>, num_steps<span class="op">=</span><span class="dv">100</span>, sample<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage within a model for a latent time series</span></span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> my_model(data<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Prior on the volatility of the random walk</span></span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>volatility <span class="op">=</span> m.dist.half_cauchy(scale<span class="op">=</span><span class="fl">1.0</span>, name<span class="op">=</span><span class="st">'volatility'</span>)</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a><span class="co"># The latent random walk</span></span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>latent_process <span class="op">=</span> m.dist.gaussian_random_walk(</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span>volatility,</span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>num_steps<span class="op">=</span><span class="bu">len</span>(data) <span class="cf">if</span> data <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> <span class="dv">10</span>,</span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'latent_process'</span></span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Observation model</span></span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Assumes the observed data is the latent process plus some noise</span></span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a>obs_noise <span class="op">=</span> m.dist.half_cauchy(scale<span class="op">=</span><span class="fl">1.0</span>, name<span class="op">=</span><span class="st">'obs_noise'</span>)</span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> m.plate(<span class="st">'time'</span>, <span class="bu">len</span>(data) <span class="cf">if</span> data <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span> <span class="cf">else</span> <span class="dv">10</span>):</span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a><span class="cf">return</span> m.dist.normal(loc<span class="op">=</span>latent_process, scale<span class="op">=</span>obs_noise, obs<span class="op">=</span>data, name<span class="op">=</span><span class="st">'obs'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-22" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-22">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gaussianrandomwalk</p>
<hr>
</section>
</section>
<section id="gaussian-state-space" class="level3">
<h3 class="anchored" data-anchor-id="gaussian-state-space">Gaussian State Space</h3>
<p>Samples from a Gaussian state space model.</p>
<section id="args-31" class="level4">
<h4 class="anchored" data-anchor-id="args-31">Args:</h4>
<div class="sourceCode" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gaussian_state_space(</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>num_steps,</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>transition_matrix,</span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>covariance_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>precision_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>num_steps</em> (int): Number of steps.</p></li>
<li><p><em>transition_matrix</em> (jnp.ndarray): State space transition matrix ${A}`.</p></li>
<li><p><em>covariance_matrix</em> (jnp.ndarray, optional): Covariance of the innovation noise <span class="math inline">\epsilon</span>. Defaults to None.</p></li>
<li><p>*precision_matrix (jnp.ndarray, optional): Precision matrix of the innovation noise <span class="math inline">\epsilon</span>. Defaults to None.</p></li>
<li><p>*scale_tril (jnp.ndarray, optional): Scale matrix of the innovation noise <span class="math inline">{\epsilon}</span>. Defaults to None.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. Defaults to None.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-31" class="level4">
<h4 class="anchored" data-anchor-id="returns-31">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI GaussianStateSpace distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the GaussianStateSpace distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-29" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-29">Example Usage:</h4>
<div class="sourceCode" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>m.dist.gaussian_state_space(num_steps<span class="op">=</span><span class="dv">5</span>, transition_matrix<span class="op">=</span>jnp.array([[<span class="fl">0.5</span>]]), covariance_matrix <span class="op">=</span>  jnp.array([[<span class="fl">1.0</span>, <span class="fl">0.6</span>],[<span class="fl">0.6</span>, <span class="fl">1.0</span>]]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-23" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-23">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gaussianstate</p>
<hr>
</section>
</section>
<section id="geometric-distribution." class="level3">
<h3 class="anchored" data-anchor-id="geometric-distribution.">Geometric distribution.</h3>
<p>The Geometric distribution models the number of independent Bernoulli trials needed to obtain the first success (or equivalently the number of failures before the first success). Each trial has the same probability of success <span class="math inline">p</span>, and trials are independent.</p>
<ul>
<li><span class="math inline">X</span> = number of trials until the first success (so support <span class="math inline">1,2,3,\dots</span>).</li>
</ul>
<p><span class="math display">
P(X = k) = (1 - p)^{,k-1} ; p,\quad k = 1,2,\dots
</span></p>
<section id="args-32" class="level4">
<h4 class="anchored" data-anchor-id="args-32">Args:</h4>
<div class="sourceCode" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>bi.dist.geometric(</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb63-13"><a href="#cb63-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb63-14"><a href="#cb63-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray, optional): Probability of success on each trial. Must be between 0 and 1.</p></li>
<li><p><em>logits</em> (jnp.ndarray, optional): Log-odds of success on each trial. <code>probs = jax.nn.sigmoid(logits)</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-32" class="level4">
<h4 class="anchored" data-anchor-id="returns-32">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Geometric distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Geometric distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-30" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-30">Example Usage:</h4>
<div class="sourceCode" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>m.dist.geometric(probs<span class="op">=</span><span class="fl">0.5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlgeometric" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlgeometric">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#geometric</h4>
</section>
<section id="references-17" class="level4">
<h4 class="anchored" data-anchor-id="references-17">References:</h4>
<ul>
<li>https://www.britannica.com/topic/geometric-distribution “Geometric distribution | Definition, Formula, Examples, Illustration, &amp; Applications | Britannica”</li>
<li>https://en.wikipedia.org/wiki/Geometric_distribution “Geometric distribution”</li>
<li>https://distribution-explorer.github.io/discrete/geometric.html “Geometric distribution — Probability Distribution Explorer documentation”</li>
</ul>
<hr>
</section>
</section>
<section id="geometriclogits" class="level3">
<h3 class="anchored" data-anchor-id="geometriclogits">GeometricLogits</h3>
<p>Samples from a GeometricLogits distribution, which models the number of failures before the first success in a sequence of independent Bernoulli trials. It is parameterized by logits, which are transformed into probabilities using the sigmoid function.</p>
<p><span class="math display">
P(X = k) = (1 - p)^k , p,
</span></p>
<p>where:</p>
<ul>
<li>X is the number of failures before the first success.</li>
<li>k is the number of failures.</li>
<li>p is the probability of success on each trial (derived from the logits).</li>
</ul>
<section id="args-33" class="level4">
<h4 class="anchored" data-anchor-id="args-33">Args:</h4>
<div class="sourceCode" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>bi.dist.geometric_logits(</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>logits</em> (jnp.ndarray): Log-odds parameterization of the probability of success.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-33" class="level4">
<h4 class="anchored" data-anchor-id="returns-33">Returns:</h4>
<p>BI GeometricLogits distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the GeometricLogits distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-31" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-31">Example Usage:</h4>
<div class="sourceCode" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>m.dist.geometric_logits(logits<span class="op">=</span>jnp.zeros(<span class="dv">10</span>), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-24" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-24">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#geometriclogits</p>
</section>
<section id="references-18" class="level4">
<h4 class="anchored" data-anchor-id="references-18">References:</h4>
<p>https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/Geometric</p>
<hr>
</section>
</section>
<section id="geometricprobs" class="level3">
<h3 class="anchored" data-anchor-id="geometricprobs">GeometricProbs</h3>
<p>Samples from a Geometric</p>
<p>The Geometric distribution models the number of trials until the first success in a sequence of independent Bernoulli trials, where each trial has the same probability of success.</p>
<p><span class="math display">
P(X = k) = (1 - p)^k p
</span></p>
<section id="args-34" class="level4">
<h4 class="anchored" data-anchor-id="args-34">Args:</h4>
<div class="sourceCode" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>bi.dist.geometric_probs(</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): Probability of success on each trial. Must be between 0 and 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-34" class="level4">
<h4 class="anchored" data-anchor-id="returns-34">Returns:</h4>
<p>BI GeometricProbs distribution object (for model building). JAX array of samples drawn from the GeometricProbs distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-32" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-32">Example Usage:</h4>
<div class="sourceCode" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>m.dist.geometric_probs(probs<span class="op">=</span><span class="fl">0.5</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-25" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-25">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#geometricprobs</p>
<hr>
</section>
</section>
<section id="gompertz" class="level3">
<h3 class="anchored" data-anchor-id="gompertz">Gompertz</h3>
<p>The Gompertz distribution is a distribution with support on the positive real line that is closely related to the Gumbel distribution. This implementation follows the notation used in the Wikipedia entry for the Gompertz distribution. See https://en.wikipedia.org/wiki/Gompertz_distribution.</p>
<p>The probability density function (PDF) is:</p>
<p><span class="math display">
F(x) = 1 - \exp!\bigl[-a (e^x - 1)\bigr], \quad x \ge 0.
</span></p>
<section id="args-35" class="level4">
<h4 class="anchored" data-anchor-id="args-35">Args:</h4>
<div class="sourceCode" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gompertz(</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter. Must be positive.</p></li>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-35" class="level4">
<h4 class="anchored" data-anchor-id="returns-35">Returns:</h4>
<p>BI Gompertz distribution object: When <code>sample=False</code> (for model building). JAX array: When <code>sample=True</code> (for direct sampling). BI distribution object: When <code>create_obj=True</code> (for advanced use cases).</p>
</section>
<section id="example-usage-33" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-33">Example Usage:</h4>
<div class="sourceCode" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>m.dist.gompertz(concentration<span class="op">=</span><span class="fl">1.0</span>, rate<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-26" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-26">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gompertz #### References: https://en.wikipedia.org/wiki/Gompertz_distribution</p>
<hr>
</section>
</section>
<section id="gumbel" class="level3">
<h3 class="anchored" data-anchor-id="gumbel">Gumbel</h3>
<p>Samples from a Gumbel (or Extreme Value) distribution.</p>
<p>The Gumbel distribution is a continuous probability distribution named after German mathematician Carl Gumbel. It is often used to model the distribution of maximum values in a sequence of independent random variables.</p>
<p><span class="math display">
F(x;\mu,\beta) = \exp\bigl( -\exp\bigl( -\tfrac{x - \mu}{\beta} \bigr) \bigr), \quad x \in {R}.
</span></p>
<section id="args-36" class="level4">
<h4 class="anchored" data-anchor-id="args-36">Args:</h4>
<div class="sourceCode" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>bi.dist.gumbel(</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray or float, optional): Location parameter. Defaults to 0.0.</p></li>
<li><p><em>scale</em> (jnp.ndarray or float, optional): Scale parameter. Must be positive. Defaults to 1.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as event dimensions (used in model building). Defaults to None.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. Defaults to None.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-36" class="level4">
<h4 class="anchored" data-anchor-id="returns-36">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Gumbel distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Gumbel distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-34" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-34">Example Usage:</h4>
<div class="sourceCode" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>m.dist.gumbel(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-27" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-27">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#gumbel #### References: https://en.wikipedia.org/wiki/Gumbel_distribution</p>
<hr>
</section>
</section>
<section id="halfcauchy" class="level3">
<h3 class="anchored" data-anchor-id="halfcauchy">HalfCauchy</h3>
<p>The HalfCauchy distribution is a probability distribution that is half of the Cauchy distribution. It is defined on the positive real numbers and is often used in situations where only positive values are relevant.</p>
<p><span class="math display">   
f(x) = \frac{1}{2} \cdot \frac{1}{\pi \cdot \frac{1}{scale} \cdot (x^2 + \frac{1}{scale^2})}
</span></p>
<section id="args-37" class="level4">
<h4 class="anchored" data-anchor-id="args-37">Args:</h4>
<div class="sourceCode" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>bi.dist.half_cauchy(</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (jnp.ndarray): The scale parameter of the Cauchy distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-37" class="level4">
<h4 class="anchored" data-anchor-id="returns-37">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI HalfCauchy distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the HalfCauchy distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-35" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-35">Example Usage:</h4>
<div class="sourceCode" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>m.dist.half_cauchy(scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlhalfcauchy" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlhalfcauchy">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#halfcauchy</h4>
<hr>
</section>
</section>
<section id="halfnormal" class="level3">
<h3 class="anchored" data-anchor-id="halfnormal">HalfNormal</h3>
<p>Samples from a HalfNormal distribution.</p>
<p>The HalfNormal distribution is a distribution of the absolute value of a normal random variable. It is defined by a location parameter (implicitly 0) and a scale parameter.</p>
<p><span class="math display">
f(x; \sigma) = \frac{2}{\pi,\sigma} ; \frac{1}{1 + \bigl(x/\sigma\bigr)^2}, \qquad x \ge 0.
</span></p>
<section id="args-38" class="level4">
<h4 class="anchored" data-anchor-id="args-38">Args:</h4>
<div class="sourceCode" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>bi.dist.half_normal(</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb75-7"><a href="#cb75-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb75-8"><a href="#cb75-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb75-9"><a href="#cb75-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb75-10"><a href="#cb75-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb75-11"><a href="#cb75-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb75-12"><a href="#cb75-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb75-13"><a href="#cb75-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (float, array): The scale parameter of the distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-38" class="level4">
<h4 class="anchored" data-anchor-id="returns-38">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI HalfNormal distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the HalfNormal distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-36" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-36">Example Usage:</h4>
<div class="sourceCode" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>m.dist.half_normal(scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-28" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-28">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#halfnormal</p>
</section>
<section id="references-19" class="level4">
<h4 class="anchored" data-anchor-id="references-19">References:</h4>
<p>https://distribution-explorer.github.io/continuous/halfcauchy.html</p>
<hr>
</section>
</section>
<section id="improper-uniform" class="level3">
<h3 class="anchored" data-anchor-id="improper-uniform">Improper Uniform</h3>
<p>A helper distribution with zero :meth:<code>log_prob</code> over the <code>support</code> domain.</p>
<section id="args-39" class="level4">
<h4 class="anchored" data-anchor-id="args-39">Args:</h4>
<div class="sourceCode" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>bi.dist.improper_uniform(</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>support,</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a>batch_shape,</span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>event_shape,</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb77-10"><a href="#cb77-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb77-11"><a href="#cb77-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb77-12"><a href="#cb77-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb77-13"><a href="#cb77-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb77-14"><a href="#cb77-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb77-15"><a href="#cb77-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>support</em> (numpyro.distributions.constraints.Constraint): The support of this distribution.</p></li>
<li><p><em>batch_shape</em> (tuple): Batch shape of this distribution. It is usually safe to set <code>batch_shape=()</code>.</p></li>
<li><p><em>event_shape</em> (tuple): Event shape of this distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-39" class="level4">
<h4 class="anchored" data-anchor-id="returns-39">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI ImproperUniform distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the ImproperUniform distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-37" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-37">Example Usage:</h4>
<div class="sourceCode" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> sample</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> numpyro.distributions <span class="im">import</span> ImproperUniform, Normal, constraints</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> model():</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> sample(<span class="st">'x'</span>, ImproperUniform(constraints.ordered_vector, (), event_shape<span class="op">=</span>(<span class="dv">10</span>,)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-29" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-29">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#improperuniform</p>
<hr>
</section>
</section>
<section id="inverse-gamma" class="level3">
<h3 class="anchored" data-anchor-id="inverse-gamma">Inverse Gamma</h3>
<p>The InverseGamma distribution is a two-parameter family of continuous probability distributions. It is defined by its shape <span class="math inline">\alpha</span> and rate <span class="math inline">\beta</span> parameters. It is often used as a prior distribution for precision parameters (inverse variance) in Bayesian statistics.</p>
<p><span class="math display">
f(x; \alpha, \beta) = \frac{\beta^\alpha}{\Gamma(\alpha)} x^{-(\alpha + 1)} \exp\left(-\frac{\beta}{x}\right), \quad x &gt; 0
</span></p>
<p>where: * (()) is the Gamma function.</p>
<section id="args-40" class="level4">
<h4 class="anchored" data-anchor-id="args-40">Args:</h4>
<div class="sourceCode" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>bi.dist.inverse_gamma(</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter <span class="math inline">\alpha</span> of the InverseGamma distribution. Must be positive.</p></li>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter <span class="math inline">\beta</span> of the InverseGamma distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-40" class="level4">
<h4 class="anchored" data-anchor-id="returns-40">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI InverseGamma distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the InverseGamma distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-38" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-38">Example Usage:</h4>
<div class="sourceCode" id="cb80"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>m.dist.inverse_gamma(concentration<span class="op">=</span><span class="fl">2.0</span>, rate<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-30" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-30">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#inversegamma</p>
<hr>
</section>
</section>
<section id="kumaraswamy" class="level3">
<h3 class="anchored" data-anchor-id="kumaraswamy">Kumaraswamy</h3>
<p>The Kumaraswamy distribution is a continuous probability distribution defined on the interval [0, 1]. It is a flexible distribution that can take on various shapes depending on its parameters.</p>
<p><span class="math display">
f(x; a, b) = a b x^{a-1} (1 - x^a)^{b-1}, \quad x \in (0, 1)
</span></p>
<p>where: * ( a ) and ( b ) are shape parameters. * ( x ) is the random variable.</p>
<section id="args-41" class="level4">
<h4 class="anchored" data-anchor-id="args-41">Args:</h4>
<div class="sourceCode" id="cb81"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>bi.dist.kumaraswamy(</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>concentration1,</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>concentration0,</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb81-8"><a href="#cb81-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb81-9"><a href="#cb81-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb81-10"><a href="#cb81-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb81-11"><a href="#cb81-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb81-12"><a href="#cb81-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb81-13"><a href="#cb81-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb81-14"><a href="#cb81-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration1</em> (jnp.ndarray): The first shape parameter. Must be positive.</p></li>
<li><p><em>concentration0</em> (jnp.ndarray): The second shape parameter. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-41" class="level4">
<h4 class="anchored" data-anchor-id="returns-41">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Kumaraswamy distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Kumaraswamy distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-39" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-39">Example Usage:</h4>
<div class="sourceCode" id="cb82"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>m.dist.kumaraswamy(concentration1<span class="op">=</span><span class="fl">2.0</span>, concentration0<span class="op">=</span><span class="fl">3.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-31" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-31">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#kumaraswamy #### Rerferences: https://en.wikipedia.org/wiki/Kumaraswamy_distribution</p>
<hr>
</section>
</section>
<section id="laplace" class="level3">
<h3 class="anchored" data-anchor-id="laplace">Laplace</h3>
<p>Samples from a Laplace distribution, also known as the double exponential distribution. The Laplace distribution is defined by its location parameter (loc) and scale parameter (scale).</p>
<p><span class="math display">
f(x \mid \mu, b) = \frac{1}{2b} \exp\left( -\frac{|x - \mu|}{b} \right)
</span></p>
<p>Where: * (<span class="math inline">\mu</span>) is the <strong>location parameter</strong>, indicating the peak of the distribution. * (b &gt; 0) is the <strong>scale parameter</strong>, controlling the spread of the distribution.</p>
<section id="args-42" class="level4">
<h4 class="anchored" data-anchor-id="args-42">Args:</h4>
<div class="sourceCode" id="cb83"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>bi.dist.laplace(</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray): Location parameter of the Laplace distribution.</p></li>
<li><p><em>sample</em> (jnp.ndarray): Scale parameter of the Laplace distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-42" class="level4">
<h4 class="anchored" data-anchor-id="returns-42">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Laplace distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Laplace distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-40" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-40">Example Usage:</h4>
<div class="sourceCode" id="cb84"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>m.dist.laplace(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-32" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-32">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#laplace</p>
<hr>
</section>
</section>
<section id="left-truncated" class="level3">
<h3 class="anchored" data-anchor-id="left-truncated">Left Truncated</h3>
<p>Samples from a left-truncated distribution.</p>
<p>A left-truncated distribution is a probability distribution obtained by restricting the support of another distribution to values greater than a specified lower bound. This is useful when dealing with data that is known to be greater than a certain value. All the “mass” below (or equal to) (<span class="math inline">a</span>) is <strong>excluded</strong> (not just unobserved, but removed from the sample/analysis).</p>
<!--
Let (X) be a random variable with PDF ($f_X(x)$) and CDF ($F_X(x)$). Choose a truncation point ($a$). Then define a new random variable $Y = X \mid (X > a)$. The PDF and CDF of ($Y$) are:

* **PDF** for $ y > a $:
$$
f_Y(y) = \frac{f_X(y)}{1 - F_X(a)}
$$


(Here $1 - F_X(a) = P(X > a)$ is the normalising constant)
-->
<section id="args-43" class="level4">
<h4 class="anchored" data-anchor-id="args-43">Args:</h4>
<div class="sourceCode" id="cb85"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>bi.dist.left_truncated_distribution(</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>base_dist</em>: The base distribution to truncate. Must be univariate and have real support.</p></li>
<li><p><em>low</em>: The lower truncation bound. Values less than this are excluded from the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-43" class="level4">
<h4 class="anchored" data-anchor-id="returns-43">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI LeftTruncatedDistribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the LeftTruncatedDistribution distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmllefttruncateddistribution" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmllefttruncateddistribution">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#lefttruncateddistribution</h4>
</section>
<section id="references-20" class="level4">
<h4 class="anchored" data-anchor-id="references-20">References:</h4>
<p>https://en.wikipedia.org/wiki/Truncated_distribution https://encyclopediaofmath.org/wiki/Truncated_distribution</p>
<hr>
</section>
</section>
<section id="levy" class="level3">
<h3 class="anchored" data-anchor-id="levy">Levy</h3>
<p>Samples from a Levy distribution. The Lévy distribution is a continuous probability distribution on the positive real line (or shifted positive line) that is <em>heavy-tailed</em>, <em>skewed</em>, and arises naturally in connection with stable distributions (specifically the case with stability index <span class="math inline">\alpha = \tfrac12</span>. It is often used in contexts such as hitting‐time problems for Brownian motion, physics (e.g., van der Waals line‐shapes), and modelling very heavy‐tailed phenomena. Let (X) be a Lévy‐distributed random variable with <strong>location</strong> parameter <span class="math inline">\mu</span> and <strong>scale</strong> parameter (c &gt; 0). The support is <span class="math inline">x \ge \mu</span>.</p>
<p><strong>PDF</strong> <span class="math display">
f(x; \mu, c) = \sqrt{\frac{c}{2\pi}} ; \frac{1}{(x - \mu)^{3/2}} ; \exp!\Bigl( -\frac{c}{2 (x - \mu)} \Bigr),
\quad x \ge \mu.
</span></p>
<section id="args-44" class="level4">
<h4 class="anchored" data-anchor-id="args-44">Args:</h4>
<div class="sourceCode" id="cb86"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>bi.dist.levy(</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>loc,</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>scale,</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray): Location parameter.</p></li>
<li><p><em>sample</em> (jnp.ndarray): Scale parameter.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-44" class="level4">
<h4 class="anchored" data-anchor-id="returns-44">Returns:</h4>
<p>BI Levy distribution object: When <code>sample=False</code> (for model building). JAX array: When <code>sample=True</code> (for direct sampling). BI distribution object: When <code>create_obj=True</code> (for advanced use cases).</p>
</section>
<section id="example-usage-41" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-41">Example Usage:</h4>
<div class="sourceCode" id="cb87"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>m.dist.levy(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-33" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-33">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#levy</p>
<hr>
</section>
</section>
<section id="lewandowski-kurowicka-joe-lkj" class="level3">
<h3 class="anchored" data-anchor-id="lewandowski-kurowicka-joe-lkj">Lewandowski Kurowicka Joe (LKJ)</h3>
<p>The LKJ distribution is controlled by the concentration parameter <span class="math inline">\eta</span> to make the probability of the correlation matrix $M` proportional to <span class="math inline">\det(M)^{\eta - 1}</span>. When <span class="math inline">\eta = 1</span>, the distribution is uniform over correlation matrices. When <span class="math inline">\eta &gt; 1</span>, the distribution favors samples with large determinants. When <span class="math inline">\eta &lt; 1</span>, the distribution favors samples with small determinants.</p>
<p><span class="math display">
P(M) \propto |\det(M)|^{\eta - 1}
</span></p>
<section id="args-45" class="level4">
<h4 class="anchored" data-anchor-id="args-45">Args:</h4>
<div class="sourceCode" id="cb88"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>bi.dist.lkj(</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>dimension,</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>concentration<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>sample_method<span class="op">=</span><span class="st">'onion'</span>,</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb88-8"><a href="#cb88-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb88-9"><a href="#cb88-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb88-10"><a href="#cb88-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb88-11"><a href="#cb88-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb88-12"><a href="#cb88-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb88-13"><a href="#cb88-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb88-14"><a href="#cb88-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb88-15"><a href="#cb88-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>dimension</em> (int): The dimension of the correlation matrices.</p></li>
<li><p><em>concentration</em> (ndarray): The concentration/shape parameter of the distribution (often referred to as eta). Must be positive.</p></li>
<li><p><em>sample_method</em> (str): Either “cvine” or “onion”. Both methods are proposed in [1] and offer the same distribution over correlation matrices. But they are different in how to generate samples. Defaults to “onion”.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-45" class="level4">
<h4 class="anchored" data-anchor-id="returns-45">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI LKJ distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the LKJ distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-42" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-42">Example Usage:</h4>
<div class="sourceCode" id="cb89"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>m.dist.lkj(dimension<span class="op">=</span><span class="dv">2</span>, concentration<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-34" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-34">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#lkj #### References: https://en.wikipedia.org/wiki/Kumaraswamy_distribution</p>
<hr>
</section>
</section>
<section id="lkj-cholesky" class="level3">
<h3 class="anchored" data-anchor-id="lkj-cholesky">LKJ Cholesky</h3>
<p>The LKJ (Leonard-Kjærgaard-Jørgensen) Cholesky distribution is a family of distributions on symmetric matrices, often used as a prior for the Cholesky decomposition of a symmetric matrix. It is particularly useful in Bayesian inference for models with covariance structure. Given a lower triangular matrix (L) with unit diagonal entries, the PDF is: <span class="math display">
p(L | \eta) \propto \prod_{k=2}^d L_{kk}^{d-k+2\eta-2}
</span></p>
<p>where: * <span class="math inline">L_{kk}</span> is the diagonal element of (L). * <span class="math inline">\eta</span> is the shape parameter. * <span class="math inline">d</span> is the dimension of the matrix.</p>
<ul>
<li><span class="math inline">\eta = 1</span>: Uniform prior over correlation matrices.</li>
<li><span class="math inline">\eta &gt; 1</span>: Prior favors correlation matrices close to the identity matrix (i.e., weak correlations).</li>
<li><span class="math inline">\eta &lt; 1</span>: Prior allows stronger correlations.</li>
</ul>
<section id="args-46" class="level4">
<h4 class="anchored" data-anchor-id="args-46">Args:</h4>
<div class="sourceCode" id="cb90"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a>bi.dist.lkj_cholesky(</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>dimension,</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>concentration<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>sample_method<span class="op">=</span><span class="st">'onion'</span>,</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p>*dimension (int): The dimension of the correlation matrices.</p></li>
<li><p>*concentration (float): A parameter controlling the concentration of the distribution around the identity matrix. Higher values indicate greater concentration. Must be greater than 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
<p>Attributes: concentration (float): The concentration parameter.</p>
</section>
<section id="references-21" class="level4">
<h4 class="anchored" data-anchor-id="references-21">References:</h4>
<p>https://mc-stan.org/docs/2_21/functions-reference/cholesky-lkj-correlation-distribution.html</p>
<hr>
</section>
</section>
<section id="log-normal" class="level3">
<h3 class="anchored" data-anchor-id="log-normal">Log-Normal</h3>
<p>The Log-Normal distribution is a probability distribution defined for positive real-valued random variables, parameterized by a location parameter (loc : <span class="math inline">\mu</span>) and a scale parameter (scale). It arises when the logarithm of a random variable is normally distributed. * A random variable <span class="math inline">X</span> is <strong>log-normal</strong> if its natural logarithm <span class="math inline">\ln X</span> is (approximately) normally distributed. Equivalently, one can write: <span class="math display">
X = \exp(Y), \quad \text{where } Y \sim {N}(\mu, \sigma^2).
</span></p>
<p>Because <span class="math inline">X</span> is the exponential of a normal, it is always positive (support <span class="math inline">x&gt;0</span>). It is useful in contexts where growth, multiplicative effects, or compounding factors dominate (e.g.&nbsp;stock prices, income, sizes of biological organisms).</p>
<section id="parameters" class="level4">
<h4 class="anchored" data-anchor-id="parameters">Parameters</h4>
<ul>
<li><span class="math inline">\mu</span>: the mean of <span class="math inline">\ln X</span> (i.e.&nbsp;the location parameter in log-space)</li>
<li><span class="math inline">\sigma &gt; 0</span>: the standard deviation of <span class="math inline">\ln X</span> (i.e.&nbsp;the “scale” in log-space)</li>
</ul>
<p>Sometimes another parameterization uses a threshold (shift) <span class="math inline">\theta</span>, i.e.&nbsp;<span class="math inline">X = \theta + \exp(Y)</span>. For the two-parameter (no shift) case:</p>
<ul>
<li><strong>PDF</strong> <span class="math display">
f(x; \mu, \sigma) = \frac{1}{x,\sigma\sqrt{2\pi}} ; \exp!\Bigl( -\frac{(\ln x - \mu)^2}{2\sigma^2} \Bigr), \quad x &gt; 0.
</span></li>
</ul>
<hr>
</section>
<section id="args-47" class="level4">
<h4 class="anchored" data-anchor-id="args-47">Args:</h4>
<div class="sourceCode" id="cb91"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>bi.dist.log_normal(</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb91-12"><a href="#cb91-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb91-13"><a href="#cb91-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb91-14"><a href="#cb91-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (float): Location parameter.</p></li>
<li><p><em>scale</em> (float): Scale parameter.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-46" class="level4">
<h4 class="anchored" data-anchor-id="returns-46">Returns:</h4>
<p>BI LogNormal distribution object (for model building). JAX array of samples drawn from the LogNormal distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-43" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-43">Example Usage:</h4>
<div class="sourceCode" id="cb92"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>m.dist.log_normal(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmllognormal" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmllognormal">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#lognormal</h4>
</section>
<section id="references-22" class="level4">
<h4 class="anchored" data-anchor-id="references-22">References:</h4>
<p>https://en.wikipedia.org/wiki/Log-normal_distribution</p>
<hr>
</section>
</section>
<section id="log-uniform" class="level3">
<h3 class="anchored" data-anchor-id="log-uniform">Log-Uniform</h3>
<p>Samples from a Log Uniform distribution.</p>
<p>The Log Uniform distribution is defined over the positive real numbers and is the result of applying an exponential transformation to a uniform distribution over the interval [low, high]. It is often used when modeling parameters that must be positive.</p>
<p>A random variable <span class="math inline">X</span> is <strong>log-uniform</strong> on <span class="math inline">[a, b]</span>, with <span class="math inline">0 &lt; a &lt; b</span>, if <span class="math inline">\ln X</span> is uniformly distributed on <span class="math inline">[\ln a, \ln b]</span>. Equivalently, the density of <span class="math inline">X</span> is proportional to <span class="math inline">1/x</span> over that interval. This distribution is sometimes called the <em>reciprocal distribution</em>. It is useful in modeling scales spanning several orders of magnitude, where you want every decade (or log-interval) to have equal weight.</p>
<p>For <span class="math inline">x \in [a, b]</span>:</p>
<ul>
<li><p><strong>PDF</strong> <span class="math display">
f(x) = \frac{1}{x , [\ln(b) - \ln(a)]}, \quad a \le x \le b.
</span> Outside <span class="math inline">[a, b]</span>, <span class="math inline">f(x) = 0</span>.</p></li>
<li><p><strong>CDF</strong> <span class="math display">
F(x) = \Pr(X \le x) = \frac{\ln(x) - \ln(a)}{\ln(b) - \ln(a)}, \quad a \le x \le b.
</span> (For <span class="math inline">x &lt; a</span>, <span class="math inline">F(x) = 0</span>; for <span class="math inline">x &gt; b</span>, <span class="math inline">F(x) = 1</span>.</p></li>
</ul>
<section id="args-48" class="level4">
<h4 class="anchored" data-anchor-id="args-48">Args:</h4>
<div class="sourceCode" id="cb93"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>bi.dist.log_uniform(</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>low,</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>high,</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb93-10"><a href="#cb93-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb93-11"><a href="#cb93-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb93-12"><a href="#cb93-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb93-13"><a href="#cb93-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb93-14"><a href="#cb93-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>low</em> (jnp.ndarray): The lower bound of the uniform distribution’s log-space. Must be positive.</p></li>
<li><p><em>high</em> (jnp.ndarray): The upper bound of the uniform distribution’s log-space. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-47" class="level4">
<h4 class="anchored" data-anchor-id="returns-47">Returns:</h4>
<p>BI LogUniform distribution object (for model building) when <code>sample=False</code>.</p>
<p>JAX array of samples drawn from the LogUniform distribution (for direct sampling) when <code>sample=True</code>.</p>
<p>The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-44" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-44">Example Usage:</h4>
<div class="sourceCode" id="cb94"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>m.dist.log_uniform(low<span class="op">=</span><span class="fl">0.1</span>, high<span class="op">=</span><span class="fl">10.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-35" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-35">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#loguniform</p>
</section>
<section id="references-23" class="level4">
<h4 class="anchored" data-anchor-id="references-23">References:</h4>
<p>https://en.wikipedia.org/wiki/Reciprocal_distribution https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.loguniform.html https://docs.scipy.org/doc/scipy/tutorial/stats/continuous_loguniform.html</p>
<hr>
</section>
</section>
<section id="logistic" class="level3">
<h3 class="anchored" data-anchor-id="logistic">Logistic</h3>
<p>Samples from a Logistic distribution.</p>
<p>The Logistic distribution is a continuous probability distribution defined by two parameters: location and scale. It is often used to model growth processes and is closely related to the normal distribution.Its CDF is the logistic (sigmoid) function, which makes it appealing in modeling probabilities, logistic regression, and various growth models. It resembles the normal distribution in shape (bell‐shaped, symmetric) but has <strong>heavier tails</strong> (i.e.&nbsp;more probability in the extremes) and simpler closed‐form expressions for the CDF.</p>
<ul>
<li><strong>Location parameter</strong>: <span class="math inline">\mu \in {R}</span> — the center or “mean” of the distribution.</li>
<li><strong>Scale parameter</strong>: <span class="math inline">s &gt; 0</span> — controls the spread (similar role to standard deviation).</li>
<li><strong>Support</strong>: $&nbsp;x (-, +) $.</li>
</ul>
<p>Let <span class="math inline">X \sim {Logistic}(\mu, s)</span>. Then:</p>
<p><span class="math display">
F(x; \mu, s) = \frac{1}{1 + \exp\bigl(-\frac{x - \mu}{s}\bigr)}.
</span></p>
<p>Equivalently: <span class="math display">
F(x) = \frac12 + \frac12 \tanh!\bigl(\frac{x - \mu}{2s}\bigr).
</span></p>
<section id="args-49" class="level4">
<h4 class="anchored" data-anchor-id="args-49">Args:</h4>
<div class="sourceCode" id="cb95"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>bi.dist.logistic(</span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb95-11"><a href="#cb95-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb95-12"><a href="#cb95-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb95-13"><a href="#cb95-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb95-14"><a href="#cb95-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray or float): The location parameter, specifying the median of the distribution. Defaults to 0.0.</p></li>
<li><p><em>scale</em> (jnp.ndarray or float): The scale parameter, which determines the spread of the distribution. Must be positive. Defaults to 1.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-48" class="level4">
<h4 class="anchored" data-anchor-id="returns-48">Returns:</h4>
<p>BI Logistic distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the Logistic distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-45" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-45">Example Usage:</h4>
<div class="sourceCode" id="cb96"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>m.dist.logistic(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-36" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-36">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#logistic</p>
</section>
<section id="reference" class="level4">
<h4 class="anchored" data-anchor-id="reference">Reference:</h4>
<p>https://en.wikipedia.org/wiki/Logistic_distribution</p>
<hr>
</section>
</section>
<section id="low-rank-multivariate-normal" class="level3">
<h3 class="anchored" data-anchor-id="low-rank-multivariate-normal">Low Rank Multivariate Normal</h3>
<p>The <em>Low-Rank Multivariate Normal</em> (LRMVN) distribution is a parameterization of the multivariate normal distribution where the covariance matrix is expressed as a low-rank plus diagonal decomposition: <span class="math display">
\Sigma = F F^\top + D
</span></p>
<p>where <span class="math inline">F</span> is a low-rank matrix (capturing correlations) and <span class="math inline">D</span> is a diagonal matrix (capturing independent noise). This representation is often used in probabilistic modeling and variational inference to efficiently handle high-dimensional Gaussian distributions with structured covariance.</p>
<!---
For a (d)-dimensional random vector $ x \in {R}^d $,
$$
p(x \mid \mu, F, D) = \frac{1}{(2\pi)^{d/2} |\Sigma|^{1/2}} \exp\left(-\frac{1}{2}(x - \mu)^\top \Sigma^{-1} (x - \mu)\right),
$$

where $ \Sigma = F F^\top + D $.
-->
<p>Parameters: - <em>loc</em> (jnp.ndarray): Mean vector.</p>
<p>cov_factor (jnp.ndarray): Matrix used to construct the covariance matrix.</p>
<p>cov_diag (jnp.ndarray): Diagonal elements of the covariance matrix.</p>
<ul>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
<section id="example-usage-46" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-46">Example Usage:</h4>
<p>from BI import bi m = bi(‘cpu’) event_size = 100 # Our distribution has 100 dimensions rank = 5<br>
m.dist.low_rank_multivariate_normal( - <em>loc</em>=m.dist.normal(0,1, shape = (event_size,), sample=True)<em>2, cov_factor=m.dist.normal(0,1, shape = (event_size, rank), sample=True), cov_diag=jnp.exp(m.dist.normal(0,1, shape = (event_size,), sample=True)) </em> 0.1, sample=True )</p>
</section>
<section id="wrapper-of-37" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-37">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#lowrankmultivariatenormal</p>
</section>
<section id="reference-1" class="level4">
<h4 class="anchored" data-anchor-id="reference-1">Reference:</h4>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Multivariate_normal_distribution">Multivariate normal distribution – Wikipedia</a></p></li>
<li><p><a href="https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/MultivariateNormalDiagPlusLowRank">TensorFlow Probability: LowRankMultivariateNormal</a></p></li>
</ul>
<div class="sourceCode" id="cb97"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>bi.dist.low_rank_multivariate_normal(</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>loc,</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>cov_factor,</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a>cov_diag,</span>
<span id="cb97-5"><a href="#cb97-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb97-6"><a href="#cb97-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb97-7"><a href="#cb97-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb97-8"><a href="#cb97-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb97-9"><a href="#cb97-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb97-10"><a href="#cb97-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb97-11"><a href="#cb97-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb97-12"><a href="#cb97-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb97-13"><a href="#cb97-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb97-14"><a href="#cb97-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb97-15"><a href="#cb97-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="lower-truncated-power-law" class="level3">
<h3 class="anchored" data-anchor-id="lower-truncated-power-law">Lower Truncated Power Law</h3>
<p>The <em>Lower-Truncated Power-Law</em> distribution (also known as the <em>Pareto Type I</em> or <em>power-law with a lower bound</em>) models quantities that follow a heavy-tailed power-law behavior but are bounded below by a minimum value <span class="math inline">x_{min}</span>. It is commonly used to describe phenomena such as wealth distributions, city sizes, and biological scaling laws</p>
<!--

The probability density function (PDF) is given by:

For $ x \ge x_{min} > 0 $:
$$
f(x \mid \alpha, x_{min}) = \frac{\alpha - 1}{x_{min}} \left( \frac{x_{min}}{x} \right)^{\alpha}
$$

where $ \alpha > 1 $ is the shape (scaling) parameter.
-->
<section id="args-50" class="level4">
<h4 class="anchored" data-anchor-id="args-50">Args:</h4>
<div class="sourceCode" id="cb98"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>bi.dist.lower_truncated_power_law(</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>alpha,</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a>low,</span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><code>alpha</code> (jnp.ndarray): index of the power law distribution. Must be less than -1.</p></li>
<li><p><code>low</code> (jnp.ndarray): lower bound of the distribution. Must be greater than 0.</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><code>event</code> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><code>mask</code> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><code>create_obj</code> (bool): If True, returns the raw BI distribution object instead of creating a sample site.This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><code>sample</code> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><code>seed</code> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><code>obs</code> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><code>name</code> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-49" class="level4">
<h4 class="anchored" data-anchor-id="returns-49">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI LowerTruncatedPowerLaw distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the LowerTruncatedPowerLaw distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-47" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-47">Example Usage:</h4>
<div class="sourceCode" id="cb99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a>m.dist.lower_truncated_power_law(alpha<span class="op">=-</span><span class="fl">2.0</span>, low<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-38" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-38">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#lowertruncatedpowerlaw</p>
</section>
<section id="reference-2" class="level4">
<h4 class="anchored" data-anchor-id="reference-2">Reference:</h4>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Power_law#Continuous_power-law_distribution">Power-law distribution – Wikipedia</a></p></li>
<li><p><a href="https://en.wikipedia.org/wiki/Pareto_distribution">Pareto distribution – Wikipedia</a></p></li>
<li><p><a href="https://arxiv.org/abs/0706.1062">Truncated Pareto distribution</a></p></li>
</ul>
<hr>
</section>
</section>
<section id="matrix-normal" class="level3">
<h3 class="anchored" data-anchor-id="matrix-normal">Matrix Normal</h3>
<p>Samples from a Matrix Normal distribution, which is a multivariate normal distribution over matrices. The distribution is characterized by a location matrix and two lower triangular matrices that define the correlation structure. The distribution is related to the multivariate normal distribution in the following way. If <span class="math inline">X ~ MN(loc,U,V)</span> then <span class="math inline">vec(X) ~ MVN(vec(loc), kron(V,U) )</span>.</p>
<p>Let (X) be (n p), with mean matrix <span class="math display">M\in R^{n\times p}</span>, row-covariance matrix <span class="math inline">U\in R^{n\times n}</span> (positive-definite), and column-covariance matrix <span class="math inline">V\in R^{p\times p}</span>(positive-definite). Then the PDF is:</p>
<p><span class="math display">
f(X;M,U,V) = \frac{1}{(2\pi)^{\tfrac{np}{2}},|V|^{\tfrac n2},|U|^{\tfrac p2}},
\exp!\Big[-\tfrac12,{tr}!\big( V^{-1}(X-M)^\top,U^{-1}(X-M)\big)\Big].
</span></p>
<p>Also the equivalence to vec-form is given: <span class="math display">
{vec}(X) \sim  N_{np}\big({vec}(M),,V\otimes U\big). ; ; \text{([The Book of Statistical Proofs][3])}
</span></p>
<section id="args-51" class="level4">
<h4 class="anchored" data-anchor-id="args-51">Args:</h4>
<div class="sourceCode" id="cb100"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>bi.dist.matrix_normal(</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>loc,</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>scale_tril_row,</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>scale_tril_column,</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb100-13"><a href="#cb100-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb100-14"><a href="#cb100-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb100-15"><a href="#cb100-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><code>loc</code> (array_like): Location of the distribution.</p></li>
<li><p><code>scale_tril_row</code> (array_like): Lower cholesky of rows correlation matrix.</p></li>
<li><p><code>scale_tril_column</code> (array_like): Lower cholesky of columns correlation matrix.</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><code>event</code> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><code>mask</code> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><code>create_obj</code> (bool): If True, returns the raw BI distribution object instead of creating a sample site.This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><code>sample</code> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><code>seed</code> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><code>obs</code> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><code>name</code> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-50" class="level4">
<h4 class="anchored" data-anchor-id="returns-50">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI MatrixNormal distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the MatrixNormal distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-48" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-48">Example Usage:</h4>
<div class="sourceCode" id="cb101"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> jax.numpy <span class="im">as</span> jnp</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)            </span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>n_rows, n_cols <span class="op">=</span> <span class="dv">3</span>, <span class="dv">4</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a><span class="op">-</span> <span class="op">*</span>loc<span class="op">*</span> <span class="op">=</span> jnp.zeros((n_rows, n_cols))</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>U_row_cov <span class="op">=</span> jnp.array([[<span class="fl">1.0</span>, <span class="fl">0.5</span>, <span class="fl">0.2</span>],</span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>[<span class="fl">0.5</span>, <span class="fl">1.0</span>, <span class="fl">0.3</span>],</span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>[<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">1.0</span>]])</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>scale_tril_row <span class="op">=</span> jnp.linalg.cholesky(U_row_cov)</span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a>V_col_cov <span class="op">=</span> jnp.array([[<span class="fl">2.0</span>, <span class="op">-</span><span class="fl">0.8</span>, <span class="fl">0.1</span>, <span class="fl">0.4</span>],</span>
<span id="cb101-13"><a href="#cb101-13" aria-hidden="true" tabindex="-1"></a>[<span class="op">-</span><span class="fl">0.8</span>, <span class="fl">2.0</span>, <span class="fl">0.2</span>, <span class="op">-</span><span class="fl">0.2</span>],</span>
<span id="cb101-14"><a href="#cb101-14" aria-hidden="true" tabindex="-1"></a>[<span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">2.0</span>, <span class="fl">0.0</span>],</span>
<span id="cb101-15"><a href="#cb101-15" aria-hidden="true" tabindex="-1"></a>[<span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.2</span>, <span class="fl">0.0</span>, <span class="fl">2.0</span>]])</span>
<span id="cb101-16"><a href="#cb101-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-17"><a href="#cb101-17" aria-hidden="true" tabindex="-1"></a><span class="co"># The argument passed to the distribution is its Cholesky factor</span></span>
<span id="cb101-18"><a href="#cb101-18" aria-hidden="true" tabindex="-1"></a>scale_tril_column <span class="op">=</span> jnp.linalg.cholesky(V_col_cov)</span>
<span id="cb101-19"><a href="#cb101-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-20"><a href="#cb101-20" aria-hidden="true" tabindex="-1"></a>m.dist.matrix_normal(</span>
<span id="cb101-21"><a href="#cb101-21" aria-hidden="true" tabindex="-1"></a>oc<span class="op">=</span>loc, </span>
<span id="cb101-22"><a href="#cb101-22" aria-hidden="true" tabindex="-1"></a>scale_tril_row<span class="op">=</span>scale_tril_row, </span>
<span id="cb101-23"><a href="#cb101-23" aria-hidden="true" tabindex="-1"></a>scale_tril_column<span class="op">=</span>scale_tril_column, </span>
<span id="cb101-24"><a href="#cb101-24" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">True</span></span>
<span id="cb101-25"><a href="#cb101-25" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmatrixnormal_lowercase" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmatrixnormal_lowercase">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#matrixnormal_lowercase</h4>
</section>
<section id="reference-3" class="level4">
<h4 class="anchored" data-anchor-id="reference-3">Reference:</h4>
<ul>
<li><ul>
<li>https://en.wikipedia.org/wiki/Matrix_normal_distribution “Matrix normal distribution”</li>
</ul></li>
<li><ul>
<li>https://statproofbook.github.io/P/matn-pdf.html “Probability density function of the matrix-normal distribution | The Book of Statistical Proofs”</li>
</ul></li>
</ul>
<hr>
</section>
</section>
<section id="a-marginalized-finite-mixture-of-component-distributions." class="level3">
<h3 class="anchored" data-anchor-id="a-marginalized-finite-mixture-of-component-distributions.">A marginalized finite mixture of component distributions.</h3>
<p>A <strong>marginalised finite mixture of component distributions</strong> refers to a probability model in which you have a finite number <span class="math inline">K</span> of component distributions (e.g., normals, Poissons, etc.), each weighted by a mixing probability from a Categorical distribution. We <strong>marginalise out</strong> the latent assignment variable (i.e., you don’t explicitly model which component each data point came from). The resulting distribution can be either a <code>mixture_general</code> (when component distributions are a list) or a <code>mixture_same_Family</code> (when component distributions are a single distribution).</p>
<p>Suppose you have:</p>
<ul>
<li>Mixing weights <span class="math inline">\pi_1, \dots, \pi_K</span>, with <span class="math inline">\pi_j \ge 0</span> and <span class="math inline">\sum_{j=1}^K \pi_j = 1</span>.</li>
<li>Component distributions (densities or PMFs) <span class="math inline">f_j(x \mid \theta_j)</span>, for <span class="math inline">j = 1,\dots,K</span>.</li>
</ul>
<p>Then the marginalised mixture distribution for an observed <span class="math inline">X</span> is:</p>
<p><span class="math display">
p(X = x) = \sum_{j=1}^K \pi_j ; f_j(x \mid \theta_j).
</span></p>
<p>This is exactly the mixture density/PMF you get when you marginalise the latent component‑assignment variable <span class="math inline">Z \in {1,\dots,K}</span> out of the joint distribution:</p>
<p><span class="math display">
p(x) = \sum_{z=1}^K p(z = j); p(x \mid z = j) = \sum_{j=1}^K \pi_j ; f_j(x \mid \theta_j).
</span></p>
<section id="args-52" class="level4">
<h4 class="anchored" data-anchor-id="args-52">Args:</h4>
<div class="sourceCode" id="cb102"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>bi.dist.mixture(</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>mixing_distribution,</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a>component_distributions,</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb102-5"><a href="#cb102-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb102-6"><a href="#cb102-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb102-7"><a href="#cb102-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb102-8"><a href="#cb102-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb102-9"><a href="#cb102-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb102-10"><a href="#cb102-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb102-11"><a href="#cb102-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb102-12"><a href="#cb102-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb102-13"><a href="#cb102-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb102-14"><a href="#cb102-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>mixing_distribution</em> (BI.distribution): The distribution used to determine the mixing weights.</p></li>
<li><p><em>component_distributions</em> (list[BI.distribution]): The list of component distributions.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-51" class="level4">
<h4 class="anchored" data-anchor-id="returns-51">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Mixture distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Mixture distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-49" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-49">Example Usage:</h4>
<div class="sourceCode" id="cb103"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> jax <span class="im">import</span> random</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> BI <span class="im">as</span> pyro</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> pyro.distributions.Mixture(</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a>pyro.distributions.Categorical(torch.ones(<span class="dv">2</span>)),</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>[pyro.distributions.Normal(<span class="dv">0</span>, <span class="dv">1</span>), pyro.distributions.Normal(<span class="dv">2</span>, <span class="dv">1</span>)]</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> m.sample(sample_shape<span class="op">=</span>(<span class="dv">10</span>,))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixture" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixture">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#mixture</h4>
</section>
<section id="references-24" class="level4">
<h4 class="anchored" data-anchor-id="references-24">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Mixture_distribution “Mixture distribution”</li>
<li>https://pymc3-testing.readthedocs.io/en/rtd-docs/notebooks/marginalized_gaussian_mixture_model.html “Marginalized Gaussian Mixture Model — PyMC3 3.1rc3 documentation”</li>
</ul>
<hr>
</section>
</section>
<section id="mixture-general" class="level3">
<h3 class="anchored" data-anchor-id="mixture-general">Mixture General</h3>
<p>A mixture distribution is a probability distribution constructed by selecting one of several component distributions according to specified weights, and then drawing a sample from the chosen component. It allows modelling of heterogeneous populations and multimodal data.</p>
<p>Let the mixture have (K) component distributions with densities <span class="math inline">f_i(x)</span> for <span class="math inline">i = 1,\dots,K</span>. Let weights $w_i) satisfy <span class="math inline">\sum_{i=1}^K w_i = 1</span>. Then the PDF is <span class="math display">
f(x) = \sum_{i=1}^K w_i ; f_i(x).
</span></p>
<p><strong>Parameters:</strong></p>
<ul>
<li><p><strong>mixing_distribution</strong>: A <code>Categorical</code> distribution representing the mixing weights.</p></li>
<li><p><strong>component_distributions</strong>: A list of distributions representing the components of the mixture.</p></li>
<li><p>**sample (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><strong>seed (int, optional)</strong>: An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><strong>obs (jnp.ndarray, optional)</strong>: The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><strong>name (str, optional)</strong>: The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
<p><strong>#### Returns:</strong></p>
<ul>
<li><strong>When <code>sample=False</code></strong>: A BI MixtureGeneral distribution object (for model building).</li>
<li><strong>When <code>sample=True</code></strong>: A JAX array of samples drawn from the MixtureGeneral distribution (for direct sampling).</li>
<li><strong>When <code>create_obj=True</code></strong>: The raw BI distribution object (for advanced use cases).</li>
</ul>
<p><strong>#### Example Usage:</strong></p>
<p>from BI import bi m = bi(‘cpu’) m.dist.mixture_general( mixing_distribution=m.dist.categorical(probs=jnp.array([0.3, 0.7]), create_obj = True),<br>
component_distributions=[m.dist.normal(loc=0.0, scale=1.0, create_obj=True),m.dist.normal (loc=0.0, scale=1.0, create_obj=True)], sample = True )</p>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixturegeneral" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixturegeneral">Wrapper of: <a href="https://num.pyro.ai/en/stable/distributions.html#mixturegeneral">https://num.pyro.ai/en/stable/distributions.html#mixturegeneral</a></h4>
</section>
<section id="reference-4" class="level4">
<h4 class="anchored" data-anchor-id="reference-4">Reference:<br>
</h4>
<ul>
<li><ul>
<li>https://preliz.readthedocs.io/en/latest/distributions/gallery/mixture.html “Mixture Distribution — PreliZ 0.22.0 documentation”</li>
</ul></li>
<li><ul>
<li>https://reliability.readthedocs.io/en/latest/Mixture%20models.html “Mixture models — reliability 0.9.0 documentation”</li>
</ul></li>
<li><ul>
<li>https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.Mixture.html “Mixture — SciPy v1.16.2 Manual”</li>
</ul></li>
</ul>
<div class="sourceCode" id="cb104"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>bi.dist.mixture_general(</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>mixing_distribution,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>component_distributions,</span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>support<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb104-6"><a href="#cb104-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb104-7"><a href="#cb104-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb104-8"><a href="#cb104-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb104-9"><a href="#cb104-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb104-10"><a href="#cb104-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb104-11"><a href="#cb104-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb104-12"><a href="#cb104-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb104-13"><a href="#cb104-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb104-14"><a href="#cb104-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb104-15"><a href="#cb104-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="finite-mixture-of-component-distributions-from-the-same-family." class="level3">
<h3 class="anchored" data-anchor-id="finite-mixture-of-component-distributions-from-the-same-family.">Finite mixture of component distributions from the same family.</h3>
<p>A <em>Mixture (Same-Family)</em> distribution is a finite mixture in which <strong>all components come from the same parametric family</strong> (for example, all Normal distributions but with different parameters), and are combined via mixing weights. The class is typically denoted as:</p>
<p>Let the mixture have (K) components. Let weights (w_i), (_{i=1}^K w_i = 1). Let the component family have density (f(x _i)) for each component (i). Then the mixture’s PDF is <span class="math display">
f_X(x) = \sum_{i=1}^K w_i ; f(x \mid \theta_i).
</span></p>
<p>where each <span class="math inline">f(x \mid \theta_i)</span> is from the same family with parameter <span class="math inline">\theta_i</span>.</p>
<section id="args-53" class="level4">
<h4 class="anchored" data-anchor-id="args-53">Args:</h4>
<div class="sourceCode" id="cb105"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>bi.dist.mixture_same_family(</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>mixing_distribution,</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>component_distribution,</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb105-10"><a href="#cb105-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb105-11"><a href="#cb105-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb105-12"><a href="#cb105-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb105-13"><a href="#cb105-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb105-14"><a href="#cb105-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><strong>Distribution Args</strong>:</p></li>
<li><p><strong>mixing_distribution</strong>: A <code>Categorical</code> distribution representing the mixing weights.</p></li>
<li><p><strong>component_distributions</strong>: A list of distributions representing the components of the mixture.</p></li>
<li><p><strong>Sampling / Modeling Args</strong>:</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><code>event</code> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><code>mask</code> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><code>create_obj</code> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p>**sample (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><strong>seed (int, optional)</strong>: An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><strong>obs (jnp.ndarray, optional)</strong>: The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><strong>name (str, optional)</strong>: The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-52" class="level4">
<h4 class="anchored" data-anchor-id="returns-52">Returns:</h4>
<ul>
<li><strong>When <code>sample=False</code></strong>: A BI MixtureSameFamily distribution object (for model building).</li>
<li><strong>When <code>sample=True</code></strong>: A JAX array of samples drawn from the MixtureSameFamily distribution (for direct sampling).</li>
<li><strong>When <code>create_obj=True</code></strong>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-50" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-50">Example Usage:</h4>
<div class="sourceCode" id="cb106"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>m.dist.mixture_same_family(</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>mixing_distribution<span class="op">=</span>m.dist.categorical(probs<span class="op">=</span>jnp.array([<span class="fl">0.3</span>, <span class="fl">0.7</span>]), create_obj <span class="op">=</span> <span class="va">True</span>), </span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a>component_distribution<span class="op">=</span>m.dist.normal(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, shape <span class="op">=</span> (<span class="dv">2</span>,), create_obj<span class="op">=</span><span class="va">True</span>),</span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>sample <span class="op">=</span> <span class="va">True</span></span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixture--to_jax-truesame-family" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmixture--to_jax-truesame-family">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#mixture-, to_jax = True)same-family</h4>
</section>
<section id="reference-5" class="level4">
<h4 class="anchored" data-anchor-id="reference-5">Reference:</h4>
<ul>
<li><a href="https://cseweb.ucsd.edu/~elkan/250Bwinter2011/mixturemodels.pdf">Mixture distributions</a></li>
</ul>
<hr>
</section>
</section>
<section id="multinomial" class="level3">
<h3 class="anchored" data-anchor-id="multinomial">Multinomial</h3>
<p>The multinomial distribution is a discrete probability distribution that generalizes the Binomial distribution to the case of <strong>more than two possible outcomes per trial</strong>. Specifically:</p>
<ul>
<li>You perform a fixed number <span class="math inline">n</span> of independent trials.</li>
<li>On each trial there are <span class="math inline">k</span> possible outcomes, labeled (1,2,,k).</li>
<li>The probability of outcome <span class="math inline">i</span> on any trial is <span class="math inline">p_i</span>, where <span class="math inline">p_1 + p_2 + \cdots + p_k = 1</span>.</li>
<li>Let <span class="math inline">X_i</span> be the number of trials (out of <span class="math inline">n</span>) that produced outcome <span class="math inline">i</span>. Then the vector $ X = (X_1, , X_k)$ follows a multinomial distribution.</li>
</ul>
<p>It reduces to the binomial when (k=2).</p>
<p>PMF (Probability Mass Function):</p>
<p>For $ x = (x_1,,x_k)$ with each <span class="math inline">x_i \ge 0</span> integer and <span class="math inline">\sum_{i=1}^k x_i = n</span>, <span class="math display">
\Pr(X_1 = x_1, \dots, X_k = x_k)
= \frac{n!}{x_1!,x_2!\cdots x_k!}
; p_1^{,x_1}; p_2^{,x_2} \cdots p_k^{,x_k}.
</span></p>
<p>If <span class="math inline">\sum_i x_i \neq n</span>, the probability is 0.</p>
<section id="args-54" class="level4">
<h4 class="anchored" data-anchor-id="args-54">Args:</h4>
<div class="sourceCode" id="cb107"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a>bi.dist.multinomial(</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>total_count_max<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-10"><a href="#cb107-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb107-11"><a href="#cb107-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb107-12"><a href="#cb107-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb107-13"><a href="#cb107-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb107-14"><a href="#cb107-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb107-15"><a href="#cb107-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb107-16"><a href="#cb107-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>total_count</em> (int or jnp.ndarray): The number of trials.</p></li>
<li><p><em>probs</em> (jnp.ndarray, optional): Event probabilities. Must sum to 1.</p></li>
<li><p><em>logits</em> (jnp.ndarray, optional): Event log probabilities.</p></li>
<li><p>*total_count_max (int, optional): An optional integer providing an upper bound on <code>total_count</code>. This is used for performance optimization with <code>lax.scan</code> when <code>total_count</code> is a dynamic JAX tracer, helping to avoid recompilation.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-53" class="level4">
<h4 class="anchored" data-anchor-id="returns-53">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Multinomial distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Multinomial distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-51" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-51">Example Usage:</h4>
<div class="sourceCode" id="cb108"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>m.dist.multinomial(total_count<span class="op">=</span><span class="dv">10</span>, probs<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmultinomial" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmultinomial">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#multinomial</h4>
</section>
<section id="references-25" class="level4">
<h4 class="anchored" data-anchor-id="references-25">References:</h4>
<ul>
<li>https://online.stat.psu.edu/stat504/book/export/html/667 “2.3 - The Multinomial Distribution - STAT ONLINE”</li>
<li>https://en.wikipedia.org/wiki/Multinomial_distribution “Multinomial distribution”</li>
<li>https://faculty.washington.edu/yenchic/20A_stat512/Lec7_Multinomial.pdf “[PDF] Lecture 7: Multinomial distribution”</li>
</ul>
<hr>
</section>
</section>
<section id="multinomial-logits" class="level3">
<h3 class="anchored" data-anchor-id="multinomial-logits">Multinomial Logits</h3>
<p>A <em>multinomial logits</em> distribution refers to a categorical (or more generally multinomial) distribution over (K) classes whose probabilities are given via the softmax of a vector of logits. That is, given a vector of real-valued logits <span class="math inline">\ell = (\ell_1, …, \ell_K)</span>, the class probabilities are: <span class="math display">
p_k = \frac{\exp(\ell_k)}{\sum_{j=1}^K \exp(\ell_j)}.
</span> Then a single draw from the distribution yields one of the (K) classes (or for a multinomial count version, counts over the classes) with those probabilities.</p>
<p><strong>Categorical version (one draw):</strong> Let <span class="math inline">X \in {1,2,\dots,K}</span> be the class. Then <span class="math display">
\Pr(X=k \mid \ell) ;=; p_k = \frac{\exp(\ell_k)}{\sum_{j=1}^K \exp(\ell_j)}.
</span> This directly uses the logits <span class="math inline">\ell</span>.</p>
<p><strong>Multinomial version (n draws):</strong> If you draw (n) independent draws (or equivalently count vector) with probabilities <span class="math inline">p = (p_1,…,p_K)</span>, then for a count vector <span class="math inline">x = (x_1,…,x_K)</span> with <span class="math inline">\sum_k x_k = n</span>, <span class="math display">
\Pr(X = x \mid n, p) = \frac{n!}{x_1! \cdots x_K!} ; \prod_{k=1}^K p_k^{x_k}
</span></p>
<p>and <span class="math inline">p_k</span> is given by the softmax of logits.</p>
<section id="args-55" class="level4">
<h4 class="anchored" data-anchor-id="args-55">Args:</h4>
<div class="sourceCode" id="cb109"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>bi.dist.multinomial_logits(</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>total_count_max<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb109-8"><a href="#cb109-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb109-9"><a href="#cb109-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb109-10"><a href="#cb109-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb109-11"><a href="#cb109-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb109-12"><a href="#cb109-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb109-13"><a href="#cb109-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb109-14"><a href="#cb109-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb109-15"><a href="#cb109-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>logits</em> (jnp.ndarray): Logits for each outcome. Must be at least one-dimensional.</p></li>
<li><p><em>total_count</em> (jnp.ndarray): The total number of trials.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-54" class="level4">
<h4 class="anchored" data-anchor-id="returns-54">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI MultinomialLogits distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the MultinomialLogits distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-52" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-52">Example Usage:</h4>
<div class="sourceCode" id="cb110"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>m.dist.multinomial_logits(logits<span class="op">=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">0.5</span>], dtype<span class="op">=</span>jnp.float32), total_count<span class="op">=</span>jnp.array(<span class="dv">5</span>, dtype<span class="op">=</span>jnp.int32), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-39" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-39">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#multinomiallogits</p>
</section>
<section id="reference-6" class="level4">
<h4 class="anchored" data-anchor-id="reference-6">Reference:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Multinomial_distribution</li>
</ul>
<hr>
</section>
</section>
<section id="multinomial-probs" class="level3">
<h3 class="anchored" data-anchor-id="multinomial-probs">Multinomial Probs</h3>
<p>Samples from a Multinomial distribution.</p>
<p>The Multinomial distribution models the number of times each of several discrete outcomes occurs in a fixed number of trials. Each trial independently results in one of several outcomes, and each outcome has a probability of occurring.</p>
<section id="args-56" class="level4">
<h4 class="anchored" data-anchor-id="args-56">Args:</h4>
<div class="sourceCode" id="cb111"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>bi.dist.multinomial_probs(</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>total_count<span class="op">=</span><span class="dv">1</span>,</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a>total_count_max<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb111-9"><a href="#cb111-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb111-10"><a href="#cb111-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb111-11"><a href="#cb111-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb111-12"><a href="#cb111-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb111-13"><a href="#cb111-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb111-14"><a href="#cb111-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb111-15"><a href="#cb111-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>probs</em> (jnp.ndarray): Vector of probabilities for each outcome. Must sum to 1.</p></li>
<li><p><em>total_count</em> (jnp.ndarray): The number of trials.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-55" class="level4">
<h4 class="anchored" data-anchor-id="returns-55">Returns:</h4>
<p>BI MultinomialProbs distribution object (for model building). JAX array of samples drawn from the MultinomialProbs distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-53" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-53">Example Usage:</h4>
<div class="sourceCode" id="cb112"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>m.dist.multinomial_probs(probs<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), total_count<span class="op">=</span><span class="dv">10</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-40" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-40">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#multinomialprobs</p>
<hr>
</section>
</section>
<section id="multivariate-normal" class="level3">
<h3 class="anchored" data-anchor-id="multivariate-normal">Multivariate Normal</h3>
<p>The Multivariate Normal distribution, also known as the Gaussian distribution in multiple dimensions, is a probability distribution that arises frequently in statistics and machine learning. It is defined by its mean vector and covariance matrix, which describe the central tendency and spread of the distribution, respectively.</p>
<p><span class="math display">
p(x) = \frac{1}{\sqrt{(2\pi)^n |\Sigma|}} \exp\left(-\frac{1}{2}(x - \mu)^T \Sigma^{-1} (x - \mu)\right)
</span></p>
<p>where: - <span class="math inline">x</span> is a <span class="math inline">n</span>-dimensional vector of random variables. - <span class="math inline">\mu</span> is the mean vector. - <span class="math inline">\Sigma</span> is the covariance matrix.</p>
<section id="args-57" class="level4">
<h4 class="anchored" data-anchor-id="args-57">Args:</h4>
<div class="sourceCode" id="cb113"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>bi.dist.multivariate_normal(</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a>covariance_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a>precision_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-6"><a href="#cb113-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-7"><a href="#cb113-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb113-8"><a href="#cb113-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-9"><a href="#cb113-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-10"><a href="#cb113-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb113-11"><a href="#cb113-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb113-12"><a href="#cb113-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb113-13"><a href="#cb113-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb113-14"><a href="#cb113-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb113-15"><a href="#cb113-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb113-16"><a href="#cb113-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (tuple): The mean vector of the distribution.</p></li>
<li><p><em>covariance_matrix</em> (jnp.ndarray, optional): The covariance matrix of the distribution. Must be positive definite.</p></li>
<li><p><em>precision_matrix</em> (jnp.ndarray, optional): The precision matrix (inverse of the covariance matrix) of the distribution. Must be positive definite.</p></li>
<li><p><em>scale_tril</em> (jnp.ndarray, optional): The lower triangular Cholesky decomposition of the covariance matrix.</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-56" class="level4">
<h4 class="anchored" data-anchor-id="returns-56">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI MultivariateNormal distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the MultivariateNormal distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-54" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-54">Example Usage:</h4>
<div class="sourceCode" id="cb114"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a>m.dist.multivariate_normal(</span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="op">-</span> <span class="op">*</span>loc<span class="op">*=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">0.0</span>, <span class="op">-</span><span class="fl">2.0</span>]), </span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>covariance_matrix<span class="op">=</span>jnp.array([[ <span class="fl">2.0</span>,  <span class="fl">0.7</span>, <span class="op">-</span><span class="fl">0.3</span>],</span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>[ <span class="fl">0.7</span>,  <span class="fl">1.0</span>,  <span class="fl">0.5</span>],</span>
<span id="cb114-7"><a href="#cb114-7" aria-hidden="true" tabindex="-1"></a>[<span class="op">-</span><span class="fl">0.3</span>,  <span class="fl">0.5</span>,  <span class="fl">1.5</span>]]), </span>
<span id="cb114-8"><a href="#cb114-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">True</span></span>
<span id="cb114-9"><a href="#cb114-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmultivariate-normal" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlmultivariate-normal">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#multivariate-normal</h4>
</section>
<section id="reference-7" class="level4">
<h4 class="anchored" data-anchor-id="reference-7">Reference:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Multivariate_normal_distribution">Multivariate normal distribution – Wikipedia</a></li>
</ul>
<hr>
</section>
</section>
<section id="multivariate-students-t" class="level3">
<h3 class="anchored" data-anchor-id="multivariate-students-t">Multivariate Student’s t</h3>
<p>The Multivariate Student’s t distribution is a generalization of the Student’s t distribution to multiple dimensions. It is a heavy-tailed distribution that is often used to model data that is not normally distributed.</p>
<p>The PDF of the multivariate Student’s t-distribution for a random vector ( {x} ^d ) is given by:</p>
<p><span class="math display">
f({x}) = \frac{\Gamma\left(\frac{\nu + d}{2}\right)}{\Gamma\left(\frac{\nu}{2}\right) \nu^{d/2} \pi^{d/2} | \Sigma|^{1/2}} \left(1 + \frac{1}{\nu} ({x} - \mu)^T \Sigma^{-1} ({x} - \mu)\right)^{-(\nu + d)/2}
</span></p>
<p>where:</p>
<ul>
<li>$ () $ is the Gamma function.</li>
<li>$ $ is the mean vector.</li>
<li>$ $ is the scale (covariance) matrix.</li>
<li>$ $ is the degrees of freedom.</li>
<li>$ d $ is the dimensionality of $ {x} $.</li>
</ul>
<section id="args-58" class="level4">
<h4 class="anchored" data-anchor-id="args-58">Args:</h4>
<div class="sourceCode" id="cb115"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a>bi.dist.multivariate_student_t(</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>df,</span>
<span id="cb115-3"><a href="#cb115-3" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb115-4"><a href="#cb115-4" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb115-5"><a href="#cb115-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb115-6"><a href="#cb115-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb115-7"><a href="#cb115-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb115-8"><a href="#cb115-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb115-9"><a href="#cb115-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb115-10"><a href="#cb115-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb115-11"><a href="#cb115-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb115-12"><a href="#cb115-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb115-13"><a href="#cb115-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb115-14"><a href="#cb115-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb115-15"><a href="#cb115-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>df</em> (jnp.ndarray): Degrees of freedom, must be positive.</p></li>
<li><p><em>loc</em> (jnp.ndarray): Location vector, representing the mean of the distribution. scale_tril (jnp.ndarray): Lower triangular matrix defining the scale.</p></li>
<li><p><code>shape</code> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-57" class="level4">
<h4 class="anchored" data-anchor-id="returns-57">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI MultivariateStudentT distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the MultivariateStudentT distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-55" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-55">Example Usage:</h4>
<div class="sourceCode" id="cb116"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> jax.numpy <span class="im">as</span> jnp</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb116-4"><a href="#cb116-4" aria-hidden="true" tabindex="-1"></a>m.dist.multivariate_student_t(</span>
<span id="cb116-5"><a href="#cb116-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="dv">2</span>,</span>
<span id="cb116-6"><a href="#cb116-6" aria-hidden="true" tabindex="-1"></a><span class="op">-</span> <span class="op">*</span>loc<span class="op">*=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">0.0</span>, <span class="op">-</span><span class="fl">2.0</span>]), </span>
<span id="cb116-7"><a href="#cb116-7" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span>jnp.linalg.cholesky(</span>
<span id="cb116-8"><a href="#cb116-8" aria-hidden="true" tabindex="-1"></a>jnp.array([[ <span class="fl">2.0</span>,  <span class="fl">0.7</span>, <span class="op">-</span><span class="fl">0.3</span>],</span>
<span id="cb116-9"><a href="#cb116-9" aria-hidden="true" tabindex="-1"></a>[ <span class="fl">0.7</span>,  <span class="fl">1.0</span>,  <span class="fl">0.5</span>],</span>
<span id="cb116-10"><a href="#cb116-10" aria-hidden="true" tabindex="-1"></a>[<span class="op">-</span><span class="fl">0.3</span>,  <span class="fl">0.5</span>,  <span class="fl">1.5</span>]])), </span>
<span id="cb116-11"><a href="#cb116-11" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">True</span></span>
<span id="cb116-12"><a href="#cb116-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-41" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-41">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#multivariatestudentt</p>
</section>
<section id="reference-8" class="level4">
<h4 class="anchored" data-anchor-id="reference-8">Reference:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Multivariate_Student%27s_t-distribution">Multivariate Student-t distribution – Wikipedia</a></li>
</ul>
<hr>
</section>
</section>
<section id="negative-binomial" class="level3">
<h3 class="anchored" data-anchor-id="negative-binomial">Negative Binomial</h3>
<p>The Negative Binomial distribution models the number of failures (or the total number of trials) in a sequence of independent Bernoulli trials with success probability (p), until a specified number <code>total_count</code> (r) of successes is achieved. It is often used as a count-data model when the variance exceeds the mean (“overdispersion”) relative to a Poisson.</p>
<p>PMF (Probability Mass Function): Given (X) = number of failures before the (r)th success (so (X = 0,1,2,)), with (r) successes fixed, success probability (p), failure probability (q = 1 - p):</p>
<p><span class="math display">
\Pr(X = k) ;=; \binom{k + r - 1}{k} ; p^r ; q^k, \quad k = 0,1,2,\dots
</span></p>
<section id="args-59" class="level4">
<h4 class="anchored" data-anchor-id="args-59">Args:</h4>
<div class="sourceCode" id="cb117"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>bi.dist.negative_binomial(</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a>total_count,</span>
<span id="cb117-3"><a href="#cb117-3" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-4"><a href="#cb117-4" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-5"><a href="#cb117-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-6"><a href="#cb117-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb117-7"><a href="#cb117-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-8"><a href="#cb117-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-9"><a href="#cb117-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb117-10"><a href="#cb117-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb117-11"><a href="#cb117-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb117-12"><a href="#cb117-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb117-13"><a href="#cb117-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb117-14"><a href="#cb117-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb117-15"><a href="#cb117-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>total_count</em> (jnp.ndarray): The total number of events.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-58" class="level4">
<h4 class="anchored" data-anchor-id="returns-58">Returns:</h4>
<p>BI NegativeBinomial distribution object (for model building). JAX array of samples drawn from the NegativeBinomial distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-56" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-56">Example Usage:</h4>
<div class="sourceCode" id="cb118"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb118-2"><a href="#cb118-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb118-3"><a href="#cb118-3" aria-hidden="true" tabindex="-1"></a>m.dist.negative_binomial(total_count<span class="op">=</span><span class="fl">5.0</span>,probs <span class="op">=</span> jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomial" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomial">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#negativebinomial</h4>
</section>
<section id="references-26" class="level4">
<h4 class="anchored" data-anchor-id="references-26">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Negative_binomial_distribution</li>
<li>https://www.statology.org/negative-binomial-distribution/</li>
</ul>
<hr>
</section>
</section>
<section id="section" class="level3">
<h3 class="anchored" data-anchor-id="section"></h3>
<p>Samples from a Negative Binomial</p>
<p>The NB2 parameterisation of the negative-binomial distribution is a count distribution used for modelling over-dispersed count data (variance &gt; mean). It is defined such that the variance grows <strong>quadratically</strong> in the mean:</p>
<p><span class="math display">
mathrm{Var}(Y) = \mu + \alpha,\mu^2,
</span> where (= {E}[Y]) and (&gt;0) is the dispersion (heterogeneity) parameter. Because of this quadratic variance growth, it is called the NB2 family.</p>
<p><span class="math display">
P(k) = \frac{\Gamma(k + \alpha)}{\Gamma(k + 1) \Gamma(\alpha)} \left(\frac{\beta}{\alpha + \beta}\right)^k \left(1 - \frac{\beta}{\alpha + \beta}\right)^k
</span></p>
<p>One commonly used form of the NB2 parameterisation is obtained via a Poisson‐Gamma mixture: <span class="math display">
Y \mid \lambda \sim \text{Poisson}(\lambda), \quad \lambda \sim \text{Gamma}!\Big(\frac{1}{\alpha},,\frac{\mu,\alpha}{1}\Big)
</span></p>
<p>which marginalises to:</p>
<p><span class="math display">
\Pr(Y = y) = \frac{\Gamma(y + \tfrac1\alpha)}{\Gamma(\tfrac1\alpha); y!}; \bigg(\frac{1}{1 + \alpha,\mu}\bigg)^{1/\alpha}; \bigg(\frac{\alpha,\mu}{1 + \alpha,\mu}\bigg)^{y}, \quad y = 0,1,2,\dots
</span> This parameterisation matches the NB2 form with mean <span class="math inline">\mu</span> and dispersion <span class="math inline">\alpha</span>. (See Hilbe (2011) forderivation.)</p>
<section id="args-60" class="level4">
<h4 class="anchored" data-anchor-id="args-60">Args:</h4>
<div class="sourceCode" id="cb119"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>bi.dist.negative_binomial2(</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>mean,</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb119-6"><a href="#cb119-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb119-7"><a href="#cb119-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb119-8"><a href="#cb119-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb119-9"><a href="#cb119-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb119-10"><a href="#cb119-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb119-11"><a href="#cb119-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb119-12"><a href="#cb119-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb119-13"><a href="#cb119-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb119-14"><a href="#cb119-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>mean</em> (jnp.ndarray or float): The mean of the distribution. This is equivalent to the <code>mu</code> parameter.</p></li>
<li><p><em>concentration</em> (jnp.ndarray or float): The concentration parameter. This is equivalent to the <span class="math inline">alpha</span> parameter.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-59" class="level4">
<h4 class="anchored" data-anchor-id="returns-59">Returns:</h4>
<p>BI NegativeBinomial distribution object: When <code>sample=False</code> (for model building). jnp.ndarray: A JAX array of samples drawn from the NegativeBinomial distribution (for direct sampling). BI NegativeBinomial distribution object: When <code>create_obj=True</code> (for advanced use cases).</p>
</section>
<section id="example-usage-57" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-57">Example Usage:</h4>
<div class="sourceCode" id="cb120"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>m.dist.negative_binomial2(mean<span class="op">=</span><span class="fl">2.0</span>, concentration<span class="op">=</span><span class="fl">3.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomial-1" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomial-1">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#negativebinomial</h4>
</section>
<section id="reference-9" class="level4">
<h4 class="anchored" data-anchor-id="reference-9">Reference:</h4>
<ul>
<li>[Negative Binomial Regression]9https://www.cambridge.org/core/books/negative-binomial-regression/12D6281A46B9A980DC6021080C9419E7)</li>
<li><a href="https://assets.cambridge.org/97805211/98158/excerpt/9780521198158_excerpt.pdf">Negative Binomial Regression: Second Edition</a></li>
</ul>
<hr>
</section>
</section>
<section id="negative-binomial-logits" class="level3">
<h3 class="anchored" data-anchor-id="negative-binomial-logits">Negative Binomial Logits</h3>
<p>Samples from a Negative Binomial Logits distribution.</p>
<p>The Negative Binomial Logits distribution is a generalization of the Negative Binomial distribution where the parameter ‘r’ (number of successes) is expressed as a function of a logit parameter. This allows for more flexible modeling of count data.</p>
<section id="args-61" class="level4">
<h4 class="anchored" data-anchor-id="args-61">Args:</h4>
<div class="sourceCode" id="cb121"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a>bi.dist.negative_binomial_logits(</span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>total_count,</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb121-11"><a href="#cb121-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb121-12"><a href="#cb121-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb121-13"><a href="#cb121-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb121-14"><a href="#cb121-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>total_count</em> (jnp.ndarray): The parameter controlling the shape of the distribution. Represents the total number of trials.</p></li>
<li><p><em>logits</em> (jnp.ndarray): The log-odds parameter. Related to the probability of success.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-60" class="level4">
<h4 class="anchored" data-anchor-id="returns-60">Returns:</h4>
<p>Negative Binomial Logits: A BI Negative Binomial Logits distribution object (for model building).</p>
<p>jnp.ndarray: A JAX array of samples drawn from the Negative Binomial Logits distribution (for direct sampling).</p>
<p>Negative Binomial Logits: The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-58" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-58">Example Usage:</h4>
<div class="sourceCode" id="cb122"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a>m.dist.negative_binomial_logits(total_count<span class="op">=</span><span class="fl">5.0</span>, logits<span class="op">=</span><span class="fl">0.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-42" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-42">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#Negative Binomial Logits</p>
<hr>
</section>
</section>
<section id="negative-binomial-with-probabilities." class="level3">
<h3 class="anchored" data-anchor-id="negative-binomial-with-probabilities.">Negative Binomial with probabilities.</h3>
<p>The Negative Binomial distribution models the number of failures before the first success in a sequence of independent Bernoulli trials. It is characterized by two parameters: ‘concentration’ (r) and ‘rate’ (p). In this implementation, the ‘concentration’ parameter is derived from ‘total_count’ and the ‘rate’ parameter is derived from ‘probs’.</p>
<section id="args-62" class="level4">
<h4 class="anchored" data-anchor-id="args-62">Args:</h4>
<div class="sourceCode" id="cb123"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>bi.dist.negative_binomial_probs(</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a>total_count,</span>
<span id="cb123-3"><a href="#cb123-3" aria-hidden="true" tabindex="-1"></a>probs,</span>
<span id="cb123-4"><a href="#cb123-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb123-5"><a href="#cb123-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb123-6"><a href="#cb123-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb123-7"><a href="#cb123-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb123-8"><a href="#cb123-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb123-9"><a href="#cb123-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb123-10"><a href="#cb123-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb123-11"><a href="#cb123-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb123-12"><a href="#cb123-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb123-13"><a href="#cb123-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb123-14"><a href="#cb123-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>total_count</em> (jnp.ndarray): A numeric vector, matrix, or array representing the parameter.</p></li>
<li><p><em>probs</em> (jnp.ndarray): A numeric vector representing event probabilities. Must sum to 1.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-61" class="level4">
<h4 class="anchored" data-anchor-id="returns-61">Returns:</h4>
<p>BI NegativeBinomialProbs distribution object (for model building). JAX array of samples drawn from the NegativeBinomialProbs distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-59" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-59">Example Usage:</h4>
<div class="sourceCode" id="cb124"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb124-2"><a href="#cb124-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb124-3"><a href="#cb124-3" aria-hidden="true" tabindex="-1"></a>m.dist.negative_binomial_probs(total_count<span class="op">=</span><span class="fl">10.0</span>, probs <span class="op">=</span> jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomialprobs" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlnegativebinomialprobs">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#negativebinomialprobs</h4>
<hr>
</section>
</section>
<section id="normal" class="level3">
<h3 class="anchored" data-anchor-id="normal">Normal</h3>
<p>Samples from a Normal (Gaussian) distribution.</p>
<p>The Normal distribution is characterized by its mean (loc) and standard deviation (scale). It’s a continuous probability distribution that arises frequently in statistics and probability theory.</p>
<p><span class="math display">   
f(x \mid \mu, \sigma) = \frac{1}{\sqrt{2\pi}\sigma}\exp!\left(-\frac{(x - \mu)^2}{2\sigma^2}\right)
</span></p>
<section id="args-63" class="level4">
<h4 class="anchored" data-anchor-id="args-63">Args:</h4>
<div class="sourceCode" id="cb125"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>bi.dist.normal(</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb125-8"><a href="#cb125-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb125-9"><a href="#cb125-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb125-10"><a href="#cb125-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb125-11"><a href="#cb125-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb125-12"><a href="#cb125-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb125-13"><a href="#cb125-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb125-14"><a href="#cb125-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (jnp.ndarray): The mean of the distribution.</p></li>
<li><p><em>scale</em> (jnp.ndarray): The standard deviation of the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-62" class="level4">
<h4 class="anchored" data-anchor-id="returns-62">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Normal distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Normal distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-60" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-60">Example Usage:</h4>
<div class="sourceCode" id="cb126"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb126-3"><a href="#cb126-3" aria-hidden="true" tabindex="-1"></a>m.dist.normal(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-43" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-43">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#normal</p>
</section>
<section id="references-27" class="level4">
<h4 class="anchored" data-anchor-id="references-27">References:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Normal_distribution#Probability_density_function">Wikipedia: Normal distribution – PDF</a></li>
<li><a href="https://mathworld.wolfram.com/NormalDistribution.html">Wolfram MathWorld: Normal Distribution</a></li>
</ul>
<hr>
</section>
</section>
<section id="ordered-logistic" class="level3">
<h3 class="anchored" data-anchor-id="ordered-logistic">Ordered Logistic</h3>
<p>The ordered logistic distribution is used for modeling <strong>ordinal</strong> outcome variables $ Y $ (i.e., categories with a natural order) via a latent continuous predictor $ $ and a set of increasing <em>cut-points</em>. When $ $ crosses thresholds, the observed <span class="math inline">Y</span></p>
<!--
### PMF (Probability Mass Function)

Let $ \eta \in  R $ be the linear predictor (for example $ \eta = x^\top \beta $), let cutpoints $ c = (c_1, \dots,c_{K-1}) $ satisfy $ c_1 < c_2 < \cdots < c_{K-1} $. Define also $ c_0 = -\infty $, $ c_K = +\infty $.
Then for $k \in {1,\dots,K}$,
$$
\Pr(Y = k \mid \eta, c) =
\begin{cases}
1 - {logit}^{-1}(\eta - c_1), k = 1, [6pt]
{logit}^{-1}(\eta - c_{k-1}) - {logit}^{-1}(\eta - c_{k}),  1 < k < K, [6pt]
{logit}^{-1}(\eta - c_{K-1}),  k = K.
\end{cases}
$$

Here $ {logit}^{-1}(z) = \frac{1}{1 + e^{-z}} $.
In other words:
$$
\Pr(Y \le k \mid \eta, c) = {logit}^{-1}(\eta - c_k),
$$
and so
$$
\Pr(Y = k) = \Pr(Y \le k) - \Pr(Y \le k-1).
$$
-->
<p>This formulation appears in e.g.&nbsp;the CDF‐link setup of the proportional odds model.</p>
<section id="args-64" class="level4">
<h4 class="anchored" data-anchor-id="args-64">Args:</h4>
<div class="sourceCode" id="cb127"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>bi.dist.ordered_logistic(</span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a>predictor,</span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a>cutpoints,</span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb127-6"><a href="#cb127-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb127-7"><a href="#cb127-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb127-8"><a href="#cb127-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb127-9"><a href="#cb127-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb127-10"><a href="#cb127-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb127-11"><a href="#cb127-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb127-12"><a href="#cb127-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb127-13"><a href="#cb127-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb127-14"><a href="#cb127-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>predictor</em> (jnp.ndarray): Prediction in real domain; typically this is output of a linear model.</p></li>
<li><p><em>cutpoints</em> (jnp.ndarray): Positions in real domain to separate categories.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-63" class="level4">
<h4 class="anchored" data-anchor-id="returns-63">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI OrderedLogistic distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the OrderedLogistic distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-61" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-61">Example Usage:</h4>
<div class="sourceCode" id="cb128"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>m.dist.ordered_logistic(predictor<span class="op">=</span>jnp.array([<span class="fl">0.2</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span>]), cutpoints<span class="op">=</span>jnp.array([<span class="op">-</span><span class="fl">1.0</span>, <span class="fl">0.0</span>, <span class="fl">1.0</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-44" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-44">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#orderedlogistic</p>
</section>
<section id="reference-10" class="level4">
<h4 class="anchored" data-anchor-id="reference-10">Reference:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Ordered_logit</li>
<li>https://mc-stan.org/docs/2_21/functions-reference/ordered-logistic-distribution.html</li>
<li>https://labdisia.disia.unifi.it/grilli/files/Papers/Ordered_Logit.pdf”</li>
</ul>
<hr>
</section>
</section>
<section id="pareto" class="level3">
<h3 class="anchored" data-anchor-id="pareto">Pareto</h3>
<p>Samples from a Pareto distribution.</p>
<p>The <strong>Pareto distribution</strong>, named after economist Vilfredo Pareto, is a <strong>power-law</strong> probability distribution used to describe phenomena with “rich-get-richer” or “heavy-tail” properties — for example, income distribution, city sizes, or wealth concentration. It is characterized by:</p>
<ul>
<li>a <strong>scale parameter</strong> $ x_m &gt; 0 $ (the minimum possible value), and</li>
<li>a <strong>shape parameter</strong> $ &gt; 0 $ (which controls the tail heaviness).</li>
</ul>
<p>A random variable $ X (, x_m) $ takes values $ x x_m $.</p>
<p><span class="math display">
f(x \mid \alpha, x_m) =
\begin{cases}
\dfrac{\alpha x_m^\alpha}{x^{\alpha + 1}},  x \ge x_m, \
0,  x &lt; x_m.
\end{cases}
</span></p>
<section id="args-65" class="level4">
<h4 class="anchored" data-anchor-id="args-65">Args:</h4>
<div class="sourceCode" id="cb129"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>bi.dist.pareto(</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>scale,</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a>alpha,</span>
<span id="cb129-4"><a href="#cb129-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb129-5"><a href="#cb129-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb129-6"><a href="#cb129-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb129-7"><a href="#cb129-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb129-8"><a href="#cb129-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb129-9"><a href="#cb129-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb129-10"><a href="#cb129-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb129-11"><a href="#cb129-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb129-12"><a href="#cb129-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb129-13"><a href="#cb129-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb129-14"><a href="#cb129-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (jnp.ndarray or float): Scale parameter of the Pareto distribution. Must be positive.</p></li>
<li><p><em>alpha</em> (jnp.ndarray or float): Shape parameter of the Pareto distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-64" class="level4">
<h4 class="anchored" data-anchor-id="returns-64">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI Pareto distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the Pareto distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-62" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-62">Example Usage:</h4>
<div class="sourceCode" id="cb130"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>m.dist.pareto(scale<span class="op">=</span><span class="fl">2.0</span>, alpha<span class="op">=</span><span class="fl">3.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-45" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-45">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#pareto</p>
</section>
<section id="reference-11" class="level4">
<h4 class="anchored" data-anchor-id="reference-11">Reference:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Pareto_distribution#Probability_density_function">Wikipedia: Pareto distribution – PDF</a></li>
<li>Arnold, B. C. (2015). <em>Pareto Distributions</em>, Second Edition. CRC Press.</li>
</ul>
<hr>
</section>
</section>
<section id="poisson" class="level3">
<h3 class="anchored" data-anchor-id="poisson">Poisson</h3>
<p>The <strong>Poisson distribution</strong> models the probability of observing a given number of events <span class="math inline">k</span> occurring in a fixed interval of time or space when these events happen independently and at a constant average rate $ &gt; 0 $. It is widely used for modeling <strong>count data</strong>, such as the number of emails received per hour or mutations in a DNA strand per unit length.</p>
<p>Formally,</p>
<p><span class="math display">
K \sim \text{Poisson}(\lambda)
</span></p>
<p>where <span class="math inline">\lambda</span> is both the <strong>mean</strong> and <strong>variance</strong> of the distribution.</p>
<p>The probability mass function (PMF) of the Poisson distribution is given by: <span class="math display">
P(K = k \mid \lambda) =
\frac{e^{-\lambda} \lambda^k}{k!},
\quad k = 0, 1, 2, \dots
</span></p>
<section id="args-66" class="level4">
<h4 class="anchored" data-anchor-id="args-66">Args:</h4>
<div class="sourceCode" id="cb131"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a>bi.dist.poisson(</span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>rate,</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a>is_sparse<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb131-7"><a href="#cb131-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb131-8"><a href="#cb131-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb131-9"><a href="#cb131-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb131-10"><a href="#cb131-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb131-11"><a href="#cb131-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb131-12"><a href="#cb131-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb131-13"><a href="#cb131-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb131-14"><a href="#cb131-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter, representing the average number of events.</p></li>
<li><p><em>is_sparse</em> (bool, optional): Indicates whether the <code>rate</code> parameter is sparse. If <code>True</code>, a specialized sparse sampling implementation is used, which can be more efficient for models with many zero-rate components (e.g., zero-inflated models). Defaults to <code>False</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-65" class="level4">
<h4 class="anchored" data-anchor-id="returns-65">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Poisson distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Poisson distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-63" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-63">Example Usage:</h4>
<div class="sourceCode" id="cb132"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>m.dist.poisson(rate<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlpoisson" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlpoisson">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#poisson</h4>
</section>
<section id="references-28" class="level4">
<h4 class="anchored" data-anchor-id="references-28">References:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Poisson_distribution#Probability_mass_function">Wikipedia: Poisson distribution – PMF</a></li>
<li>Johnson, N. L., Kotz, S., &amp; Kemp, A. W. (1992). <em>Univariate Discrete Distributions</em> (2nd ed.). Wiley.</li>
</ul>
<hr>
</section>
</section>
<section id="projected-normal" class="level3">
<h3 class="anchored" data-anchor-id="projected-normal">Projected Normal</h3>
<p>The projected normal distribution arises by taking a multivariate normal vector $ X N (, ) $ in $ R^n$ and projecting it to the unit sphere. This distribution is commonly used in directional statistics (data on circles or spheres) and supports asymmetric and even multimodal behaviours depending on parameters.</p>
<section id="args-67" class="level4">
<h4 class="anchored" data-anchor-id="args-67">Args:</h4>
<div class="sourceCode" id="cb133"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>bi.dist.projected_normal(</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter, representing the direction towards which the samples are concentrated. Must be a JAX array with at least one dimension.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-66" class="level4">
<h4 class="anchored" data-anchor-id="returns-66">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI ProjectedNormal distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the ProjectedNormal distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-64" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-64">Example Usage:</h4>
<div class="sourceCode" id="cb134"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>m.dist.projected_normal(concentration<span class="op">=</span>jnp.array([<span class="fl">1.0</span>, <span class="fl">3.0</span>, <span class="fl">2.0</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-46" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-46">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#projectednormal</p>
<hr>
</section>
</section>
<section id="relaxed-bernoulli" class="level3">
<h3 class="anchored" data-anchor-id="relaxed-bernoulli">Relaxed Bernoulli</h3>
<p>The Relaxed Bernoulli is a continuous distribution on the interval <span class="math inline">(0,1)</span> that smoothly approximates the discrete Bernoulli distribution (which has support <span class="math inline">{0,1}</span>). It was introduced to allow for <em>differentiable</em> sampling of approximate binary random variables, which is useful in variational inference and other gradient‐based optimization settings.</p>
<p>The probability density function (PDF) is defined as:</p>
<ul>
<li>It has a <em>temperature</em> parameter <span class="math inline">\tau &gt; 0</span> controlling the smoothness: as <span class="math inline">\tau \to 0</span>, the distribution concentrates near <span class="math inline">{0, 1}</span> (thus approximating the discrete Bernoulli).</li>
<li>For larger <span class="math inline">\tau</span>, the sample becomes more “soft” and less binary, e.g., converging toward 0.5 for very high <span class="math inline">\tau</span>.</li>
<li>It is parameterised by a probability <span class="math inline">p</span> (or equivalently logits <span class="math inline">\ell = \log\frac{p}{1-p}</span>).</li>
</ul>
<p>Let <span class="math inline">X \in (0,1)</span> follow a Relaxed Bernoulli (Binary Concrete) distribution with location parameter <span class="math inline">\alpha &gt; 0</span> (sometimes expressed via logits) and temperature parameter <span class="math inline">\lambda &gt; 0</span>. Then the PDF is:</p>
<p><span class="math display">
p_{,\alpha,;\lambda}(x) =
\frac{;\lambda,\alpha,x^{-\lambda - 1},(1 - x)^{-\lambda - 1};}
{\Bigl(\alpha,x^{-\lambda} + (1 - x)^{-\lambda}\Bigr)^{2}},
\quad \text{for } 0 &lt; x &lt; 1.
</span></p>
<section id="args-68" class="level4">
<h4 class="anchored" data-anchor-id="args-68">Args:</h4>
<div class="sourceCode" id="cb135"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>bi.dist.relaxed_bernoulli(</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>temperature,</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>probs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb135-11"><a href="#cb135-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb135-12"><a href="#cb135-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb135-13"><a href="#cb135-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb135-14"><a href="#cb135-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb135-15"><a href="#cb135-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>temperature</em> (jnp.ndarray): The temperature parameter. Must be greater than 0.</p></li>
<li><p><em>probs</em> (jnp.ndarray, optional): The probability of success. Must be in the interval <code>[0, 1]</code>. Only one of <code>probs</code> or <code>logits</code> can be specified.</p></li>
<li><p><em>logits</em> (jnp.ndarray, optional): The log-odds of success. Must be in the interval <code>[-inf, inf]</code>. Only one of <code>probs</code> or <code>logits</code> can be specified.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-67" class="level4">
<h4 class="anchored" data-anchor-id="returns-67">Returns:</h4>
<p>BI RelaxedBernoulli distribution object (for model building) when <code>sample=False</code>. A JAX array of samples drawn from the RelaxedBernoulli distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-65" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-65">Example Usage:</h4>
<div class="sourceCode" id="cb136"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>m.dist.relaxed_bernoulli(temperature<span class="op">=</span><span class="fl">1.0</span>, probs <span class="op">=</span> jnp.array([<span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>]), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlrelaxedbernoulli" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlrelaxedbernoulli">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#relaxedbernoulli</h4>
</section>
<section id="references-29" class="level4">
<h4 class="anchored" data-anchor-id="references-29">References</h4>
<ul>
<li>https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/RelaxedBernoulli “tfp.distributions.RelaxedBernoulli | TensorFlow Probability”</li>
<li>https://rstudio.github.io/tfprobability/reference/tfd_relaxed_bernoulli.html “RelaxedBernoulli distribution with temperature and logits …”</li>
</ul>
<hr>
</section>
</section>
<section id="relaxed-bernoulli-logits" class="level3">
<h3 class="anchored" data-anchor-id="relaxed-bernoulli-logits">Relaxed Bernoulli Logits</h3>
<p>The Relaxed Bernoulli (logits) is a <strong>continuous</strong> relaxation of the standard Bernoulli distribution, parameterised by logits (or probabilities) and a <strong>temperature</strong> parameter. Rather than producing strictly 0 or 1, it produces values in the continuous interval (0, 1). As the temperature → 0 the distribution approximates a Bernoulli; as temperature → <span class="math inline">\infty</span> the distribution approximates a uniform distribution.</p>
<p>It is used in variational inference and deep-learning contexts to allow gradient-based optimisation through otherwise discrete Bernoulli draws (see the “Concrete” or “Gumbel-Softmax” literature: The Concrete Distribution: A Continuous Relaxation of Discrete Random Variables and Categorical Reparameterization with Gumbel‑Softmax ).</p>
<p>Because this is a continuous relaxation, the “density” is defined over the unit interval (0,1) rather than a pmf at {0,1}. Using logistic or Gumbel‐softmax style construction one can express it as:</p>
<p>If logits = ( ), and temperature = ( &gt; 0 ), then one generates <span class="math display">
u \sim {Uniform}(0,1),\quad
g = -\log(-\log(u)),\quad
x = \sigma\big((\ell + g)/\tau\big)
</span> where $ () $ is the logistic (sigmoid) function. Then <span class="math inline">x \in (0,1)</span> has the RelaxedBernoulli distribution.</p>
<p>The exact density formula can be derived via change of variables from logistic/Gumbel, but is somewhat involved (including Jacobian of sigmoid transform).</p>
<section id="args-69" class="level4">
<h4 class="anchored" data-anchor-id="args-69">Args:</h4>
<div class="sourceCode" id="cb137"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>bi.dist.relaxed_bernoulli_logits(</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>temperature,</span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>logits,</span>
<span id="cb137-4"><a href="#cb137-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb137-5"><a href="#cb137-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb137-6"><a href="#cb137-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb137-7"><a href="#cb137-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb137-8"><a href="#cb137-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb137-9"><a href="#cb137-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb137-10"><a href="#cb137-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb137-11"><a href="#cb137-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb137-12"><a href="#cb137-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb137-13"><a href="#cb137-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb137-14"><a href="#cb137-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>temperature</em> (jnp.ndarray): The temperature parameter, must be positive.</p></li>
<li><p><em>logits</em> (jnp.ndarray): The logits parameter.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-68" class="level4">
<h4 class="anchored" data-anchor-id="returns-68">Returns:</h4>
<p>RelaxedBernoulliLogits: A BI RelaxedBernoulliLogits distribution object (for model building). jnp.ndarray: A JAX array of samples drawn from the RelaxedBernoulliLogits distribution (for direct sampling). RelaxedBernoulliLogits: The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-66" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-66">Example Usage:</h4>
<div class="sourceCode" id="cb138"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>m.dist.relaxed_bernoulli_logits(temperature<span class="op">=</span><span class="fl">1.0</span>, logits<span class="op">=</span><span class="fl">0.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-47" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-47">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#relaxed-bernoulli-logits</p>
</section>
<section id="reference-12" class="level4">
<h4 class="anchored" data-anchor-id="reference-12">Reference:</h4>
<ul>
<li>https://arxiv.org/pdf/1611.00712</li>
</ul>
<hr>
</section>
</section>
<section id="right-truncated" class="level3">
<h3 class="anchored" data-anchor-id="right-truncated">Right Truncated</h3>
<p>A right-truncated distribution is a statistical distribution that arises when the possible values of a random variable are restricted to be below a certain specified value <code>high</code>. In essence, the right tail of the original distribution is “cut off” at a particular point, and the remaining probability is redistributed among the allowable values. This type of distribution is common in various fields where there are inherent upper limits or observational constraints.</p>
<p>The probability density function (PDF) for a continuous right-truncated distribution or the probability mass function (PMF) for a discrete one is derived from the original distribution by normalizing it over the restricted range.</p>
<p><strong>For a continuous random variable X:</strong></p>
<p>If the original probability density function is <code>f(x)</code> and the cumulative distribution function is <code>F(x)</code>, and the distribution is right-truncated at a value <code>b</code> (meaning <code>x ≤ b</code>), the new PDF, <code>f_T(x)</code>, is: <span class="math display">
f_{trunc}(x \mid X \le b) =
\begin{cases}
\displaystyle \frac{,f(x),}{,F(b),},  x \le b, \
0,  x &gt; b.
\end{cases}
</span></p>
<p><strong>For a discrete random variable X:</strong></p>
<p>If the original probability mass function is <code>P(X = k)</code> and the cumulative distribution function is <code>F(k)</code>, and the distribution is right-truncated at a value <code>b</code> (meaning <code>k ≤ b</code>), the new PMF, <code>P_T(X = k)</code>, is:</p>
<p><span class="math display">P_T(X=k) = \begin{cases} \frac{P(X=k)}{F(b)}  \text{if } k \le b \\ 0  \text{if } k &gt; b \end{cases}</span></p>
<section id="args-70" class="level4">
<h4 class="anchored" data-anchor-id="args-70">Args:</h4>
<div class="sourceCode" id="cb139"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>bi.dist.right_truncated_distribution(</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb139-13"><a href="#cb139-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb139-14"><a href="#cb139-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>base_dist</em>: The base distribution to truncate. Must be a univariate distribution with real support.</p></li>
<li><p><em>high</em> (float, jnp.ndarray, optional): The upper truncation point. The support of the new distribution is <span class="math inline">-\infty, \text{high}</span>. Defaults to 0.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-69" class="level4">
<h4 class="anchored" data-anchor-id="returns-69">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI RightTruncatedDistribution distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the RightTruncatedDistribution distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-67" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-67">Example Usage:</h4>
<div class="sourceCode" id="cb140"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>m.dist.right_truncated_distribution(base_dist <span class="op">=</span> m.dist.normal(<span class="dv">0</span>,<span class="dv">1</span>, create_obj <span class="op">=</span> <span class="va">True</span>), high<span class="op">=</span><span class="dv">0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-48" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-48">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#righttruncateddistribution</p>
</section>
<section id="references-30" class="level4">
<h4 class="anchored" data-anchor-id="references-30">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Truncated_normal_distribution</li>
</ul>
<hr>
</section>
</section>
<section id="sine-bivariate-von-mises" class="level3">
<h3 class="anchored" data-anchor-id="sine-bivariate-von-mises">Sine Bivariate Von Mises</h3>
<p><a href="https://en.wikipedia.org/wiki/Bivariate_von_Mises_distribution">In probability theory and statistics, the bivariate von Mises distribution is a probability distribution describing values on a torus. It may be thought of as an analogue on the torus of the bivariate normal distribution. The distribution belongs to the field of directional statistics. The general bivariate von Mises distribution was first proposed by Kanti Mardia in 1975. One of its variants is today used in the field of bioinformatics to formulate a probabilistic model of protein structure in atomic detail, such as backbone-dependent rotamer libraries.</a></p>
<section id="args-71" class="level4">
<h4 class="anchored" data-anchor-id="args-71">Args:</h4>
<div class="sourceCode" id="cb141"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>bi.dist.sine_bivariate_vonmises(</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>phi_loc,</span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>psi_loc,</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>phi_concentration,</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>psi_concentration,</span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a>correlation<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a>weighted_correlation<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb141-10"><a href="#cb141-10" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-11"><a href="#cb141-11" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-12"><a href="#cb141-12" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb141-13"><a href="#cb141-13" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb141-14"><a href="#cb141-14" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb141-15"><a href="#cb141-15" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb141-16"><a href="#cb141-16" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb141-17"><a href="#cb141-17" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>phi_loc</em> (jnp.ndarray): The location parameter for the first angle (phi).</p></li>
<li><p><em>psi_loc</em> (jnp.ndarray): The location parameter for the second angle (psi).</p></li>
<li><p><em>phi_concentration</em> (jnp.ndarray): The concentration parameter for the first angle (phi). Must be positive.</p></li>
<li><p><em>psi_concentration</em> (jnp.ndarray): The concentration parameter for the second angle (psi). Must be positive.</p></li>
<li><p><em>correlation</em> (jnp.ndarray, optional): The correlation parameter between the two angles. One of <code>correlation</code> or <code>weighted_correlation</code> must be specified.</p></li>
<li><p><em>weighted_correlation</em> (jnp.ndarray, optional): An alternative correlation parameter. One of <code>correlation</code> or <code>weighted_correlation</code> must be specified.</p></li>
<li><p><em>validate_args</em> (bool, optional): Whether to enable validation of distribution parameters. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations. If provided, events with a <code>True</code> mask will be conditioned on <code>obs</code>, while the remaining events will be treated as latent variables. Defaults to <code>None</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>shape</em> (tuple, optional): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int, optional): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to <code>False</code>.</p></li>
</ul>
</section>
<section id="returns-70" class="level4">
<h4 class="anchored" data-anchor-id="returns-70">Returns:</h4>
<p>BI.primitives.Messenger: A BI sample site object when used in a model context (<code>sample=False</code>). jnp.ndarray: A JAX array of samples drawn from the SineBivariateVonMises distribution (for direct sampling, <code>sample=True</code>). numpyro.distributions.Distribution: The raw BI distribution object (if <code>create_obj=True</code>).</p>
</section>
<section id="example-usage-68" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-68">Example Usage:</h4>
<div class="sourceCode" id="cb142"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> jax.numpy <span class="im">as</span> jnp</span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Direct sampling</span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> m.dist.sine_bivariate_vonmises(</span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>phi_loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a>psi_loc<span class="op">=</span>jnp.pi,</span>
<span id="cb142-9"><a href="#cb142-9" aria-hidden="true" tabindex="-1"></a>phi_concentration<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb142-10"><a href="#cb142-10" aria-hidden="true" tabindex="-1"></a>psi_concentration<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb142-11"><a href="#cb142-11" aria-hidden="true" tabindex="-1"></a>correlation<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb142-12"><a href="#cb142-12" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb142-13"><a href="#cb142-13" aria-hidden="true" tabindex="-1"></a><span class="op">-</span> <span class="op">*</span>shape<span class="op">*=</span>(<span class="dv">10</span>,)</span>
<span id="cb142-14"><a href="#cb142-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb142-15"><a href="#cb142-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb142-16"><a href="#cb142-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Usage within a model</span></span>
<span id="cb142-17"><a href="#cb142-17" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> my_model():</span>
<span id="cb142-18"><a href="#cb142-18" aria-hidden="true" tabindex="-1"></a>angles <span class="op">=</span> m.dist.sine_bivariate_vonmises(</span>
<span id="cb142-19"><a href="#cb142-19" aria-hidden="true" tabindex="-1"></a>phi_loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb142-20"><a href="#cb142-20" aria-hidden="true" tabindex="-1"></a>psi_loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb142-21"><a href="#cb142-21" aria-hidden="true" tabindex="-1"></a>phi_concentration<span class="op">=</span><span class="fl">2.0</span>,</span>
<span id="cb142-22"><a href="#cb142-22" aria-hidden="true" tabindex="-1"></a>psi_concentration<span class="op">=</span><span class="fl">2.0</span>,</span>
<span id="cb142-23"><a href="#cb142-23" aria-hidden="true" tabindex="-1"></a>weighted_correlation<span class="op">=</span><span class="fl">0.9</span>,</span>
<span id="cb142-24"><a href="#cb142-24" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'angles'</span></span>
<span id="cb142-25"><a href="#cb142-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb142-26"><a href="#cb142-26" aria-hidden="true" tabindex="-1"></a><span class="co"># ... rest of the model</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-49" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-49">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#sinebivariatevonmises</p>
<hr>
</section>
</section>
<section id="sine-skewing" class="level3">
<h3 class="anchored" data-anchor-id="sine-skewing">Sine-skewing</h3>
<p>The sine-skewed von Mises distribution is an extension of the symmetric circular (or toroidal) von Mises (or bivariate von Mises) distribution to allow for skewness (asymmetry) via a sine-based skewing function. It is used to model directional data on the circle (or torus) that depart from symmetry.<a href="https://arxiv.org/abs/1902.02579">1</a></p>
<p><strong>Parameters:</strong></p>
<ul>
<li><p><strong>base_dist:</strong> Base density on a d-dimensional torus. Supported base distributions include: 1D :class:<code>~numpyro.distributions.VonMises</code>, :class:<code>~numnumpyro.distributions.SineBivariateVonMises</code>, 1D :class:<code>~numpyro.distributions.ProjectedNormal</code>, and :class:<code>~numpyro.distributions.Uniform</code> (-pi, pi).</p></li>
<li><p><strong>skewness:</strong> Skewness of the distribution.</p></li>
<li><p><strong>sample (bool, optional):</strong> A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><strong>seed (int, optional):</strong> An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><strong>obs (jnp.ndarray, optional):</strong> The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><strong>name (str, optional):</strong> The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
<p><strong>PDF:</strong></p>
<p>The probability density function (PDF) of the Sine Skewed X distribution is not explicitly defined here, but it is derived from the base distribution and the skewness parameter.</p>
<section id="example-usage-69" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-69">Example Usage:</h4>
<p>from num.pyro import distributions as dist import num.pyro as pyro import num.numpy as np</p>
<p>m = pyro.distributions.Normal(loc=0.0, scale=1.0) skewness = np.array([0.5, 0.5]) sine_skewed = dist.SineSkewed(base_dist=m, skewness=skewness) samples = sine_skewed.sample((1000,))</p>
<div class="sourceCode" id="cb143"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>bi.dist.sine_skewed(</span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>base_dist: numpyro.distributions.distribution.Distribution,</span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>skewness,</span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb143-6"><a href="#cb143-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb143-7"><a href="#cb143-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb143-8"><a href="#cb143-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb143-9"><a href="#cb143-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb143-10"><a href="#cb143-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb143-11"><a href="#cb143-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb143-12"><a href="#cb143-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb143-13"><a href="#cb143-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb143-14"><a href="#cb143-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="softlaplace" class="level3">
<h3 class="anchored" data-anchor-id="softlaplace">SoftLaplace</h3>
<p>Smooth distribution with Laplace-like tail behavior.</p>
<p>This distribution corresponds to the log-convex density:</p>
<p><span class="math display">
z = (value - loc) / scale
log_prob = log(2 / pi) - log(scale) - logaddexp(z, -z)
</span></p>
<p>Like the Laplace density, this density has the heaviest possible tails (asymptotically) while still being log-convex. Unlike the Laplace distribution, this distribution is infinitely differentiable everywhere, and is thus suitable for HMC and Laplace approximation.</p>
<section id="args-72" class="level4">
<h4 class="anchored" data-anchor-id="args-72">Args:</h4>
<div class="sourceCode" id="cb144"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>bi.dist.soft_laplace(</span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a>loc,</span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>scale,</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb144-6"><a href="#cb144-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb144-7"><a href="#cb144-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb144-8"><a href="#cb144-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb144-9"><a href="#cb144-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb144-10"><a href="#cb144-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb144-11"><a href="#cb144-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb144-12"><a href="#cb144-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb144-13"><a href="#cb144-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb144-14"><a href="#cb144-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em>: Location parameter.</p></li>
<li><p><em>scale</em>: Scale parameter.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-71" class="level4">
<h4 class="anchored" data-anchor-id="returns-71">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI SoftLaplace distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the SoftLaplace distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-70" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-70">Example Usage:</h4>
<div class="sourceCode" id="cb145"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a>m.dist.soft_laplace(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-50" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-50">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#softlaplace</p>
<hr>
</section>
</section>
<section id="students-t" class="level3">
<h3 class="anchored" data-anchor-id="students-t">Student’s t</h3>
<p>The Student’s t-distribution is a probability distribution that arises in hypothesis testing involving the mean of a normally distributed population when the population standard deviation is unknown. It is similar to the normal distribution, but has heavier tails, making it more robust to outliers. For large $ <span class="math inline">, it converges to the Normal distribution.</span>$ X t_(, ) $$ where:</p>
<ul>
<li>$ $ is the <strong>location (mean)</strong> parameter</li>
<li>$ &gt; 0 $ is the <strong>scale</strong> parameter</li>
<li>$ &gt; 0 $ is the <strong>degrees of freedom</strong> controlling the tail heaviness</li>
</ul>
<p>The probability density function (pdf) of the Student’s t distribution is: <span class="math display">
f(x \mid \nu, \mu, \sigma) = \frac{\Gamma!\left(\frac{\nu+1}{2}\right)}
{\Gamma!\left(\frac{\nu}{2}\right)\sqrt{\pi\nu}\sigma}
\left[1 + \frac{1}{\nu}
\left(\frac{x - \mu}{\sigma}\right)^2
\right]^{-\frac{\nu+1}{2}}
</span></p>
<ul>
<li>The heavier tails (for small <span class="math inline">ν</span>) allow for larger outliers.</li>
<li>For $ $, this approaches $ {N}(, ^2) $.</li>
</ul>
<section id="args-73" class="level4">
<h4 class="anchored" data-anchor-id="args-73">Args:</h4>
<div class="sourceCode" id="cb146"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a>bi.dist.student_t(</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>df,</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb146-7"><a href="#cb146-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb146-8"><a href="#cb146-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb146-9"><a href="#cb146-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb146-10"><a href="#cb146-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb146-11"><a href="#cb146-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb146-12"><a href="#cb146-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb146-13"><a href="#cb146-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb146-14"><a href="#cb146-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb146-15"><a href="#cb146-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>df</em> (jnp.ndarray): Degrees of freedom, must be positive.</p></li>
<li><p><em>loc</em> (jnp.ndarray): Location parameter, defaults to 0.0.</p></li>
<li><p><em>scale</em> (jnp.ndarray): Scale parameter, defaults to 1.0.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>.expand(shape)</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-72" class="level4">
<h4 class="anchored" data-anchor-id="returns-72">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI StudentT distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the StudentT distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-71" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-71">Example Usage:</h4>
<div class="sourceCode" id="cb147"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a>m.dist.student_t(df <span class="op">=</span> <span class="dv">2</span>, loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-51" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-51">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#studentt</p>
</section>
<section id="references-31" class="level4">
<h4 class="anchored" data-anchor-id="references-31">References:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Student%27s_t-distribution#Probability_density_function">Wikipedia – Student’s t distribution: PDF</a></li>
</ul>
<hr>
</section>
</section>
<section id="truncated-cauchy" class="level3">
<h3 class="anchored" data-anchor-id="truncated-cauchy">Truncated Cauchy</h3>
<p>The Cauchy distribution, also known as the Lorentz distribution, is a continuous probability distribution that appears frequently in various areas of mathematics and physics. It is characterized by its heavy tails, which extend to infinity. The truncated version limits the support of the Cauchy distribution to a specified interval.</p>
<p>Let the base Cauchy have parameters <span class="math inline">\mu</span> (location) and <span class="math inline">\gamma &gt; 0</span> (scale). Let the truncation bounds be <span class="math inline">a</span> (<code>low</code>) and <span class="math inline">b</span> (<code>high</code>) (with <span class="math inline">a &lt; b</span>). Then the PDF of the truncated Cauchy at <span class="math inline">x \in [a,b]</span> is: <span class="math display">
f_{T}(x) = \frac{1}{\pi,\gamma};\frac{1}{1 + \bigl(\tfrac{x - \mu}{\gamma}\bigr)^2} ;\Bigg/ ; \left[F_{{Cauchy}}!\left(\tfrac{b - \mu}{\gamma}\right) - F_{{Cauchy}}!\left(\tfrac{a - \mu}{\gamma}\right)\right],
</span> and <span class="math inline">f_{T}(x)=0</span> for <span class="math inline">x &lt; a</span> or <span class="math inline">x &gt; b</span>.</p>
<p>Here <span class="math inline">F_{{Cauchy}}</span> is the standard Cauchy CDF $with parameters <span class="math inline">\mu,\gamma)</span>.</p>
<section id="args-74" class="level4">
<h4 class="anchored" data-anchor-id="args-74">Args:</h4>
<div class="sourceCode" id="cb148"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>bi.dist.truncated_cauchy(</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-9"><a href="#cb148-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-10"><a href="#cb148-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb148-11"><a href="#cb148-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb148-12"><a href="#cb148-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb148-13"><a href="#cb148-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb148-14"><a href="#cb148-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb148-15"><a href="#cb148-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb148-16"><a href="#cb148-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (float): Location parameter of the Cauchy distribution.</p></li>
<li><p><em>scale</em> (float): Scale parameter of the Cauchy distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-73" class="level4">
<h4 class="anchored" data-anchor-id="returns-73">Returns:</h4>
<p>BI TruncatedCauchy distribution object (for model building) when <code>sample=False</code>.</p>
<p>JAX array of samples drawn from the TruncatedCauchy distribution (for direct sampling) when <code>sample=True</code>.</p>
<p>The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-72" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-72">Example Usage:</h4>
<div class="sourceCode" id="cb149"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a>m.dist.truncated_cauchy(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncatedcauchy" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncatedcauchy">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#truncatedcauchy</h4>
</section>
<section id="references-32" class="level4">
<h4 class="anchored" data-anchor-id="references-32">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Cauchy_distribution “Cauchy distribution”</li>
<li>https://www.tandfonline.com/doi/full/10.1080/00207390600595223 “Full article: A truncated Cauchy distribution”</li>
<li>https://rstudio.github.io/tfprobability/reference/tfd_truncated_cauchy.html “The Truncated Cauchy distribution. — tfd_truncated_cauchy”</li>
</ul>
<hr>
</section>
</section>
<section id="truncated" class="level3">
<h3 class="anchored" data-anchor-id="truncated">Truncated</h3>
<p>Samples from a Truncated Distribution.</p>
<p>A <strong>truncated distribution</strong> arises when you take a random variable <span class="math inline">X</span> that originally has some distribution (with PDF <span class="math inline">f_X(x)</span> and CDF <span class="math inline">F_X(x)</span>) and you restrict attention only to those values of <span class="math inline">X</span> that are <em>above</em> a given truncation point <span class="math inline">a</span>. In other words you only observe <span class="math inline">X</span> when <span class="math inline">X &gt; a</span>. All the “mass” below (or equal to) <span class="math inline">a</span> is <strong>excluded</strong> (not just unobserved, but removed from the sample/analysis). This differs from <em>censoring</em>, where values below a threshold might be known (for example “&lt; a”), but here they are entirely excluded from the domain. Left truncation is common in many applied fields — for instance:</p>
<ul>
<li>In survival analysis: subjects whose event time happens before the study start are not included.</li>
<li>In insurance or losses: only losses above a deductible (threshold) are recorded, so the loss distribution is left-truncated at that deductible.</li>
<li>In industrial/life-data: items used before data collection start, so only those with lifetime &gt; some lower bound are observed.</li>
</ul>
<p>Let (<span class="math inline">X</span>) be a random variable with PDF. Choose truncation points <span class="math inline">a &lt; X \le b</span>. Then the truncated distribution (<span class="math inline">Y = X \mid a &lt; X \le b</span>) has:<br>
<span class="math display">
f_Y(y) = \frac{f_X(y)}{F_X(b) - F_X(a)}, \quad \text{for } a &lt; y \le b
</span> and (0) outside that interval. * The lower bound (a) can be <span class="math inline">-\infty</span> (so only an upper truncation), and the upper bound (<span class="math inline">b</span>) can be <span class="math inline">+\infty</span> (so only a lower truncation). * If both bounds are finite (i.e., doubly truncated), that is a valid and often‐used scenario.</p>
<section id="args-75" class="level4">
<h4 class="anchored" data-anchor-id="args-75">Args:</h4>
<div class="sourceCode" id="cb150"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>bi.dist.truncated_distribution(</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-5"><a href="#cb150-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-6"><a href="#cb150-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb150-7"><a href="#cb150-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-8"><a href="#cb150-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-9"><a href="#cb150-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb150-10"><a href="#cb150-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb150-11"><a href="#cb150-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb150-12"><a href="#cb150-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb150-13"><a href="#cb150-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb150-14"><a href="#cb150-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb150-15"><a href="#cb150-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>base_dist</em>: The base distribution to be truncated. This should be a univariate distribution. Currently, only the following distributions are supported: Cauchy, Laplace, Logistic, Normal, and StudentT.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-74" class="level4">
<h4 class="anchored" data-anchor-id="returns-74">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI TruncatedDistribution distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the TruncatedDistribution distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-73" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-73">Example Usage:</h4>
<div class="sourceCode" id="cb151"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb151-2"><a href="#cb151-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb151-3"><a href="#cb151-3" aria-hidden="true" tabindex="-1"></a>m.dist.truncated_distribution(base_dist <span class="op">=</span> m.dist.normal(<span class="dv">0</span>,<span class="dv">1</span>, create_obj <span class="op">=</span> <span class="va">True</span>), high<span class="op">=</span><span class="dv">1</span>, low <span class="op">=</span> <span class="dv">0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncateddistribution" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncateddistribution">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#truncateddistribution</h4>
</section>
<section id="references-33" class="level4">
<h4 class="anchored" data-anchor-id="references-33">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Truncated_distribution “Truncated distribution”</li>
<li>https://pages.stern.nyu.edu/~wgreene/Text/Edition8/PDF/M19_GREE1366_08_SE_C19.pdf “19”</li>
<li>https://search.r-project.org/CRAN/refmans/LaplacesDemon/html/dist.Truncated.html “R: Truncated Distributions”</li>
<li>https://dspace.cuni.cz/bitstream/handle/20.500.11956/127921/130308341.pdf?isAllowed=y&amp;sequence=1&amp;utm_source.com “Truncated data thesis”</li>
</ul>
<hr>
</section>
</section>
<section id="truncated-normal" class="level3">
<h3 class="anchored" data-anchor-id="truncated-normal">Truncated Normal</h3>
<p>A truncated normal distribution is derived from a normal (Gaussian) random variable by restricting (truncating) its domain to an interval <span class="math inline">[a, b]</span> (which could be one‐sided, e.g., (a) only or (b) only). It is defined by its location (<code>loc</code>), scale (<code>scale</code>), lower bound <span class="math inline">a</span> (<code>low</code>), and upper bound <span class="math inline">b</span> (<code>high</code>). In effect: if <span class="math inline">X \sim  N(\mu, \sigma^2)</span>, then the truncated version <span class="math inline">Y = X | (a \le X \le b)</span> has the same “shape” but only supports values in <span class="math inline">[a,b]</span>. This is used when you know that values outside a range are impossible or not observed (e.g., measurement limits, natural bounds).</p>
<p>Let <span class="math inline">X \sim  N(\mu, \sigma^2)</span>. Define truncation bounds <span class="math inline">a &lt; b</span> (could be <span class="math inline">a = -\infty</span> or <span class="math inline">b = +\infty</span> for one‐sided truncation). Then for <span class="math inline">y \in [a,b]</span>,</p>
<p><span class="math display">
f_Y(y) = \frac{1}{\sigma} ; \frac{\varphi!\bigl(\frac{y - \mu}{\sigma}\bigr)}{\Phi!\bigl(\frac{b - \mu}{\sigma}\bigr) - \Phi!\bigl(\frac{a - \mu}{\sigma}\bigr)} ,
</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">\varphi(z)</span> is the standard normal PDF, <span class="math inline">\varphi(z) = \frac1{\sqrt{2\pi}} e^{-z^2/2}</span>.</li>
<li><span class="math inline">\Phi(z)</span> is the standard normal CDF.</li>
<li>The denominator normalises the density so the total probability over (<span class="math inline">[a,b]</span>) is 1.</li>
</ul>
<p>For values <span class="math inline">y &lt; a</span> or <span class="math inline">y &gt; b</span>, <span class="math inline">f_Y(y) = 0</span>.</p>
<section id="args-76" class="level4">
<h4 class="anchored" data-anchor-id="args-76">Args:</h4>
<div class="sourceCode" id="cb152"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a>bi.dist.truncated_normal(</span>
<span id="cb152-2"><a href="#cb152-2" aria-hidden="true" tabindex="-1"></a>loc<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb152-3"><a href="#cb152-3" aria-hidden="true" tabindex="-1"></a>scale<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb152-4"><a href="#cb152-4" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-5"><a href="#cb152-5" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-6"><a href="#cb152-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-7"><a href="#cb152-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb152-8"><a href="#cb152-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-9"><a href="#cb152-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-10"><a href="#cb152-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb152-11"><a href="#cb152-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb152-12"><a href="#cb152-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb152-13"><a href="#cb152-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb152-14"><a href="#cb152-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb152-15"><a href="#cb152-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb152-16"><a href="#cb152-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>loc</em> (float): The location parameter of the normal distribution.</p></li>
<li><p><em>sample</em> (float): The scale parameter of the normal distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-75" class="level4">
<h4 class="anchored" data-anchor-id="returns-75">Returns:</h4>
<p>BI TruncatedNormal distribution object (for model building). JAX array of samples drawn from the TruncatedNormal distribution (for direct sampling). The raw BI distribution object (for advanced use cases).</p>
</section>
<section id="example-usage-74" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-74">Example Usage:</h4>
<div class="sourceCode" id="cb153"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>m.dist.truncated_normal(loc<span class="op">=</span><span class="fl">0.0</span>, scale<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncatednormal_lowercase" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltruncatednormal_lowercase">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#truncatednormal_lowercase</h4>
</section>
<section id="references-34" class="level4">
<h4 class="anchored" data-anchor-id="references-34">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Truncated_normal_distribution “Truncated normal distribution”</li>
<li>https://people.math.sc.edu/burkardt/m_src/truncated_normal/truncated_normal.html “TRUNCATED_NORMAL - The Truncated Normal Distribution”</li>
<li>https://www.statisticshowto.com/truncated-normal-distribution/ “Truncated Distribution / Truncated Normal Distribution - Statistics …”</li>
<li>https://real-statistics.com/normal-distribution/truncated-normal-distribution/ “Truncated Normal Distribution | Real Statistics Using Excel”</li>
<li>https://stats.stackexchange.com/questions/525894/understanding-the-pdf-of-a-truncated-normal-distribution “Understanding the pdf of a truncated normal distribution”</li>
</ul>
<hr>
</section>
</section>
<section id="truncated-polyagamma" class="level3">
<h3 class="anchored" data-anchor-id="truncated-polyagamma">Truncated PolyaGamma</h3>
<p>Samples from a Truncated PolyaGamma distribution.</p>
<p>This distribution is a truncated version of the PolyaGamma distribution, defined over the interval [0, truncation_point]. It is often used in Bayesian non-parametric models.</p>
<section id="args-77" class="level4">
<h4 class="anchored" data-anchor-id="args-77">Args:</h4>
<div class="sourceCode" id="cb154"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>bi.dist.truncated_polya_gamma(</span>
<span id="cb154-2"><a href="#cb154-2" aria-hidden="true" tabindex="-1"></a>batch_shape<span class="op">=</span>(),</span>
<span id="cb154-3"><a href="#cb154-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb154-4"><a href="#cb154-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb154-5"><a href="#cb154-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb154-6"><a href="#cb154-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb154-7"><a href="#cb154-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb154-8"><a href="#cb154-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb154-9"><a href="#cb154-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb154-10"><a href="#cb154-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb154-11"><a href="#cb154-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb154-12"><a href="#cb154-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb154-13"><a href="#cb154-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>batch_shape</em> (tuple): The shape of the batch dimension.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions.</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-76" class="level4">
<h4 class="anchored" data-anchor-id="returns-76">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Truncated PolyaGamma distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Truncated PolyaGamma distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-75" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-75">Example Usage:</h4>
<div class="sourceCode" id="cb155"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb155-2"><a href="#cb155-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb155-3"><a href="#cb155-3" aria-hidden="true" tabindex="-1"></a>m.dist.truncated_polya_gamma(batch_shape<span class="op">=</span>(), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-52" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-52">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#truncatedpolygammadistribution</p>
<hr>
</section>
</section>
<section id="two-sided-truncated" class="level3">
<h3 class="anchored" data-anchor-id="two-sided-truncated">Two Sided Truncated</h3>
<p>A “two-sided truncated distribution” is a general concept: you take a base continuous distribution and <strong>restrict it</strong> to an interval ([<code>low</code>, <code>high</code>]), discarding all mass outside, then <strong>renormalize</strong> so the inner portion integrates to 1. I’ll spell out the general formulas, caveats, sampling strategies, and special cases (e.g.&nbsp;truncated normal) to illustrate.</p>
<p><span class="math display">
f(x) =
\begin{cases}
\dfrac{p(x)}{P(\text{low} \le X \le \text{high})},  \text{if } \text{low} \le x \le \text{high}, \\[6pt]
0,  \text{otherwise}.
\end{cases}
</span></p>
<p>where <span class="math inline">p(x)</span> is the probability density function of the base distribution.</p>
<section id="args-78" class="level4">
<h4 class="anchored" data-anchor-id="args-78">Args:</h4>
<div class="sourceCode" id="cb156"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb156-1"><a href="#cb156-1" aria-hidden="true" tabindex="-1"></a>bi.dist.two_sided_truncated_distribution(</span>
<span id="cb156-2"><a href="#cb156-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb156-3"><a href="#cb156-3" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb156-4"><a href="#cb156-4" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb156-5"><a href="#cb156-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb156-6"><a href="#cb156-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb156-7"><a href="#cb156-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb156-8"><a href="#cb156-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb156-9"><a href="#cb156-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb156-10"><a href="#cb156-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb156-11"><a href="#cb156-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb156-12"><a href="#cb156-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb156-13"><a href="#cb156-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb156-14"><a href="#cb156-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb156-15"><a href="#cb156-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>base_dist</em>: The base distribution to truncate.</p></li>
<li><p><em>low</em>: The lower bound for truncation.</p></li>
<li><p><em>high</em>: The upper bound for truncation.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-77" class="level4">
<h4 class="anchored" data-anchor-id="returns-77">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI TwoSidedTruncatedDistribution distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the TwoSidedTruncatedDistribution distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltwosidedtruncateddistribution" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmltwosidedtruncateddistribution">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#twosidedtruncateddistribution</h4>
<hr>
</section>
</section>
<section id="uniform" class="level3">
<h3 class="anchored" data-anchor-id="uniform">Uniform</h3>
<p>The Uniform distribution is the simplest continuous distribution: every value in the interval ([a, b]) is <strong>equally likely</strong>. It is widely used for modeling complete randomness within a fixed range, random sampling, and as a building block for other distributions.</p>
<p><span class="math display">   
f(x) =
\begin{cases}
\dfrac{1}{b - a},  a \le x \le b, [2mm]
0,  \text{otherwise}.
\end{cases}
</span></p>
<section id="args-79" class="level4">
<h4 class="anchored" data-anchor-id="args-79">Args:</h4>
<div class="sourceCode" id="cb157"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>bi.dist.uniform(</span>
<span id="cb157-2"><a href="#cb157-2" aria-hidden="true" tabindex="-1"></a>low<span class="op">=</span><span class="fl">0.0</span>,</span>
<span id="cb157-3"><a href="#cb157-3" aria-hidden="true" tabindex="-1"></a>high<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb157-4"><a href="#cb157-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb157-5"><a href="#cb157-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb157-6"><a href="#cb157-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb157-7"><a href="#cb157-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb157-8"><a href="#cb157-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb157-9"><a href="#cb157-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb157-10"><a href="#cb157-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb157-11"><a href="#cb157-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb157-12"><a href="#cb157-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb157-13"><a href="#cb157-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb157-14"><a href="#cb157-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>low</em> (jnp.ndarray): The lower bound of the uniform interval.</p></li>
<li><p><em>high</em> (jnp.ndarray): The upper bound of the uniform interval.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>. Defaults to False.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-78" class="level4">
<h4 class="anchored" data-anchor-id="returns-78">Returns:</h4>
<p>BI Uniform distribution object (for model building) when <code>sample=False</code>.</p>
<p>JAX array of samples drawn from the Uniform distribution (for direct sampling) when <code>sample=True</code>.</p>
<p>The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-76" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-76">Example Usage:</h4>
<div class="sourceCode" id="cb158"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb158-1"><a href="#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb158-2"><a href="#cb158-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb158-3"><a href="#cb158-3" aria-hidden="true" tabindex="-1"></a>m.dist.uniform(low<span class="op">=</span><span class="fl">0.0</span>, high<span class="op">=</span><span class="fl">1.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-53" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-53">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#uniform</p>
</section>
<section id="references-35" class="level4">
<h4 class="anchored" data-anchor-id="references-35">References:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Uniform_distribution_%28continuous%29">Wikipedia – Uniform distribution (continuous)</a></li>
</ul>
<hr>
</section>
</section>
<section id="unit" class="level3">
<h3 class="anchored" data-anchor-id="unit">Unit</h3>
<ul>
<li>The Unit distribution is the simplest continuous probability distribution.</li>
<li>Every number in ([0,1]) is equally likely.</li>
<li>Often used in simulation, Monte Carlo methods, and as a building block for generating other distributions via the inverse-CDF method. <span class="math display">
f(x) =
\begin{cases}
1,  0 \le x \le 1, [1mm]
0,  \text{otherwise}.
\end{cases}
</span></li>
</ul>
<section id="args-80" class="level4">
<h4 class="anchored" data-anchor-id="args-80">Args:</h4>
<div class="sourceCode" id="cb159"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>bi.dist.unit(</span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a>log_factor,</span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb159-5"><a href="#cb159-5" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb159-6"><a href="#cb159-6" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb159-7"><a href="#cb159-7" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb159-8"><a href="#cb159-8" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb159-9"><a href="#cb159-9" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb159-10"><a href="#cb159-10" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb159-11"><a href="#cb159-11" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb159-12"><a href="#cb159-12" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb159-13"><a href="#cb159-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>log_factor</em> (jnp.ndarray): Log factor for the unit distribution. This parameter determines the <em>shape</em> and batch size of the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-79" class="level4">
<h4 class="anchored" data-anchor-id="returns-79">Returns:</h4>
<p>BI Unit distribution object: When <code>sample=False</code> (for model building). jnp.ndarray: A JAX array of samples drawn from the Unit distribution (for direct sampling). BI Unit distribution object: When <code>create_obj=True</code> (for advanced use cases).</p>
</section>
<section id="example-usage-77" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-77">Example Usage:</h4>
<div class="sourceCode" id="cb160"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb160-1"><a href="#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb160-2"><a href="#cb160-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb160-3"><a href="#cb160-3" aria-hidden="true" tabindex="-1"></a>m.dist.unit(log_factor<span class="op">=</span>jnp.ones(<span class="dv">5</span>), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-54" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-54">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#unit</p>
<hr>
</section>
</section>
<section id="weibull" class="level3">
<h3 class="anchored" data-anchor-id="weibull">Weibull</h3>
<p>Samples from a Weibull distribution.</p>
<p>The Weibull distribution is widely used for modeling <strong>lifetime or reliability data</strong>. Its shape parameter (k) controls the hazard function:</p>
<ul>
<li>(k &lt; 1): decreasing hazard (infant mortality)</li>
<li>(k = 1): constant hazard → reduces to <strong>Exponential distribution</strong></li>
<li>(k &gt; 1): increasing hazard (aging/failure over time)</li>
</ul>
<p><span class="math display">
f(x \mid \lambda, k) =
\begin{cases}
\dfrac{k}{\lambda} \left(\dfrac{x}{\lambda}\right)^{k-1} e^{-(x/\lambda)^k},  x \ge 0, [2mm]
0,  x &lt; 0
\end{cases}
</span></p>
<section id="args-81" class="level4">
<h4 class="anchored" data-anchor-id="args-81">Args:</h4>
<div class="sourceCode" id="cb161"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a>bi.dist.weibull(</span>
<span id="cb161-2"><a href="#cb161-2" aria-hidden="true" tabindex="-1"></a>scale,</span>
<span id="cb161-3"><a href="#cb161-3" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb161-4"><a href="#cb161-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb161-5"><a href="#cb161-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb161-6"><a href="#cb161-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb161-7"><a href="#cb161-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb161-8"><a href="#cb161-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb161-9"><a href="#cb161-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb161-10"><a href="#cb161-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb161-11"><a href="#cb161-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb161-12"><a href="#cb161-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb161-13"><a href="#cb161-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb161-14"><a href="#cb161-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (jnp.ndarray): The scale parameter of the Weibull distribution. Must be positive.</p></li>
<li><p><em>concentration</em> (jnp.ndarray): The concentration parameter of the Weibull distribution. Must be positive.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-80" class="level4">
<h4 class="anchored" data-anchor-id="returns-80">Returns:</h4>
<p>BI Weibull distribution object (for model building) when <code>sample=False</code>. JAX array of samples drawn from the Weibull distribution (for direct sampling) when <code>sample=True</code>. The raw BI distribution object (for advanced use cases) when <code>create_obj=True</code>.</p>
</section>
<section id="example-usage-78" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-78">Example Usage:</h4>
<div class="sourceCode" id="cb162"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb162-1"><a href="#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb162-2"><a href="#cb162-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb162-3"><a href="#cb162-3" aria-hidden="true" tabindex="-1"></a>m.dist.weibull(scale<span class="op">=</span><span class="fl">1.0</span>, concentration<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-55" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-55">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#weibull</p>
</section>
<section id="references-36" class="level4">
<h4 class="anchored" data-anchor-id="references-36">References:</h4>
<ul>
<li><p><a href="https://en.wikipedia.org/wiki/Weibull_distribution#Probability_density_function">Wikipedia – Weibull distribution</a></p></li>
<li><p>Weibull, W. (1951). <em>A Statistical Distribution Function of Wide Applicability.</em> Journal of Applied Mechanics.</p></li>
</ul>
<hr>
</section>
</section>
<section id="wishart" class="level3">
<h3 class="anchored" data-anchor-id="wishart">Wishart</h3>
<p>The Wishart distribution is a multivariate distribution used to model positive definite matrices, often representing covariance matrices. It’s commonly used in Bayesian statistics and machine learning, particularly in models involving covariance estimation.</p>
<p><span class="math display">
X \sim \text{Wishart}(scale_matrix=V, concentration=n)
</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">p</span> = dimension of the square matrices</li>
<li><span class="math inline">{V} \in {R}^{p \times p}</span> = positive definite <strong>scale matrix</strong> (<code>scale_matrix</code>)</li>
<li><span class="math inline">n \ge p</span> = <strong>degrees of freedom</strong> (<code>concentration</code>)</li>
</ul>
<p>The Wishart distribution is the <strong>multivariate generalization of the chi-squared distribution</strong>. It is commonly used as the distribution of the <strong>sample covariance matrix</strong> for multivariate normal samples.</p>
<p><span class="math display">
f({X} \mid {V}, n) =
\frac{ |{X}|^{(n-p-1)/2} \exp\big(-\frac{1}{2} {tr}({V}^{-1} {X}) \big) }
{ 2^{np/2} |{V}|^{n/2} \Gamma_p(n/2) }, \quad {X} \succ 0
</span></p>
<p>where:</p>
<ul>
<li><span class="math inline">|{X}|</span> = determinant of <span class="math inline">{X}</span></li>
<li><span class="math inline">{tr}(\cdot)</span> = trace</li>
<li><span class="math inline">\Gamma_p(\cdot)</span> = multivariate Gamma function:</li>
</ul>
<section id="args-82" class="level4">
<h4 class="anchored" data-anchor-id="args-82">Args:</h4>
<div class="sourceCode" id="cb163"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>bi.dist.wishart(</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a>scale_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>rate_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-5"><a href="#cb163-5" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-6"><a href="#cb163-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-7"><a href="#cb163-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb163-8"><a href="#cb163-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-9"><a href="#cb163-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-10"><a href="#cb163-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb163-11"><a href="#cb163-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb163-12"><a href="#cb163-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb163-13"><a href="#cb163-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb163-14"><a href="#cb163-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb163-15"><a href="#cb163-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb163-16"><a href="#cb163-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>concentration</em> (jnp.ndarray): Positive concentration parameter analogous to the concentration of a Gamma distribution. The concentration must be larger than the dimensionality of the scale matrix.</p></li>
<li><p><em>scale_matrix</em> (jnp.ndarray, optional): Scale matrix analogous to the inverse rate of a Gamma distribution.</p></li>
<li><p><em>rate_matrix</em> (jnp.ndarray, optional): Rate matrix anaologous to the rate of a Gamma distribution.</p></li>
<li><p><em>scale_tril</em> (jnp.ndarray, optional): Cholesky decomposition of the :code:<code>scale_matrix</code>.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool, optional): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool, optional): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-81" class="level4">
<h4 class="anchored" data-anchor-id="returns-81">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI Wishart distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the Wishart distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-79" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-79">Example Usage:</h4>
<div class="sourceCode" id="cb164"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb164-1"><a href="#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb164-2"><a href="#cb164-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb164-3"><a href="#cb164-3" aria-hidden="true" tabindex="-1"></a>m.dist.wishart(concentration<span class="op">=</span><span class="fl">5.0</span>, scale_matrix<span class="op">=</span>jnp.eye(<span class="dv">2</span>), sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-56" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-56">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#wishart</p>
</section>
<section id="references-37" class="level4">
<h4 class="anchored" data-anchor-id="references-37">References:</h4>
<ul>
<li><a href="https://en.wikipedia.org/wiki/Wishart_distribution">Wikipedia – Wishart distribution</a></li>
</ul>
<hr>
</section>
</section>
<section id="wishart-cholesky" class="level3">
<h3 class="anchored" data-anchor-id="wishart-cholesky">Wishart Cholesky</h3>
<ul>
<li>The <strong>Wishart</strong> distribution is a distribution over positive definite matrices, often used as a prior for covariance or precision matrices in multivariate normal models.</li>
<li>The <strong>Cholesky parameterization</strong> of the Wishart (called “wishart_cholesky” in Stan, for example) reparameterizes the Wishart over its <strong>lower (or upper) triangular Cholesky factor</strong>. This is useful for numerical stability and unconstrained parameterization in Bayesian sampling frameworks.</li>
<li>In this parameterization, one works with a lower‐triangular matrix <span class="math inline">L_W</span> such that <span class="math display">
\Sigma = L_W L_W^\top
</span> and imposes a density over <span class="math inline">L_W</span> corresponding to the induced Wishart density on ().</li>
</ul>
<p>The probability density function (PDF) is given by:</p>
<p>If the dimension is <span class="math inline">K</span>, degrees of freedom <span class="math inline">\nu &gt; K - 1</span>, and <span class="math inline">L_S</span> is the lower‐triangular Cholesky factor of the scale matrix <span class="math inline">S</span>, then the log density for <span class="math inline">L_W</span> under the <strong>wishart_cholesky</strong> distribution is:</p>
<p><span class="math display">
\log p(L_W \mid \nu, L_S)
= \log \bigl[ \text{Wishart}(L_W L_W^\top \mid \nu, L_S L_S^\top) \bigr] + \log \bigl| J_{f^{-1}} \bigr|
</span></p>
<p>where <span class="math inline">J_{f^{-1}}</span> is the Jacobian of the transformation from <span class="math inline">L_W</span> to <span class="math inline">\Sigma = L_W L_W^\top</span>.</p>
<p>The Jacobian term (absolute log determinant) is:</p>
<p><span class="math display">
\log |J_{f^{-1}}| = K \log 2 ; \sum_{k=1}^K (K - k + 1),\log L_{W_{k,k}}.
</span></p>
<p>Thus the density includes both the usual Wishart density on <span class="math inline">\Sigma</span> plus this extra Jacobian (which penalizes or weights the diagonal entries of <span class="math inline">L_W</span>).</p>
<section id="parameters-1" class="level4">
<h4 class="anchored" data-anchor-id="parameters-1">Parameters</h4>
<ul>
<li><p>concentration: (Tensor) Positive concentration parameter analogous to the concentration of a :class:<code>Gamma</code> distribution. The concentration must be larger than the dimensionality of the scale matrix.</p></li>
<li><p>scale_matrix: (Tensor, optional) Scale matrix analogous to the inverse rate of a :class:<code>Gamma</code> distribution. If not provided, <code>rate_matrix</code> or <code>scale_tril</code> must be.</p></li>
<li><p>rate_matrix: (Tensor, optional) Rate matrix anaologous to the rate of a :class:<code>Gamma</code> distribution. If not provided, <code>scale_matrix</code> or <code>scale_tril</code> must be.</p></li>
<li><p>scale_tril: (Tensor, optional) Cholesky decomposition of the :code:<code>scale_matrix</code>. If not provided, <code>scale_matrix</code> or <code>rate_matrix</code> must be.</p></li>
<li><p>sample (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p>seed (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p>obs (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p>name (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="references-38" class="level4">
<h4 class="anchored" data-anchor-id="references-38">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Wishart_distribution “Wishart distribution”</li>
<li>https://mc-stan.org/docs/2_32/functions-reference/wishart-cholesky-distribution.html”28.2 Wishart distribution, Cholesky Parameterization - Stan”</li>
</ul>
<div class="sourceCode" id="cb165"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>bi.dist.wishart_cholesky(</span>
<span id="cb165-2"><a href="#cb165-2" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb165-3"><a href="#cb165-3" aria-hidden="true" tabindex="-1"></a>scale_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-4"><a href="#cb165-4" aria-hidden="true" tabindex="-1"></a>rate_matrix<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-5"><a href="#cb165-5" aria-hidden="true" tabindex="-1"></a>scale_tril<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-6"><a href="#cb165-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-7"><a href="#cb165-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb165-8"><a href="#cb165-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-9"><a href="#cb165-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-10"><a href="#cb165-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb165-11"><a href="#cb165-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb165-12"><a href="#cb165-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb165-13"><a href="#cb165-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb165-14"><a href="#cb165-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb165-15"><a href="#cb165-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb165-16"><a href="#cb165-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<hr>
</section>
</section>
<section id="generic-zero-inflated" class="level3">
<h3 class="anchored" data-anchor-id="generic-zero-inflated">Generic Zero Inflated</h3>
<p>A Zero-Inflated distribution combines a base distribution with a Bernoulli distribution to model data with an excess of zero values. It assumes that each observation is either drawn from the base distribution or is a zero with probability determined by the Bernoulli distribution (the “gate”). A <strong>zero-inflated distribution</strong> arises when you take a random variable <span class="math inline">X</span> that originally has some distribution (with PMF <span class="math inline">f_X(x)</span> ) <em>and</em> you add extra mass at zero. In other words you only observe <span class="math inline">X</span> when <span class="math inline">X \ge 0</span> (if count) and you assume there are more zeros than what the original distribution predicts — so you mix in a point-mass at zero. A zero-inflated model assumes two processes: a “structural zero” process with probability <span class="math inline">\pi</span>, and another process (the base count/distribution) with probability <span class="math inline">1 - \pi</span>, which itself may generate zeros <em>or</em> non-zeros.</p>
<ul>
<li><p>When / Why use it:</p></li>
<li><p><strong>Data collection constraints</strong>: You can only observe values in a process when some condition is met; others are automatically zero (for example, when the event cannot happen for some units).</p></li>
<li><p><strong>Mixtures of processes</strong>: Some observations are “structural zeros” (no risk) and others follow a regular count process (with risk).</p></li>
<li><p><strong>Overdispersion &amp; excess zeros</strong>: If you try a standard count distribution (Poisson, Negative Binomial) and you observe many more zeros than predicted (given the non-zero counts), a zero-inflated alternative may fit better.</p></li>
<li><p>The general PMF form is (for discrete count models): <span class="math display">
P(X=0) = \pi + (1 - \pi),P_{\text{base}}(0),
\quad
P(X=k) = (1 - \pi),P_{\text{base}}(k), ; k &gt; 0.
</span></p></li>
</ul>
<section id="args-83" class="level4">
<h4 class="anchored" data-anchor-id="args-83">Args:</h4>
<div class="sourceCode" id="cb166"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>bi.dist.zero_inflated_distribution(</span>
<span id="cb166-2"><a href="#cb166-2" aria-hidden="true" tabindex="-1"></a>base_dist,</span>
<span id="cb166-3"><a href="#cb166-3" aria-hidden="true" tabindex="-1"></a>gate<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-4"><a href="#cb166-4" aria-hidden="true" tabindex="-1"></a>gate_logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-5"><a href="#cb166-5" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-6"><a href="#cb166-6" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb166-7"><a href="#cb166-7" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-8"><a href="#cb166-8" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-9"><a href="#cb166-9" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb166-10"><a href="#cb166-10" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb166-11"><a href="#cb166-11" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb166-12"><a href="#cb166-12" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb166-13"><a href="#cb166-13" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb166-14"><a href="#cb166-14" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb166-15"><a href="#cb166-15" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>base_dist</em> (Distribution): The base distribution to be zero-inflated (e.g., Poisson, NegativeBinomial).</p></li>
<li><p><em>gate</em> (jnp.ndarray, optional): Probability of extra zeros (between 0 and 1).</p></li>
<li><p><em>gate_logits</em> (jnp.ndarray, optional): Log-odds of extra zeros.</p></li>
<li><p><em>validate_args</em> (bool, optional): Whether to validate parameter values. Defaults to None.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-82" class="level4">
<h4 class="anchored" data-anchor-id="returns-82">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI ZeroInflatedDistribution distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the ZeroInflatedDistribution distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-80" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-80">Example Usage:</h4>
<div class="sourceCode" id="cb167"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>m.dist.zero_inflated_distribution(base_dist<span class="op">=</span>m.dist.poisson(rate<span class="op">=</span><span class="dv">5</span>, create_obj <span class="op">=</span> <span class="va">True</span>), gate <span class="op">=</span> <span class="fl">0.3</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflateddistribution" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflateddistribution">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#zeroinflateddistribution</h4>
</section>
<section id="references-39" class="level4">
<h4 class="anchored" data-anchor-id="references-39">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Zero-inflated_model “Zero-inflated model”</li>
<li>https://larmarange.github.io/guide-R/analyses_avancees/modeles-zero-inflated.html “48 Modèles de comptage zero-inflated et hurdle – guide-R”</li>
</ul>
<hr>
</section>
</section>
<section id="zero-inflated-negative-binomial" class="level3">
<h3 class="anchored" data-anchor-id="zero-inflated-negative-binomial">Zero-Inflated Negative Binomial</h3>
<p>A Zero-Inflated Negative Binomial distribution is used for count data that exhibit <strong>both</strong> (a) over-dispersion relative to a Poisson (i.e., variance &gt; mean) <em>and</em> (b) an excess of zero counts beyond what a standard Negative Binomial would predict. It assumes two latent processes:</p>
<ol type="1">
<li>With probability $ $ (sometimes denoted <span class="math inline">\psi</span> or “zero-inflation probability”) you are in a “structural zero” state → you observe a zero.</li>
<li>With probability <span class="math inline">1 - \pi</span>, you come from a regular Negative Binomial distribution (with parameters e.g.&nbsp;mean $ $ and dispersion parameter $ $ or size/r parameter) and then you might observe zero or a positive count.</li>
</ol>
<p>Thus the model is a mixture of a point‐mass at zero + a Negative Binomial for counts.</p>
<p>This distribution combines a Negative Binomial distribution with a binary gate variable. Observations are either drawn from the Negative Binomial distribution with probability (1 - gate) or are treated as zero with probability ‘gate’. This models data with excess zeros compared to what a standard Negative Binomial distribution would predict.</p>
<p>Let <span class="math inline">X</span> denote the count random variable, support $ {0,1,2,} $. Denote:</p>
<ul>
<li><span class="math inline">\pi</span> = probability of being in the “always zero” (structural zero) process, with $ 0 $.</li>
<li>The count process is NB with parameters: mean <span class="math inline">\mu&gt;0</span> and dispersion/shape parameter <span class="math inline">\alpha&gt;0</span> (or equivalently size parameter (r)). Then:</li>
</ul>
<p><span class="math display">
P(X = 0) = \pi ;+; (1 - \pi); P_{\text{NB}}(0;\mu,\alpha),
</span> <span class="math display">
P(X = k) = (1 - \pi); P_{\text{NB}}(k;\mu,\alpha), \quad k = 1,2,3,\dots
</span></p>
<p>Here <span class="math inline">P_{\text{NB}}(k;\mu,\alpha)</span> is the PMF of the negative binomial distribution for count <span class="math inline">k</span>.</p>
<section id="args-84" class="level4">
<h4 class="anchored" data-anchor-id="args-84">Args:</h4>
<div class="sourceCode" id="cb168"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb168-1"><a href="#cb168-1" aria-hidden="true" tabindex="-1"></a>bi.dist.zero_inflated_negative_binomial2(</span>
<span id="cb168-2"><a href="#cb168-2" aria-hidden="true" tabindex="-1"></a>mean,</span>
<span id="cb168-3"><a href="#cb168-3" aria-hidden="true" tabindex="-1"></a>concentration,</span>
<span id="cb168-4"><a href="#cb168-4" aria-hidden="true" tabindex="-1"></a>gate<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-5"><a href="#cb168-5" aria-hidden="true" tabindex="-1"></a>gate_logits<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-6"><a href="#cb168-6" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-7"><a href="#cb168-7" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb168-8"><a href="#cb168-8" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-9"><a href="#cb168-9" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-10"><a href="#cb168-10" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb168-11"><a href="#cb168-11" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb168-12"><a href="#cb168-12" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb168-13"><a href="#cb168-13" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb168-14"><a href="#cb168-14" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb168-15"><a href="#cb168-15" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb168-16"><a href="#cb168-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>mean</em> (jnp.ndarray or float): The mean of the Negative Binomial 2 distribution.</p></li>
<li><p><em>concentration</em> (jnp.ndarray or float): The concentration parameter of the Negative Binomial 2 distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-83" class="level4">
<h4 class="anchored" data-anchor-id="returns-83">Returns:</h4>
<ul>
<li>When <code>sample=False</code>: A BI ZeroInflatedNegativeBinomial2 distribution object (for model building).</li>
<li>When <code>sample=True</code>: A JAX array of samples drawn from the ZeroInflatedNegativeBinomial2 distribution (for direct sampling).</li>
<li>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</li>
</ul>
</section>
<section id="example-usage-81" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-81">Example Usage:</h4>
<div class="sourceCode" id="cb169"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>m.dist.zero_inflated_negative_binomial2(mean<span class="op">=</span><span class="fl">2.0</span>, concentration<span class="op">=</span><span class="fl">3.0</span>, gate <span class="op">=</span> <span class="fl">0.3</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflatednegativebinomial2" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflatednegativebinomial2">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#zeroinflatednegativebinomial2</h4>
</section>
<section id="references-40" class="level4">
<h4 class="anchored" data-anchor-id="references-40">References:</h4>
<ul>
<li>https://www.ewadirect.com/proceedings/tns/article/view/27624 “Social Networks Count Data: Negative Binomial Distributions and Extensions Versus Poisson and Bernoulli Models”</li>
<li>https://pmc.ncbi.nlm.nih.gov/articles/PMC7880198/ “Bayesian Zero-Inflated Negative Binomial Regression Based on Pólya-Gamma Mixtures - PMC”</li>
<li>https://search.r-project.org/CRAN/refmans/emdbook/html/dzinbinom.html “R: Zero-inflated negative binomial distribution”</li>
</ul>
<hr>
</section>
</section>
<section id="a-zero-inflated-poisson" class="level3">
<h3 class="anchored" data-anchor-id="a-zero-inflated-poisson">A Zero Inflated Poisson</h3>
<p>The Zero-Inflated Poisson distribution is a discrete count‐distribution designed for data with <em>more zeros</em> than would be expected under a standard Poisson. Essentially, it assumes two underlying processes:</p>
<ul>
<li>With probability <span class="math inline">\pi</span> you are in a “structural zero” state (i.e., you automatically get a zero count).</li>
<li>With probability <span class="math inline">1 - \pi</span> you draw from a standard Poisson distribution with parameter <span class="math inline">\lambda</span>.</li>
</ul>
<p>This results in a mixture distribution that places more mass at zero than a Poisson alone would. It’s widely used in, for instance, ecology (species counts with many zeros), insurance/claims problems, and any count‐data setting with excess zeros.</p>
<p>The probability density function (PDF) is given by:</p>
<p>Let <span class="math inline">X</span> be a random variable following a Zero-Inflated Poisson with parameters</p>
<ul>
<li><span class="math inline">\pi\in[0,1]</span>: zero-inflation probability (chance of being structural zero)</li>
<li><span class="math inline">\lambda&gt;0</span>: mean of the Poisson component</li>
</ul>
<p>Then: <span class="math display">
P(X = k) =
\begin{cases}
\pi ;+; (1 - \pi),e^{-\lambda},  \text{if } k = 0, [6pt]
(1 - \pi),\dfrac{\lambda^k e^{-\lambda}}{k!},  \text{if } k = 1,2,3,\dots
\end{cases}
</span><br>
(Here the Poisson component contributes also a mass at zero of <span class="math inline">(1 - \pi)e^{-\lambda}</span>, plus the extra <span class="math inline">\pi</span> mass.)</p>
<section id="args-85" class="level4">
<h4 class="anchored" data-anchor-id="args-85">Args:</h4>
<div class="sourceCode" id="cb170"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>bi.dist.zero_inflated_poisson(</span>
<span id="cb170-2"><a href="#cb170-2" aria-hidden="true" tabindex="-1"></a>gate,</span>
<span id="cb170-3"><a href="#cb170-3" aria-hidden="true" tabindex="-1"></a>rate<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb170-4"><a href="#cb170-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb170-5"><a href="#cb170-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb170-6"><a href="#cb170-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb170-7"><a href="#cb170-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb170-8"><a href="#cb170-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb170-9"><a href="#cb170-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb170-10"><a href="#cb170-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb170-11"><a href="#cb170-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb170-12"><a href="#cb170-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb170-13"><a href="#cb170-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb170-14"><a href="#cb170-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>gate</em> (jnp.ndarray): The probability of observing a zero.</p></li>
<li><p><em>rate</em> (jnp.ndarray): The rate parameter of the underlying Poisson distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-84" class="level4">
<h4 class="anchored" data-anchor-id="returns-84">Returns:</h4>
<p>BI ZeroInflatedPoisson distribution object (when <code>sample=False</code>). JAX array of samples drawn from the ZeroInflatedPoisson distribution (when <code>sample=True</code>). The raw BI distribution object (when <code>create_obj=True</code>).</p>
</section>
<section id="example-usage-82" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-82">Example Usage:</h4>
<div class="sourceCode" id="cb171"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb171-1"><a href="#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb171-2"><a href="#cb171-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb171-3"><a href="#cb171-3" aria-hidden="true" tabindex="-1"></a>m.dist.zero_inflated_poisson(gate <span class="op">=</span> <span class="fl">0.3</span>, rate<span class="op">=</span><span class="fl">2.0</span>, sample<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflatedpoisson" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-httpsnum.pyro.aienstabledistributions.htmlzeroinflatedpoisson">Wrapper of: https://num.pyro.ai/en/stable/distributions.html#zeroinflatedpoisson</h4>
</section>
<section id="references-41" class="level4">
<h4 class="anchored" data-anchor-id="references-41">References:</h4>
<ul>
<li>https://en.wikipedia.org/wiki/Zero-inflated_model “Zero-inflated model”</li>
<li>https://www.statisticshowto.com/zero-inflated-poisson-distribution “Zero-Inflated Poisson distribution - Statistics How To”</li>
</ul>
<hr>
</section>
</section>
<section id="zero-sum-normal" class="level3">
<h3 class="anchored" data-anchor-id="zero-sum-normal">Zero Sum Normal</h3>
<p>A <strong>zero-sum normal</strong> is a variant of a multivariate normal in which one (or more) linear constraint(s) force certain components to <strong>sum to zero</strong>. In practice, it’s used to model vectors of random effects (e.g.&nbsp;in hierarchical models) where the effects are constrained to sum to zero (to avoid overparameterization or enforce identifiability).</p>
<p><span class="math display">
ZSN(\sigma) = N(0, \sigma^2 (I - \tfrac{1}{n}J)) \\
\text{where} \ ~ J_{ij} = 1 \ ~ \text{and} \\
n = \text{number of zero-sum axes}
</span></p>
<p>where <span class="math inline">J</span> is the all-ones matrix and <span class="math inline">n</span> is the number of constrained elements. So the covariance is <span class="math inline">\sigma^2 (I - \tfrac{1}{n} J)</span>, which ensures that the sum of the components is 0 (since $ J = _i x_i $).</p>
<section id="args-86" class="level4">
<h4 class="anchored" data-anchor-id="args-86">Args:</h4>
<div class="sourceCode" id="cb172"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>bi.dist.zero_sum_normal(</span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>scale,</span>
<span id="cb172-3"><a href="#cb172-3" aria-hidden="true" tabindex="-1"></a>event_shape,</span>
<span id="cb172-4"><a href="#cb172-4" aria-hidden="true" tabindex="-1"></a>validate_args<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb172-5"><a href="#cb172-5" aria-hidden="true" tabindex="-1"></a>name<span class="op">=</span><span class="st">'x'</span>,</span>
<span id="cb172-6"><a href="#cb172-6" aria-hidden="true" tabindex="-1"></a>obs<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb172-7"><a href="#cb172-7" aria-hidden="true" tabindex="-1"></a>mask<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb172-8"><a href="#cb172-8" aria-hidden="true" tabindex="-1"></a>sample<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb172-9"><a href="#cb172-9" aria-hidden="true" tabindex="-1"></a>seed<span class="op">=</span><span class="va">None</span>,</span>
<span id="cb172-10"><a href="#cb172-10" aria-hidden="true" tabindex="-1"></a>shape<span class="op">=</span>(),</span>
<span id="cb172-11"><a href="#cb172-11" aria-hidden="true" tabindex="-1"></a>event<span class="op">=</span><span class="dv">0</span>,</span>
<span id="cb172-12"><a href="#cb172-12" aria-hidden="true" tabindex="-1"></a>create_obj<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb172-13"><a href="#cb172-13" aria-hidden="true" tabindex="-1"></a>to_jax<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb172-14"><a href="#cb172-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<ul>
<li><p><em>scale</em> (array_like): Standard deviation of the underlying normal distribution before the zerosum constraint is enforced.</p></li>
<li><p><em>event_shape</em> (tuple): Shape of the event dimension of the distribution.</p></li>
<li><p><em>shape</em> (tuple): A multi-purpose argument for shaping. When <code>sample=False</code> (model building), this is used with <code>shape</code> to set the distribution’s batch shape. When <code>sample=True</code> (direct sampling), this is used as <code>sample_shape</code> to draw a raw JAX array of the given shape.</p></li>
<li><p><em>event</em> (int): The number of batch dimensions to reinterpret as event dimensions (used in model building).</p></li>
<li><p><em>mask</em> (jnp.ndarray, bool): Optional boolean array to mask observations.</p></li>
<li><p><em>create_obj</em> (bool): If True, returns the raw BI distribution object instead of creating a sample site. This is essential for building complex distributions like <code>MixtureSameFamily</code>.</p></li>
<li><p><em>sample</em> (bool, optional): A control-flow argument. If <code>True</code>, the function will directly sample a raw JAX array from the distribution, bypassing the BI model context. If <code>False</code>, it will create a <code>BI.sample</code> site within a model. Defaults to <code>False</code>.</p></li>
<li><p><em>seed</em> (int, optional): An integer used to generate a JAX PRNGKey for reproducible sampling when <code>sample=True</code>. This argument has no effect when <code>sample=False</code>, as randomness is handled by BI’s inference engine. Defaults to 0.</p></li>
<li><p><em>obs</em> (jnp.ndarray, optional): The observed value for this random variable. If provided, the sample site is conditioned on this value, and the function returns the observed value. If <code>None</code>, the site is treated as a latent (unobserved) random variable. Defaults to <code>None</code>.</p></li>
<li><p><em>name</em> (str, optional): The name of the sample site in a BI model. This is used to uniquely identify the random variable. Defaults to ‘x’.</p></li>
</ul>
</section>
<section id="returns-85" class="level4">
<h4 class="anchored" data-anchor-id="returns-85">Returns:</h4>
<ul>
<li><p>When <code>sample=False</code>: A BI ZeroSumNormal distribution object (for model building).</p></li>
<li><p>When <code>sample=True</code>: A JAX array of samples drawn from the ZeroSumNormal distribution (for direct sampling).</p></li>
<li><p>When <code>create_obj=True</code>: The raw BI distribution object (for advanced use cases).</p></li>
</ul>
</section>
<section id="example-usage-83" class="level4">
<h4 class="anchored" data-anchor-id="example-usage-83">Example Usage:</h4>
<div class="sourceCode" id="cb173"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb173-1"><a href="#cb173-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> BI <span class="im">import</span> bi</span>
<span id="cb173-2"><a href="#cb173-2" aria-hidden="true" tabindex="-1"></a>m <span class="op">=</span> bi(<span class="st">'cpu'</span>)</span>
<span id="cb173-3"><a href="#cb173-3" aria-hidden="true" tabindex="-1"></a>m.dist.zero_sum_normal(scale<span class="op">=</span><span class="fl">1.0</span>, event_shape <span class="op">=</span> (<span class="dv">2</span>,), sample <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="wrapper-of-57" class="level4">
<h4 class="anchored" data-anchor-id="wrapper-of-57">Wrapper of:</h4>
<p>https://num.pyro.ai/en/stable/distributions.html#zerosumnormal</p>
</section>
<section id="references-42" class="level4">
<h4 class="anchored" data-anchor-id="references-42">References:</h4>
<p>https://www.pymc.io/projects/docs/en/latest/api/distributions/generated/pymc.ZeroSumNormal.html</p>
<hr>


</section>
</section>

</main> <!-- /main -->
<script src="https://unpkg.com/@popperjs/core@2"></script>
<script src="https://unpkg.com/tippy.js@6"></script>
<script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/marked/marked.min.js"></script>
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async="" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>

<script>
function convertMarkdownToHTML(markdown) {
  const options = {
    breaks: true,
    gfm: true,
  };
  const htmlContent = marked.parse(markdown, options);
  return htmlContent;
}

tippy('#Elasticity', {
  content: convertMarkdownToHTML('The elasticity of the response variable with respect to the exposure is the percent change in the response with respect to a percent change in the exposure. For example, an elasticity of 1.0 means that doubling the exposure time will double the response'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});


tippy('#Uncertainty', {
  content: convertMarkdownToHTML('Bayesian analysis provides a complete probability distribution for each parameter. This full distribution (rather than a single point estimate) allows you to assess uncertainty in a natural and intuitive way.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#priorKnowledge', {
  content: convertMarkdownToHTML('When data are sparse or noisy, incorporating prior knowledge through the prior distribution can improve the model’s performance. This is especially useful in fields where previous research or expert opinion is available.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#Flexibility', {
  content: convertMarkdownToHTML('Bayesian approaches naturally extend to complex settings such as hierarchical models, models with latent variables, or situations with missing data. Their ability to blend different sources of information makes them ideal for modern, data-rich applications.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#PPCs', {
  content: convertMarkdownToHTML('PPCs involve generating *dummy data* (or simulated observations) from the posterior predictive distribution and comparing these simulated datasets with the actual observed data.By comparing summary statistics (e.g., means, variances) or visualizations (e.g., histograms, scatter plots) between \( y^{\text{rep}} \) and the observed data, we can assess if our model is capable of reproducing key features of the data.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#HPDI', {
  content: convertMarkdownToHTML('HPDI is a metric that describe the distribution of a parameter. The Highest Posterior Density Interval is the narrowest interval that contains a specified proportion (e.g., 95%) of the posterior probability. HPDIs are especially useful when the posterior distribution is skewed or multimodal as it help in understanding the uncertainty of parameter estimates and in communicating the degree of uncertainty in predictions.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#convDiag', {
  content: convertMarkdownToHTML('Convergence diagnostics are statistical tools and methods used to assess whether a Markov Chain Monte Carlo (MCMC) algorithm has sufficiently explored the target posterior distribution in Bayesian analysis. They help ensure that the chain has stabilized and that samples are reliable for inference.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#ESS', {
  content: convertMarkdownToHTML('ESS estimates how many independent samples our MCMC chain is equivalent to. Low ESS values indicate that samples are highly correlated, which might affect the reliability of posterior summaries.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#Rhat', {
  content: convertMarkdownToHTML('Rhat compares the variance between multiple chains to the variance within each chain. Values close to 1 (typically below 1.1) indicate good convergence.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

  tippy('#TP', {
    content: convertMarkdownToHTML('Trace plots are graphical tools used in MCMC diagnostics to visualize the sampled values of a parameter over iterations. They help assess the convergence and mixing behavior of the Markov chain.'),
    allowHTML: true,
    interactive: true,
    delay: 100,
  });

  tippy('#AutoCor', {
    content: convertMarkdownToHTML('Autocorrelation plots visualize the correlation between samples in an MCMC chain as a function of the lag (distance between samples). They help assess how dependent the samples are and whether the chain is mixing efficiently.'),
    allowHTML: true,
    interactive: true,
    delay: 100,
  });

  tippy('#DesPlot', {
    content: convertMarkdownToHTML('Density plots visualize the estimated posterior distribution of a parameter based on MCMC samples. They provide a smoothed representation of the distribution, helping assess central tendency, spread, and shape.'),
    allowHTML: true,
    interactive: true,
    delay: 100,
  });



tippy('#DP', {
  content: convertMarkdownToHTML('A dependent variable  is the variable being measured or observed in an analysis. It represents the outcome whose variation depends on changes in one or morepredictor. In statistical models, it is often denoted as Y.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#IP', {
  content: convertMarkdownToHTML('An independent variable is the variable that is manipulated or controlled in an analysis to observe its effect on a dependent variable. It represents the input or cause in a cause-and-effect relationship and is often denoted as X in statistical models.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#lk', {
  content: convertMarkdownToHTML('Likelihood is a statistical concept representing the probability of observing a given set of data, assuming specific values of the *parameters* in a model.It is not the same as probability; instead, it measures how well a model with certain parameters explains the observed data.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#pdist', {
  content: convertMarkdownToHTML('A probability distribution is a mathematical function that describes the likelihood of different outcomes in a random process. It specifies how probabilities are assigned to values or ranges of a random variable. Distributions can be discrete (e.g.,Binomial, Poisson) or continuous (e.g., Normal, Exponential), depending on whether the variable takes finite or infinite values.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#intercept ', {
  content: convertMarkdownToHTML('An intercept is the value of the dependent variable when all independent variables are equal to zero in a regression model. It represents the starting point or baseline level of the outcome variable and is often denoted as α in linear models.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#slope ', {
  content: convertMarkdownToHTML('A slope is the rate of change in the dependent variable for a one-unit change in an independent variable in a regression model. It represents the strength and direction of the relationship and is often denoted as β in linear models. A positive slope indicates an increasing relationship, while a negative slope indicates a decreasing relationship.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});


tippy('#prior', {
  content: convertMarkdownToHTML('Priors are distributions that represent the value of the parameter before conditioning it on the data.'),
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#BT', {
  content: '<span style="color: white;">Bayes\' Theorem describes how to update the probability of a hypothesis given new evidence. It states that the posterior probability of a parameter is proportional to the product of the prior probability and the likelihood of the data given the parameter. This approach allows for the incorporation of prior knowledge and the use ofBayesian updating to improve the accuracy of the model.</span>',
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#BUpdating', {
    content: '<span style="color: white;">Bayesian updating is the iterative process of refining beliefs by applying Bayes\' Theorem as new data becomes available.</span>',
    allowHTML: true,
    interactive: true,
    delay: 100,
  });



tippy('#overdispersion', {
  content: '<span style="color: white;">Overdispersion refers to a situation in statistical modeling where the variability of the data exceeds what is expected under the assumed model. In this case, by the binomial distribution.</span>',
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#linkF', {
    content: '<span style="color: white;">The link function is a key component of the generalized linear model (GLM) and it is used to transform the mean of the response variable so that it can be modeled as a linear combination of the predictors. Different types of response data(continuous, binary, counts, etc.) have different natural ranges. For example, probabilities must lie between 0 and 1, and a logarithmic link ensures a positive mean for count data. The choice of link function influences how we interpret the regression coefficients.For instance, with a logit link (used in logistic regression), the coefficients relate to the log - odds of the outcome.</span>',
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#simplex', {
  content: '<span style="color: white;">A simplex is a vector of k−1 probabilities that sum to one.</span>',
  allowHTML: true,
  interactive: true,
  delay: 100,
});
  
tippy('#triptych', {
    content: '<span style="color: white;">Three panels arranged side by side to compare multiple datasets or conditions simultaneously. Each panel often represents a different aspect of the data or a different dataset, allowing for easy comparison and analysis. They are particularly useful for displaying complex relationships or patterns across multiple variables or experimental conditions.") plots are very handy for understanding the impact of interactions.</span>',
    allowHTML: true,
    interactive: true,
    delay: 100,
});
    
tippy('#hyperP', {
  content: '<span style="color: white;">A hyperprior is a prior distribution placed on the parameters (called hyperparameters) of another prior distribution. Essentially, it introduces a second layer of uncertainty by modeling the parameters of the prior as random variables, allowing for more flexibility and robustness in the Bayesian model.</span>',
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#cov', {
    content: '<span style="color: white;">A covariance matrix is a square matrix that contains the covariances between pairs of elements in a random vector. Each element in the matrix represents the covariance between two variables. The diagonal elements represent the variances of each variable, and the off-diagonal elements represent the covariance between different variables.</span>',
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#centerRF', {
  content: "<span style='color: white;'>The centered version of the varying intercept model is a statistical approach where the intercepts of a regression model vary across groups or clusters, and these intercepts are centered around a global mean. In this model, each group has its own intercept, but the differences between these intercepts and the global mean are what are explicitly estimated. This helps in understanding how each group's baseline differs from an overall average, making it easier to interpret variations between groups.</span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#NodeF', {
    content: "<span style='color: white;'>Nodal categorical (e.g., sex) or continuous (e.g., age) characteristics</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#DyadicF', {
    content: "<span style='color: white;'>Link-related categorical (e.g., same group membership) or continuous (e.g., genetic distances) characteristics between nodes</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#senderF', {
    content: "<span style='color: white;'>Node <i>i</i>'s propensity to emit a link.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#receiverF', {
    content: "<span style='color: white;'>Node <i>j</i>'s propensity to receive a link.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#DyadicF2', {
    content: "<span style='color: white;'>Propensity of node <i>i</i> and <i>j</i> to receive from each other.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#expoB', {
    content: "<span style='color: white;'>Heterogeneity in observations between individuals leading to over- and under-sampling of some individuals.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#censoB', {
    content: "<span style='color: white;'>Missing interactions within observation leading to over- and under-estimation of interactions (e.g., Females birds may have cryptic phenotype compare to male birds, leading to higher probability to miss interactions during observations).</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#assor', {
    content: "<span style='color: white;'>The tendency of nodes to connect with similar nodes.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});

tippy('#disassor', {
    content: "<span style='color: white;'>The tendency of nodes to connect with dissimilar nodes.</span>",
    allowHTML: true,
    interactive: true,
    delay: 100,
});


tippy('#chol', {
  content: "<span style='color: white;'>This matrix is special because it encodes correlations between the variables but in a computationally stable way.In multivariate normal distributions, correlation matrices are positive definite and symmetric, but they can be tricky to work with directly. The Cholesky decomposition makes computation easier and more stable because it's lower triangular and simplifies matrix operations.</span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#uncertainty', {
  content: "<span style='color: white;'>Model parameter uncertainty refers to the inherent variability in the parameter estimates due to limited or noisy data, represented as probability distributions rather than fixed values. </span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#posterior', {
  content: "<span style='color: white;'> Probability distribution of the model parameters after observing the data. </span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#scaled', {
  content: "<span style='color: white;'> Scaled data is data that has been transformed to fit within a specific range or to have specific statistical properties, such as a normal distribution with a mean of zero and a standard deviation of one. This transformation helps improve the performance of algorithms and ensures that features contribute equally to the analysis. </span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#grandMean', {
  content: "<span style='color: white;'> The average value of the outcome when all predictor variables are at their mean levels. </span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#ohe', {
  content: "<span style='color: white;'>Convert categorical variables into binary vectors, where each category is represented by a unique vector with all elements set to zero except for one element, which is set to one.</span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#indices', {
  content: "<span style='color: white;'>Process of mapping each unique category of a categorical variable to a corresponding integer index, where each category is assigned a distinct numerical value.</span>",
  allowHTML: true,
  interactive: true,
  delay: 100,
});

tippy('#softmax', {
  content: `
  <span style='color: white;'>
    takes a vector of 𝑛 real-valued scores (logits) and exponentiates each score, 
    then normalizes them by dividing each exponentiated score by the sum of all exponentiated scores. 
    The resulting output values lie in the range (0, 1) and sum to 1, making them interpretable as probabilities.<br><br>
    <strong>Formula:</strong><br>
    $$softmax(z<sub>i</sub>) = [ \\frac{e^{z_i}}{\\sum_{j=1}^{n} e^{z_j}} ]$$
  </span>
  `,  allowHTML: true,
  interactive: true,
  delay: 100,
  onShow(instance) {
    MathJax.typesetPromise([instance.reference]).then(() => {
      // Optional: if you want to do something after the typesetting
    });
  },
});

tippy('#kernel', {
  content: `
  <span style='color: white;'>
    A kernel k(x,x') is a function that computes the similarity/covariance between any two points x and x'
    in a higher-dimensional feature space without explicitly transforming them. 
    The kernel function allows for non-linear relationships by mapping input space to feature space, 
    where linear techniques can then be applied. 
    Mathematically, like for the covariance matrix it must be symmetric and positive semi-definite.
  </span>
  `,  allowHTML: true,
  interactive: true,
  delay: 100,
  onShow(instance) {
    MathJax.typesetPromise([instance.reference]).then(() => {
      // Optional: if you want to do something after the typesetting
    });
  },
});

tippy('#kernelMean0', {
  content: `
  <span style='color: white;'>
    GP mean is usually set to zero as by setting the mean to zero, 
    we center the distribution around the origin. 
    This helps in interpreting the model, as the values of <i>𝑓(𝑥)<\\i> 
    can be thought of as deviations from this mean.
  </span>
  `,  allowHTML: true,
  interactive: true,
  delay: 100,
  onShow(instance) {
    MathJax.typesetPromise([instance.reference]).then(() => {
      // Optional: if you want to do something after the typesetting
    });
  },
});

tippy('#signAmbiguity', {
  content: `
  <span style='color: white;'>
    PCA identifies directions (principal components) that capture the maximum variance in the data. However, the orientation of these components is not unique; both a vector and its negation represent the same principal direction. This means that if 𝑤 is a principal component, −𝑤 is equally valid.
  </span>
  `,  allowHTML: true,
  interactive: true,
  delay: 100,
  onShow(instance) {
    MathJax.typesetPromise([instance.reference]).then(() => {
      // Optional: if you want to do something after the typesetting
    });
  },
});

tippy('#MatCov', {
  content: `
  <span style='color: white;'>
    A covariance matrix is a square matrix Σ of size $n×n$ for n random variables $X₁, X₂, ..., Xₙ$, where each element $σ_ij$ represents the covariance between variables $X_i$ and $X_j$. The diagonal elements $σ_ii$ are the variances of each variable $X_i$.
  </span>
  `, allowHTML: true,
    interactive: true,
    delay: 100,
    onShow(instance) {
      MathJax.typesetPromise([instance.reference]).then(() => {
        // Optional: if you want to do something after the typesetting
      });
    },
});

tippy('#RFcentered', {
  content: `
  <span style='color: white;'>
    The centered version of the varying intercept in mixed effects models is a transformation where each group's intercept is expressed as a deviation from the fixed intercept. This results in zero-mean random effects ensures that the random effects are balanced around zero, enhancing model interpretability and estimation efficiency.
  </span>
  `, allowHTML: true,
    interactive: true,
    delay: 100,
    onShow(instance) {
      MathJax.typesetPromise([instance.reference]).then(() => {
        // Optional: if you want to do something after the typesetting
      });
    },
});

tippy('#LKJcorr', {
      content: `
  <span style='color: white;'>
    The Cholesky Factor is a lower triangular matrix ( $L$ ) with positive diagonal entries such that $( A = LL^T )$ for a given symmetric positive-definite matrix ( $A$ ).
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#unsupervised', {
      content: `
  <span style='color: white;'>
    An unsupervised algorithm is a method that processes unlabeled data to automatically uncover patterns, groupings, or structures without predefined output targets.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#stickProcess', {
      content: `
  <span style='color: white;'>
The stick-breaking process is a way to generate a random distribution by imagining a unit-length stick that gets broken into pieces one step at a time. At each step, a random fraction of whatever stick remains is snapped off to form a new weight, and the rest is left for future breaks. Each weight is paired with a randomly chosen parameter value. Repeating this indefinitely yields a distribution made of infinitely many weighted parameter, where earlier pieces tend to be larger and later ones smaller.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#Neurons', {
      content: `
  <span style='color: white;'>
 In a neural network, a neuron (or node) is a fundamental computational unit. It receives inputs, applies a weighted sum and an activation function to these inputs, and produces an output. It's analogous to a biological neuron, processing and transmitting signals.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#Layers', {
      content: `
  <span style='color: white;'>
Layers are groups of neurons organized in a 2D array. Each layer processes the output from the previous layer and passes its output to the next. Common layer types include input layers, hidden layers (where the main computation occurs), and output layers (which produce the final predictions).
</span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });


tippy('#Dropout', {
      content: `
  <span style='color: white;'>
 Dropout is a regularization technique where, during training, a random subset of neurons (and their connections) are temporarily ignored or "dropped out" for each training iteration. This prevents neurons from co-adapting too much and encourages the network to learn more robust features, thus reducing overfitting.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#L2', {
      content: `
  <span style='color: white;'>
L2 regularization (also known as weight decay) is a technique to prevent overfitting by adding a penalty term to the loss function. This penalty is proportional to the sum of the squares of the network's weights. It encourages the model to use smaller weights, leading to a simpler and more generalized model.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });


tippy('#Overfitting', {
      content: `
  <span style='color: white;'>
Overfitting occurs when a machine learning model learns the training data too well, including its noise and random fluctuations. As a result, the model performs very poorly on new, unseen data because it has failed to generalize the underlying patterns.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });

tippy('#VI', {
      content: `
  <span style='color: white;'>
An optimization-based method for approximating complex, intractable posterior distributions. Instead of sampling (as in MCMC), VI posits a parameterized family of simpler distributions (e.g., Gaussians) and optimizes their parameters to minimize the Kullback-Leibler (KL) divergence between the approximation and the true posterior. This effectively converts an integration problem into an optimization problem
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });
    
tippy('#SG-MCMC', {
      content: `
  <span style='color: white;'>
A class of algorithms designed to scale Markov Chain Monte Carlo methods to large datasets. It approximates the gradient of the log-posterior using mini-batches (subsets) of data rather than the full dataset. To ensure convergence to the true posterior distribution despite the noise introduced by subsampling, these methods (such as Stochastic Gradient Langevin Dynamics) introduce specific noise injection and step-size schedules.
  </span>
  `, allowHTML: true,
      interactive: true,
      delay: 100,
      onShow(instance) {
        MathJax.typesetPromise([instance.reference]).then(() => {
          // Optional: if you want to do something after the typesetting
        });
      },
    });
</script>
<script>
    // Disable Quarto code folding scripts globally
    window.quartoDisableCodeFolding = true;
  </script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = true;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./api_manip.html" class="pagination-link" aria-label="Handle data">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Handle data</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./api_diag.html" class="pagination-link" aria-label="Diagnostics">
        <span class="nav-page-text">Diagnostics</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>